// waddler/sqlite/op-sqlite/session.js
import { WaddlerQueryError } from "../../errors/index.js";
import { SQLTemplate } from "../../sql-template.js";
class OpSqliteSQLTemplate extends SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
  }
  returningData = true;
  all() {
    this.returningData = true;
    return this;
  }
  run() {
    this.returningData = false;
    return this;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      if (this.options.rowMode === "array") {
        finalRes = await this.client.executeRaw(query, params);
      } else {
        const queryResult = await this.client.execute(query, params);
        ({ rows: finalRes, ...finalMetadata } = queryResult);
      }
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  /**
   * For now, throws the Error.
   * Current implementation (a placeholder) will be replaced once cloudflare op-sqlite driver acquires streaming support or when a suitable third-party solution is found.
   */
  // eslint-disable-next-line require-yield
  async *stream() {
    throw new Error("stream is not implemented for op-sqlite yet.");
  }
}
export {
  OpSqliteSQLTemplate
};
