// waddler/sqlite/bun-sqlite/session.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var session_exports = {};
__export(session_exports, {
  BunSqliteSQLTemplate: () => BunSqliteSQLTemplate
});
module.exports = __toCommonJS(session_exports);
var import_errors = require("../../errors/index.cjs");
var import_sql_template = require("../../sql-template.cjs");
class BunSqliteSQLTemplate extends import_sql_template.SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
  }
  returningData = true;
  all() {
    this.returningData = true;
    return this;
  }
  run() {
    this.returningData = false;
    return this;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      const stmt = this.client.prepare(query);
      if (this.returningData) {
        finalRes = this.options.rowMode === "array" ? stmt.values(...params) : stmt.all(...params);
      } else {
        finalRes = stmt.run(...params);
        finalMetadata = finalRes;
      }
    } catch (error) {
      throw new import_errors.WaddlerQueryError(query, params, error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  async *stream() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    this.logger.logQuery(query, params);
    try {
      const stmt = this.client.prepare(query);
      const stream = stmt.iterate(...params);
      for await (const row of stream) {
        if (this.options.rowMode === "array") {
          const rowValues = stmt.columnNames.map((colName) => row[colName]);
          yield rowValues;
        } else {
          yield row;
        }
      }
    } catch (error) {
      throw new import_errors.WaddlerQueryError(query, params, error);
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BunSqliteSQLTemplate
});
