// waddler/mysql/mysql2/session.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var session_exports = {};
__export(session_exports, {
  MySql2SQLTemplate: () => MySql2SQLTemplate
});
module.exports = __toCommonJS(session_exports);
var import_errors = require("../../errors/index.cjs");
var import_sql_template = require("../../sql-template.cjs");
var import_utils = require("./utils.cjs");
class MySql2SQLTemplate extends import_sql_template.SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }, queryConfig = {
    sql: sqlWrapper.getQuery(dialect).sql
  }, rawQueryConfig = {
    sql: sqlWrapper.getQuery(dialect).sql,
    rowsAsArray: true
  }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
    this.queryConfig = queryConfig;
    this.rawQueryConfig = rawQueryConfig;
  }
  async execute() {
    const { params } = this.sqlWrapper.getQuery(this.dialect);
    let finalResult;
    let finalMetadata;
    try {
      const queryResult = await (this.options.rowMode === "array" ? this.client.query(this.rawQueryConfig, params) : this.client.query(this.queryConfig, params));
      finalResult = queryResult[0];
      finalMetadata = queryResult[1];
    } catch (error) {
      throw new import_errors.WaddlerQueryError(this.queryConfig.sql, params, error);
    }
    this.logger.logQuery(this.queryConfig.sql, params, finalMetadata);
    return finalResult;
  }
  async *stream() {
    let conn;
    const { params } = this.sqlWrapper.getQuery(this.dialect);
    try {
      const conn2 = ((0, import_utils.isPool)(this.client) ? await this.client.getConnection() : this.client).connection;
      const stream = conn2.query(this.queryConfig, params).stream();
      for await (const row of stream) {
        yield row;
      }
      if ((0, import_utils.isPool)(this.client)) {
        conn2.release();
      }
    } catch (error) {
      if ((0, import_utils.isPool)(this.client)) {
        conn?.release();
      }
      throw new import_errors.WaddlerQueryError(this.queryConfig.sql, params, error);
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MySql2SQLTemplate
});
