// waddler/mysql/mysql2/session.js
import { WaddlerQueryError } from "../../errors/index.js";
import { SQLTemplate } from "../../sql-template.js";
import { isPool } from "./utils.js";
class MySql2SQLTemplate extends SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }, queryConfig = {
    sql: sqlWrapper.getQuery(dialect).sql
  }, rawQueryConfig = {
    sql: sqlWrapper.getQuery(dialect).sql,
    rowsAsArray: true
  }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
    this.queryConfig = queryConfig;
    this.rawQueryConfig = rawQueryConfig;
  }
  async execute() {
    const { params } = this.sqlWrapper.getQuery(this.dialect);
    let finalResult;
    let finalMetadata;
    try {
      const queryResult = await (this.options.rowMode === "array" ? this.client.query(this.rawQueryConfig, params) : this.client.query(this.queryConfig, params));
      finalResult = queryResult[0];
      finalMetadata = queryResult[1];
    } catch (error) {
      throw new WaddlerQueryError(this.queryConfig.sql, params, error);
    }
    this.logger.logQuery(this.queryConfig.sql, params, finalMetadata);
    return finalResult;
  }
  async *stream() {
    let conn;
    const { params } = this.sqlWrapper.getQuery(this.dialect);
    try {
      const conn2 = (isPool(this.client) ? await this.client.getConnection() : this.client).connection;
      const stream = conn2.query(this.queryConfig, params).stream();
      for await (const row of stream) {
        yield row;
      }
      if (isPool(this.client)) {
        conn2.release();
      }
    } catch (error) {
      if (isPool(this.client)) {
        conn?.release();
      }
      throw new WaddlerQueryError(this.queryConfig.sql, params, error);
    }
  }
}
export {
  MySql2SQLTemplate
};
