// waddler/clickhouse/session.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var session_exports = {};
__export(session_exports, {
  ClickHouseSQLTemplate: () => ClickHouseSQLTemplate
});
module.exports = __toCommonJS(session_exports);
var import_errors = require("../errors/index.cjs");
var import_sql_template = require("../sql-template.cjs");
class ClickHouseSQLTemplate extends import_sql_template.SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.dialect = dialect;
    this.options = options;
  }
  returningData = true;
  command() {
    this.returningData = false;
    return this;
  }
  query() {
    this.returningData = true;
    return this;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes;
    let finalMetadata;
    try {
      if (this.returningData) {
        const format = this.options.rowMode === "object" ? "JSON" : "JSONCompact";
        const rawRes = await this.client.query({
          query,
          format,
          query_params: params
        });
        const jsonRes = await rawRes.json();
        const { data, ...metadata } = jsonRes;
        finalRes = data;
        finalMetadata = metadata;
      } else {
        const rawRes = await this.client.command({
          query,
          query_params: params
        });
        finalRes = rawRes;
        finalMetadata = rawRes;
      }
    } catch (error) {
      throw new import_errors.WaddlerQueryError(query, JSON.stringify(params), error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  async *stream() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    try {
      const format = this.options.rowMode === "object" ? "JSONEachRow" : "JSONCompactEachRow";
      const rawRes = await this.client.query({ query, format, query_params: params });
      const stream = rawRes.stream();
      for await (const chunk of stream) {
        for (const row of chunk) {
          yield await row.json();
        }
      }
    } catch (error) {
      throw new import_errors.WaddlerQueryError(query, params, error);
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ClickHouseSQLTemplate
});
