// waddler/clickhouse/session.js
import { WaddlerQueryError } from "../errors/index.js";
import { SQLTemplate } from "../sql-template.js";
class ClickHouseSQLTemplate extends SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.dialect = dialect;
    this.options = options;
  }
  returningData = true;
  command() {
    this.returningData = false;
    return this;
  }
  query() {
    this.returningData = true;
    return this;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes;
    let finalMetadata;
    try {
      if (this.returningData) {
        const format = this.options.rowMode === "object" ? "JSON" : "JSONCompact";
        const rawRes = await this.client.query({
          query,
          format,
          query_params: params
        });
        const jsonRes = await rawRes.json();
        const { data, ...metadata } = jsonRes;
        finalRes = data;
        finalMetadata = metadata;
      } else {
        const rawRes = await this.client.command({
          query,
          query_params: params
        });
        finalRes = rawRes;
        finalMetadata = rawRes;
      }
    } catch (error) {
      throw new WaddlerQueryError(query, JSON.stringify(params), error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  async *stream() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    try {
      const format = this.options.rowMode === "object" ? "JSONEachRow" : "JSONCompactEachRow";
      const rawRes = await this.client.query({ query, format, query_params: params });
      const stream = rawRes.stream();
      for await (const chunk of stream) {
        for (const row of chunk) {
          yield await row.json();
        }
      }
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    }
  }
}
export {
  ClickHouseSQLTemplate
};
