// drizzle-orm/pg-core/effect/session.js
import { PgBasePreparedQuery, PgSession } from "../session.js";
import { PgEffectDatabase } from "./db.js";
import { entityKind, is } from "../../entity.js";
import { assertUnreachable } from "../../utils.js";
import { fillPlaceholders, sql } from "../../sql/sql.js";
import { getMigrationsToRun } from "../../migrator.utils.js";
import { NoopCache, strategyFor } from "../../cache/core/cache.js";
import * as Effect from "effect/Effect";
import * as Cause from "effect/Cause";
import { EffectCache } from "../../cache/core/cache-effect.js";
import { EffectDrizzleQueryError, EffectTransactionRollbackError, MigratorInitError } from "../../effect-core/errors.js";
import { upgradeIfNeeded } from "../../up-migrations/effect-pg.js";

//#region src/pg-core/effect/session.ts
var PgEffectPreparedQuery = class extends PgBasePreparedQuery {
	static [entityKind] = "PgEffectPreparedQuery";
	/** @internal */
	mapper;
	constructor(executor, query, mapper, mode, logger, cache, queryMetadata, cacheConfig) {
		super(query);
		this.executor = executor;
		this.mode = mode;
		this.logger = logger;
		this.cache = cache;
		this.queryMetadata = queryMetadata;
		this.cacheConfig = cacheConfig;
		this.mapper = mapper;
		if (cache && cache.strategy() === "all" && cacheConfig === void 0) this.cacheConfig = {
			enabled: true,
			autoInvalidate: true
		};
		if (!this.cacheConfig?.enabled) this.cacheConfig = void 0;
	}
	execute(placeholderValues = {}) {
		return Effect.gen({ self: this }, function* () {
			const params = fillPlaceholders(this.query.params, placeholderValues);
			const { query: { sql }, mapper, logger } = this;
			yield* logger.logQuery(sql, params);
			const query = this.queryWithCache(sql, params, Effect.suspend(() => this.executor(params)));
			if (!mapper) return yield* query;
			return yield* query.pipe(Effect.map((rows) => mapper(rows)));
		});
	}
	queryWithCache(queryString, params, query) {
		return Effect.gen({ self: this }, function* () {
			const { cacheConfig, queryMetadata } = this;
			const cache = yield* EffectCache;
			const cacheStrat = cache && !is(cache.cache, NoopCache) ? yield* Effect.tryPromise(() => strategyFor(queryString, params, queryMetadata, cacheConfig)) : { type: "skip" };
			if (cacheStrat.type === "skip") return yield* query;
			if (cacheStrat.type === "invalidate") {
				const result = yield* query;
				yield* cache.onMutate({ tables: cacheStrat.tables });
				return result;
			}
			if (cacheStrat.type === "try") {
				const { tables, key, isTag, autoInvalidate, config } = cacheStrat;
				const fromCache = yield* cache.get(key, tables, isTag, autoInvalidate);
				if (typeof fromCache !== "undefined") return fromCache;
				const result = yield* query;
				yield* cache.put(key, result, autoInvalidate ? tables : [], isTag, config);
				return result;
			}
			assertUnreachable(cacheStrat);
		}).pipe(Effect.provideService(EffectCache, this.cache), Effect.catch((e) => {
			return Effect.fail(new EffectDrizzleQueryError({
				query: queryString,
				params,
				cause: Cause.fail(e)
			}));
		}));
	}
};
var PgEffectSession = class extends PgSession {
	static [entityKind] = "PgEffectSession";
	constructor(dialect) {
		super(dialect);
	}
	execute(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "raw", false).execute();
	}
	arrays(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "arrays", false).execute();
	}
	objects(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "objects", false).execute();
	}
};
var PgEffectTransaction = class extends PgEffectDatabase {
	static [entityKind] = "PgEffectTransaction";
	constructor(dialect, session, relations, nestedIndex = 0, parseRqbJson) {
		super(dialect, session, relations, parseRqbJson);
		this.relations = relations;
		this.nestedIndex = nestedIndex;
	}
	rollback() {
		return new EffectTransactionRollbackError();
	}
	/** @internal */
	getTransactionConfigSQL(config) {
		const chunks = [];
		if (config.isolationLevel) chunks.push(`isolation level ${config.isolationLevel}`);
		if (config.accessMode) chunks.push(config.accessMode);
		if (typeof config.deferrable === "boolean") chunks.push(config.deferrable ? "deferrable" : "not deferrable");
		return sql.raw(chunks.join(" "));
	}
	setTransaction(config) {
		return this.session.execute(sql`set transaction ${this.getTransactionConfigSQL(config)}`);
	}
};
const migrate = Effect.fn("migrate")(function* (migrations, session, config) {
	const migrationsTable = typeof config === "string" ? "__drizzle_migrations" : config.migrationsTable ?? "__drizzle_migrations";
	const migrationsSchema = typeof config === "string" ? "drizzle" : config.migrationsSchema ?? "drizzle";
	yield* session.execute(sql`CREATE SCHEMA IF NOT EXISTS ${sql.identifier(migrationsSchema)}`);
	const { newDb } = yield* upgradeIfNeeded(migrationsSchema, migrationsTable, session, migrations);
	if (newDb) {
		const migrationTableCreate = sql`
			CREATE TABLE IF NOT EXISTS ${sql.identifier(migrationsSchema)}.${sql.identifier(migrationsTable)} (
				id SERIAL PRIMARY KEY,
				hash text NOT NULL,
				created_at bigint,
				name text,
				applied_at timestamp with time zone DEFAULT now()
			)
		`;
		yield* session.execute(migrationTableCreate);
	}
	const dbMigrations = yield* session.objects(sql`select id, hash, created_at, name from ${sql.identifier(migrationsSchema)}.${sql.identifier(migrationsTable)}`);
	if (typeof config === "object" && config.init) {
		if (dbMigrations.length) return yield* new MigratorInitError({ exitCode: "databaseMigrations" });
		if (migrations.length > 1) return yield* new MigratorInitError({ exitCode: "localMigrations" });
		const [migration] = migrations;
		if (!migration) return;
		yield* session.execute(sql`insert into ${sql.identifier(migrationsSchema)}.${sql.identifier(migrationsTable)} ("hash", "created_at", "name") values(${migration.hash}, ${migration.folderMillis}, ${migration.name})`);
		return;
	}
	const migrationsToRun = getMigrationsToRun({
		localMigrations: migrations,
		dbMigrations
	});
	yield* session.transaction((tx) => Effect.gen(function* () {
		for (const migration of migrationsToRun) {
			for (const stmt of migration.sql) yield* tx.execute(sql.raw(stmt));
			yield* tx.execute(sql`insert into ${sql.identifier(migrationsSchema)}.${sql.identifier(migrationsTable)} ("hash", "created_at", "name") values(${migration.hash}, ${migration.folderMillis}, ${migration.name})`);
		}
	}));
});

//#endregion
export { PgEffectPreparedQuery, PgEffectSession, PgEffectTransaction, migrate };
