// drizzle-orm/pg-core/effect/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_pg_core_session = require('../session.cjs');
const require_pg_core_effect_db = require('./db.cjs');
let __entity_ts = require("../../entity.cjs");
let __utils_ts = require("../../utils.cjs");
let __sql_sql_ts = require("../../sql/sql.cjs");
let __migrator_utils_ts = require("../../migrator.utils.cjs");
let __cache_core_cache_ts = require("../../cache/core/cache.cjs");
let effect_Effect = require("effect/Effect");
effect_Effect = require_runtime.__toESM(effect_Effect);
let effect_Cause = require("effect/Cause");
effect_Cause = require_runtime.__toESM(effect_Cause);
let __cache_core_cache_effect_ts = require("../../cache/core/cache-effect.cjs");
let __effect_core_errors_ts = require("../../effect-core/errors.cjs");
let __up_migrations_effect_pg_ts = require("../../up-migrations/effect-pg.cjs");

//#region src/pg-core/effect/session.ts
var PgEffectPreparedQuery = class extends require_pg_core_session.PgBasePreparedQuery {
	static [__entity_ts.entityKind] = "PgEffectPreparedQuery";
	/** @internal */
	mapper;
	constructor(executor, query, mapper, mode, logger, cache, queryMetadata, cacheConfig) {
		super(query);
		this.executor = executor;
		this.mode = mode;
		this.logger = logger;
		this.cache = cache;
		this.queryMetadata = queryMetadata;
		this.cacheConfig = cacheConfig;
		this.mapper = mapper;
		if (cache && cache.strategy() === "all" && cacheConfig === void 0) this.cacheConfig = {
			enabled: true,
			autoInvalidate: true
		};
		if (!this.cacheConfig?.enabled) this.cacheConfig = void 0;
	}
	execute(placeholderValues = {}) {
		return effect_Effect.gen({ self: this }, function* () {
			const params = (0, __sql_sql_ts.fillPlaceholders)(this.query.params, placeholderValues);
			const { query: { sql }, mapper, logger } = this;
			yield* logger.logQuery(sql, params);
			const query = this.queryWithCache(sql, params, effect_Effect.suspend(() => this.executor(params)));
			if (!mapper) return yield* query;
			return yield* query.pipe(effect_Effect.map((rows) => mapper(rows)));
		});
	}
	queryWithCache(queryString, params, query) {
		return effect_Effect.gen({ self: this }, function* () {
			const { cacheConfig, queryMetadata } = this;
			const cache = yield* __cache_core_cache_effect_ts.EffectCache;
			const cacheStrat = cache && !(0, __entity_ts.is)(cache.cache, __cache_core_cache_ts.NoopCache) ? yield* effect_Effect.tryPromise(() => (0, __cache_core_cache_ts.strategyFor)(queryString, params, queryMetadata, cacheConfig)) : { type: "skip" };
			if (cacheStrat.type === "skip") return yield* query;
			if (cacheStrat.type === "invalidate") {
				const result = yield* query;
				yield* cache.onMutate({ tables: cacheStrat.tables });
				return result;
			}
			if (cacheStrat.type === "try") {
				const { tables, key, isTag, autoInvalidate, config } = cacheStrat;
				const fromCache = yield* cache.get(key, tables, isTag, autoInvalidate);
				if (typeof fromCache !== "undefined") return fromCache;
				const result = yield* query;
				yield* cache.put(key, result, autoInvalidate ? tables : [], isTag, config);
				return result;
			}
			(0, __utils_ts.assertUnreachable)(cacheStrat);
		}).pipe(effect_Effect.provideService(__cache_core_cache_effect_ts.EffectCache, this.cache), effect_Effect.catch((e) => {
			return effect_Effect.fail(new __effect_core_errors_ts.EffectDrizzleQueryError({
				query: queryString,
				params,
				cause: effect_Cause.fail(e)
			}));
		}));
	}
};
var PgEffectSession = class extends require_pg_core_session.PgSession {
	static [__entity_ts.entityKind] = "PgEffectSession";
	constructor(dialect) {
		super(dialect);
	}
	execute(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "raw", false).execute();
	}
	arrays(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "arrays", false).execute();
	}
	objects(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "objects", false).execute();
	}
};
var PgEffectTransaction = class extends require_pg_core_effect_db.PgEffectDatabase {
	static [__entity_ts.entityKind] = "PgEffectTransaction";
	constructor(dialect, session, relations, nestedIndex = 0, parseRqbJson) {
		super(dialect, session, relations, parseRqbJson);
		this.relations = relations;
		this.nestedIndex = nestedIndex;
	}
	rollback() {
		return new __effect_core_errors_ts.EffectTransactionRollbackError();
	}
	/** @internal */
	getTransactionConfigSQL(config) {
		const chunks = [];
		if (config.isolationLevel) chunks.push(`isolation level ${config.isolationLevel}`);
		if (config.accessMode) chunks.push(config.accessMode);
		if (typeof config.deferrable === "boolean") chunks.push(config.deferrable ? "deferrable" : "not deferrable");
		return __sql_sql_ts.sql.raw(chunks.join(" "));
	}
	setTransaction(config) {
		return this.session.execute(__sql_sql_ts.sql`set transaction ${this.getTransactionConfigSQL(config)}`);
	}
};
const migrate = effect_Effect.fn("migrate")(function* (migrations, session, config) {
	const migrationsTable = typeof config === "string" ? "__drizzle_migrations" : config.migrationsTable ?? "__drizzle_migrations";
	const migrationsSchema = typeof config === "string" ? "drizzle" : config.migrationsSchema ?? "drizzle";
	yield* session.execute(__sql_sql_ts.sql`CREATE SCHEMA IF NOT EXISTS ${__sql_sql_ts.sql.identifier(migrationsSchema)}`);
	const { newDb } = yield* (0, __up_migrations_effect_pg_ts.upgradeIfNeeded)(migrationsSchema, migrationsTable, session, migrations);
	if (newDb) {
		const migrationTableCreate = __sql_sql_ts.sql`
			CREATE TABLE IF NOT EXISTS ${__sql_sql_ts.sql.identifier(migrationsSchema)}.${__sql_sql_ts.sql.identifier(migrationsTable)} (
				id SERIAL PRIMARY KEY,
				hash text NOT NULL,
				created_at bigint,
				name text,
				applied_at timestamp with time zone DEFAULT now()
			)
		`;
		yield* session.execute(migrationTableCreate);
	}
	const dbMigrations = yield* session.objects(__sql_sql_ts.sql`select id, hash, created_at, name from ${__sql_sql_ts.sql.identifier(migrationsSchema)}.${__sql_sql_ts.sql.identifier(migrationsTable)}`);
	if (typeof config === "object" && config.init) {
		if (dbMigrations.length) return yield* new __effect_core_errors_ts.MigratorInitError({ exitCode: "databaseMigrations" });
		if (migrations.length > 1) return yield* new __effect_core_errors_ts.MigratorInitError({ exitCode: "localMigrations" });
		const [migration] = migrations;
		if (!migration) return;
		yield* session.execute(__sql_sql_ts.sql`insert into ${__sql_sql_ts.sql.identifier(migrationsSchema)}.${__sql_sql_ts.sql.identifier(migrationsTable)} ("hash", "created_at", "name") values(${migration.hash}, ${migration.folderMillis}, ${migration.name})`);
		return;
	}
	const migrationsToRun = (0, __migrator_utils_ts.getMigrationsToRun)({
		localMigrations: migrations,
		dbMigrations
	});
	yield* session.transaction((tx) => effect_Effect.gen(function* () {
		for (const migration of migrationsToRun) {
			for (const stmt of migration.sql) yield* tx.execute(__sql_sql_ts.sql.raw(stmt));
			yield* tx.execute(__sql_sql_ts.sql`insert into ${__sql_sql_ts.sql.identifier(migrationsSchema)}.${__sql_sql_ts.sql.identifier(migrationsTable)} ("hash", "created_at", "name") values(${migration.hash}, ${migration.folderMillis}, ${migration.name})`);
		}
	}));
});

//#endregion
exports.PgEffectPreparedQuery = PgEffectPreparedQuery;
exports.PgEffectSession = PgEffectSession;
exports.PgEffectTransaction = PgEffectTransaction;
exports.migrate = migrate;
