// drizzle-orm/node-sqlite/driver.js
import { NodeSQLiteSession } from "./session.js";
import { entityKind } from "../entity.js";
import { jitCompatCheck } from "../utils.js";
import { DefaultLogger } from "../logger.js";
import * as V1 from "../_relations.js";
import { BaseSQLiteDatabase } from "../sqlite-core/db.js";
import { SQLiteSyncDialect } from "../sqlite-core/dialect.js";
import { DatabaseSync } from "node:sqlite";

//#region src/node-sqlite/driver.ts
var NodeSQLiteDatabase = class extends BaseSQLiteDatabase {
	static [entityKind] = "NodeSQLiteDatabase";
};
function construct(client, config = {}) {
	const dialect = new SQLiteSyncDialect();
	let logger;
	if (config.logger === true) logger = new DefaultLogger();
	else if (config.logger !== false) logger = config.logger;
	let schema;
	if (config.schema) {
		const tablesConfig = V1.extractTablesRelationalConfig(config.schema, V1.createTableRelationsHelpers);
		schema = {
			fullSchema: config.schema,
			schema: tablesConfig.tables,
			tableNamesMap: tablesConfig.tableNamesMap
		};
	}
	const relations = config.relations ?? {};
	const db = new NodeSQLiteDatabase("sync", dialect, new NodeSQLiteSession(client, dialect, relations, schema, {
		logger,
		useJitMappers: jitCompatCheck(config.jit)
	}), relations, schema);
	db.$client = client;
	return db;
}
function drizzle(...params) {
	if (params[0] === void 0 || typeof params[0] === "string") return construct(params[0] === void 0 ? new DatabaseSync(":memory:") : new DatabaseSync(params[0]), params[1]);
	const { connection, client, ...drizzleConfig } = params[0];
	if (client) return construct(client, drizzleConfig);
	if (typeof connection === "object") {
		const { path, ...options } = connection;
		return construct(new DatabaseSync(path ?? ":memory:", options), drizzleConfig);
	}
	return construct(new DatabaseSync(connection ?? ":memory:"), drizzleConfig);
}
(function(_drizzle) {
	function mock(config) {
		return construct({}, config);
	}
	_drizzle.mock = mock;
})(drizzle || (drizzle = {}));

//#endregion
export { NodeSQLiteDatabase, drizzle };
