// drizzle-orm/node-sqlite/driver.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
const require_node_sqlite_session = require('./session.cjs');
let __entity_ts = require("../entity.cjs");
let __utils_ts = require("../utils.cjs");
let __logger_ts = require("../logger.cjs");
let ___relations_ts = require("../_relations.cjs");
___relations_ts = require_runtime.__toESM(___relations_ts);
let __sqlite_core_db_ts = require("../sqlite-core/db.cjs");
let __sqlite_core_dialect_ts = require("../sqlite-core/dialect.cjs");
let node_sqlite = require("node:sqlite");

//#region src/node-sqlite/driver.ts
var NodeSQLiteDatabase = class extends __sqlite_core_db_ts.BaseSQLiteDatabase {
	static [__entity_ts.entityKind] = "NodeSQLiteDatabase";
};
function construct(client, config = {}) {
	const dialect = new __sqlite_core_dialect_ts.SQLiteSyncDialect();
	let logger;
	if (config.logger === true) logger = new __logger_ts.DefaultLogger();
	else if (config.logger !== false) logger = config.logger;
	let schema;
	if (config.schema) {
		const tablesConfig = ___relations_ts.extractTablesRelationalConfig(config.schema, ___relations_ts.createTableRelationsHelpers);
		schema = {
			fullSchema: config.schema,
			schema: tablesConfig.tables,
			tableNamesMap: tablesConfig.tableNamesMap
		};
	}
	const relations = config.relations ?? {};
	const db = new NodeSQLiteDatabase("sync", dialect, new require_node_sqlite_session.NodeSQLiteSession(client, dialect, relations, schema, {
		logger,
		useJitMappers: (0, __utils_ts.jitCompatCheck)(config.jit)
	}), relations, schema);
	db.$client = client;
	return db;
}
function drizzle(...params) {
	if (params[0] === void 0 || typeof params[0] === "string") return construct(params[0] === void 0 ? new node_sqlite.DatabaseSync(":memory:") : new node_sqlite.DatabaseSync(params[0]), params[1]);
	const { connection, client, ...drizzleConfig } = params[0];
	if (client) return construct(client, drizzleConfig);
	if (typeof connection === "object") {
		const { path, ...options } = connection;
		return construct(new node_sqlite.DatabaseSync(path ?? ":memory:", options), drizzleConfig);
	}
	return construct(new node_sqlite.DatabaseSync(connection ?? ":memory:"), drizzleConfig);
}
(function(_drizzle) {
	function mock(config) {
		return construct({}, config);
	}
	_drizzle.mock = mock;
})(drizzle || (drizzle = {}));

//#endregion
exports.NodeSQLiteDatabase = NodeSQLiteDatabase;
Object.defineProperty(exports, 'drizzle', {
  enumerable: true,
  get: function () {
    return drizzle;
  }
});
