// waddler/duckdb/driver.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var driver_exports = {};
__export(driver_exports, {
  sql: () => sql,
  waddler: () => waddler
});
module.exports = __toCommonJS(driver_exports);
var import_duckdb = __toESM(require("duckdb"), 1);
var import_dialect = require("../duckdb-core/dialect.cjs");
var import_logger = require("../logger.cjs");
var import_recycling_pool = require("../recycling-pool.cjs");
var import_sql_template_params = require("../sql-template-params.cjs");
var import_sql = require("../sql.cjs");
var import_session = require("./session.cjs");
const sql = (strings, ...params) => {
  const sqlWrapper = new import_sql.SQLWrapper();
  sqlWrapper.with({ templateParams: { strings, params } });
  const dialect = new import_dialect.DuckdbDialect();
  return new import_sql_template_params.SQLQuery(sqlWrapper, dialect);
};
Object.assign(sql, import_dialect.SQLFunctions);
const createSqlTemplate = (pool, configOptions = {}) => {
  const dialect = new import_dialect.DuckdbDialect();
  let logger;
  if (configOptions.logger === true) {
    logger = new import_logger.DefaultLogger();
  } else if (configOptions.logger !== false) {
    logger = configOptions.logger;
  }
  const fn = (strings, ...params) => {
    const sql2 = new import_sql.SQLWrapper();
    sql2.with({ templateParams: { strings, params } }).prepareQuery(dialect);
    return new import_session.DuckdbSQLTemplate(sql2, pool, dialect, { logger });
  };
  Object.assign(fn, import_dialect.SQLFunctions);
  return fn;
};
function createDatabase(url, options) {
  return new Promise((resolve, reject) => {
    const db = new import_duckdb.default.Database(url, options, (err) => {
      if (err) {
        return reject(err);
      }
      resolve(db);
    });
  });
}
const createFactory = ({
  url,
  accessMode = "read_write",
  maxMemory = "512MB",
  threads = "4"
}) => {
  const factory = {
    create: async function() {
      try {
        const db = await createDatabase(url, {
          access_mode: accessMode,
          max_memory: maxMemory,
          threads
        });
        return db;
      } catch (error) {
        const newError = new Error(error.message);
        throw newError;
      }
    },
    destroy: async function(db) {
      try {
        return db.close();
      } catch (error) {
        const newError = new Error(error.message);
        throw newError;
      }
    }
  };
  return factory;
};
function waddler({
  dbUrl,
  url,
  min = 1,
  max = 1,
  accessMode = "read_write",
  maxMemory = "512MB",
  threads = "4",
  logger
}) {
  url = url === void 0 && dbUrl !== void 0 ? dbUrl : url;
  const factory = createFactory({
    url,
    accessMode,
    maxMemory,
    threads
  });
  const options = {
    max,
    // maximum size of the pool
    min
    // minimum size of the pool
  };
  const pool = new import_recycling_pool.RecyclingPool(factory, options);
  return createSqlTemplate(pool, { logger });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  sql,
  waddler
});
