// waddler/duckdb/driver.js
import duckdb from "duckdb";
import { DuckdbDialect, SQLFunctions } from "../duckdb-core/dialect.js";
import { DefaultLogger } from "../logger.js";
import { RecyclingPool } from "../recycling-pool.js";
import { SQLQuery } from "../sql-template-params.js";
import { SQLWrapper } from "../sql.js";
import { DuckdbSQLTemplate } from "./session.js";
const sql = (strings, ...params) => {
  const sqlWrapper = new SQLWrapper();
  sqlWrapper.with({ templateParams: { strings, params } });
  const dialect = new DuckdbDialect();
  return new SQLQuery(sqlWrapper, dialect);
};
Object.assign(sql, SQLFunctions);
const createSqlTemplate = (pool, configOptions = {}) => {
  const dialect = new DuckdbDialect();
  let logger;
  if (configOptions.logger === true) {
    logger = new DefaultLogger();
  } else if (configOptions.logger !== false) {
    logger = configOptions.logger;
  }
  const fn = (strings, ...params) => {
    const sql2 = new SQLWrapper();
    sql2.with({ templateParams: { strings, params } }).prepareQuery(dialect);
    return new DuckdbSQLTemplate(sql2, pool, dialect, { logger });
  };
  Object.assign(fn, SQLFunctions);
  return fn;
};
function createDatabase(url, options) {
  return new Promise((resolve, reject) => {
    const db = new duckdb.Database(url, options, (err) => {
      if (err) {
        return reject(err);
      }
      resolve(db);
    });
  });
}
const createFactory = ({
  url,
  accessMode = "read_write",
  maxMemory = "512MB",
  threads = "4"
}) => {
  const factory = {
    create: async function() {
      try {
        const db = await createDatabase(url, {
          access_mode: accessMode,
          max_memory: maxMemory,
          threads
        });
        return db;
      } catch (error) {
        const newError = new Error(error.message);
        throw newError;
      }
    },
    destroy: async function(db) {
      try {
        return db.close();
      } catch (error) {
        const newError = new Error(error.message);
        throw newError;
      }
    }
  };
  return factory;
};
function waddler({
  dbUrl,
  url,
  min = 1,
  max = 1,
  accessMode = "read_write",
  maxMemory = "512MB",
  threads = "4",
  logger
}) {
  url = url === void 0 && dbUrl !== void 0 ? dbUrl : url;
  const factory = createFactory({
    url,
    accessMode,
    maxMemory,
    threads
  });
  const options = {
    max,
    // maximum size of the pool
    min
    // minimum size of the pool
  };
  const pool = new RecyclingPool(factory, options);
  return createSqlTemplate(pool, { logger });
}
export {
  sql,
  waddler
};
