// waddler/pg/node-postgres/session.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var session_exports = {};
__export(session_exports, {
  NodePgSQLTemplate: () => NodePgSQLTemplate
});
module.exports = __toCommonJS(session_exports);
var import_pg = __toESM(require("pg"), 1);
var import_errors = require("../../errors/index.cjs");
var import_sql_template = require("../../sql-template.cjs");
const { Pool, types } = import_pg.default;
const pgTypeConfig = {
  // @ts-expect-error
  getTypeParser: (typeId, format) => {
    if (typeId === types.builtins.INTERVAL) return (val) => val;
    if (typeId === 1187) return (val) => val;
    return types.getTypeParser(typeId, format);
  }
};
class NodePgSQLTemplate extends import_sql_template.SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
    const query = this.sqlWrapper.getQuery(this.dialect).sql;
    this.queryConfig = {
      text: query,
      types: pgTypeConfig
    };
    this.rawQueryConfig = {
      rowMode: "array",
      text: query,
      types: pgTypeConfig
    };
  }
  queryConfig;
  rawQueryConfig;
  async execute() {
    const { params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      const queryResult = this.options.rowMode === "array" ? await this.client.query(this.rawQueryConfig, params) : await this.client.query(this.queryConfig, params);
      ({ rows: finalRes, ...finalMetadata } = queryResult);
    } catch (error) {
      throw new import_errors.WaddlerQueryError(this.queryConfig.text, params, error);
    }
    this.logger.logQuery(this.queryConfig.text, params, finalMetadata);
    return finalRes;
  }
  // TODO: revise: maybe I should override chunked method because we can use QueryStream with option 'batchSize' in QueryStreamConfig
  async *stream() {
    let conn;
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    try {
      conn = this.client instanceof Pool ? await this.client.connect() : this.client;
      const queryStreamObj = this.configOptions?.extensions?.find((it) => it.name === "WaddlerPgQueryStream");
      if (!queryStreamObj) {
        throw new Error(
          'To use stream feature, you would need to provide queryStream() function to waddler extensions, example: waddler("", { extensions: [queryStream()] })'
        );
      }
      const queryStream = new queryStreamObj.constructor(query, params, { types: this.queryConfig.types });
      const stream = conn.query(queryStream);
      for await (const row of stream) {
        yield row;
      }
      if (this.client instanceof Pool) {
        conn.release();
      }
    } catch (error) {
      if (this.client instanceof Pool) {
        conn?.release();
      }
      throw new import_errors.WaddlerQueryError(query, params, error);
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  NodePgSQLTemplate
});
