// @netlify/zip-it-and-ship-it/dist/manifest.js
import { promises as fs } from 'fs';
import { resolve } from 'path';
import { arch, platform } from 'process';
const MANIFEST_VERSION = 1;
export const createManifest = async ({ functions, path }) => {
    const formattedFunctions = functions.map((func) => formatFunctionForManifest(func));
    const payload = {
        functions: formattedFunctions,
        system: { arch, platform },
        timestamp: Date.now(),
        version: MANIFEST_VERSION,
    };
    await fs.writeFile(path, JSON.stringify(payload));
};
const formatFunctionForManifest = ({ bootstrapVersion, bundler, displayName, eventSubscriptions, excludedRoutes, generator, invocationMode, mainFile, memory, name, path, priority, region, trafficRules, routes, runtime, runtimeVersion, runtimeAPIVersion, schedule, timeout, vcpu, }) => {
    const manifestFunction = {
        bundler,
        displayName,
        generator,
        timeout,
        invocationMode,
        buildData: { bootstrapVersion, runtimeAPIVersion },
        mainFile,
        memory,
        name,
        priority,
        region,
        trafficRules,
        runtimeVersion,
        path: resolve(path),
        runtime,
        schedule,
        vcpu,
    };
    if (eventSubscriptions?.length) {
        manifestFunction.eventSubscriptions = eventSubscriptions;
    }
    if (routes?.length !== 0) {
        manifestFunction.routes = routes;
    }
    if (excludedRoutes && excludedRoutes.length !== 0) {
        manifestFunction.excludedRoutes = excludedRoutes;
    }
    return manifestFunction;
};
