// drizzle-orm/tidb-serverless/driver.js
import { TiDBServerlessSession } from "./session.js";
import { entityKind } from "../entity.js";
import { jitCompatCheck } from "../utils.js";
import { connect } from "@tidbcloud/serverless";
import { DefaultLogger } from "../logger.js";
import { MySqlDatabase } from "../mysql-core/db.js";
import { MySqlDialect } from "../mysql-core/dialect.js";

//#region src/tidb-serverless/driver.ts
var TiDBServerlessDatabase = class extends MySqlDatabase {
	static [entityKind] = "TiDBServerlessDatabase";
};
function construct(client, config = {}) {
	const dialect = new MySqlDialect({ useJitMappers: jitCompatCheck(config.jit) });
	let logger;
	if (config.logger === true) logger = new DefaultLogger();
	else if (config.logger !== false) logger = config.logger;
	const relations = config.relations ?? {};
	const db = new TiDBServerlessDatabase(dialect, new TiDBServerlessSession(client, dialect, void 0, relations, {
		logger,
		cache: config.cache
	}), relations);
	db.$client = client;
	db.$cache = config.cache;
	if (db.$cache) db.$cache["invalidate"] = config.cache?.onMutate;
	return db;
}
function drizzle(...params) {
	if (typeof params[0] === "string") return construct(connect({ url: params[0] }), params[1]);
	const { connection, client, ...DrizzleMySqlConfig } = params[0];
	if (client) return construct(client, DrizzleMySqlConfig);
	return construct(typeof connection === "string" ? connect({ url: connection }) : connect(connection), DrizzleMySqlConfig);
}
(function(_drizzle) {
	function mock(config) {
		return construct({}, config);
	}
	_drizzle.mock = mock;
})(drizzle || (drizzle = {}));

//#endregion
export { TiDBServerlessDatabase, drizzle };
