// drizzle-orm/tidb-serverless/session.js
import { entityKind } from "../entity.js";
import { sql } from "../sql/sql.js";
import { NoopLogger } from "../logger.js";
import { NoopCache } from "../cache/core/index.js";
import { MySqlPreparedQuery, MySqlSession, MySqlTransaction } from "../mysql-core/session.js";

//#region src/tidb-serverless/session.ts
var TiDBServerlessSession = class TiDBServerlessSession extends MySqlSession {
	static [entityKind] = "TiDBServerlessSession";
	logger;
	client;
	cache;
	constructor(baseClient, dialect, tx, relations, options = {}) {
		super(dialect);
		this.baseClient = baseClient;
		this.relations = relations;
		this.options = options;
		this.client = tx ?? baseClient;
		this.logger = options.logger ?? new NoopLogger();
		this.cache = options.cache ?? new NoopCache();
	}
	prepareQuery(query, mode, mapper, queryMetadata, cacheConfig) {
		const { client } = this;
		const queryConfig = mode === "arrays" ? { arrayMode: true } : { fullResult: true };
		const executor = async (params = []) => {
			const raw = client.execute(query.sql, params, queryConfig);
			if (mode === "arrays") return raw;
			if (mode === "objects") return raw.then((res) => res.rows);
			if (!mapper) return raw;
			return raw.then((res) => ({
				insertId: res.lastInsertId ?? 0,
				affectedRows: res.rowsAffected ?? 0
			}));
		};
		return new MySqlPreparedQuery(executor, void 0, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	async transaction(transaction) {
		const nativeTx = await this.baseClient.begin();
		try {
			const session = new TiDBServerlessSession(this.baseClient, this.dialect, nativeTx, this.relations, this.options);
			const result = await transaction(new TiDBServerlessTransaction(this.dialect, session, this.relations));
			await nativeTx.commit();
			return result;
		} catch (err) {
			await nativeTx.rollback();
			throw err;
		}
	}
};
var TiDBServerlessTransaction = class TiDBServerlessTransaction extends MySqlTransaction {
	static [entityKind] = "TiDBServerlessTransaction";
	constructor(dialect, session, relations, nestedIndex = 0) {
		super(dialect, session, relations, nestedIndex);
	}
	async transaction(transaction) {
		const savepointName = `sp${this.nestedIndex + 1}`;
		const tx = new TiDBServerlessTransaction(this.dialect, this.session, this.relations, this.nestedIndex + 1);
		await tx.execute(sql.raw(`savepoint ${savepointName}`));
		try {
			const result = await transaction(tx);
			await tx.execute(sql.raw(`release savepoint ${savepointName}`));
			return result;
		} catch (err) {
			await tx.execute(sql.raw(`rollback to savepoint ${savepointName}`));
			throw err;
		}
	}
};

//#endregion
export { TiDBServerlessSession, TiDBServerlessTransaction };
