// drizzle-orm/tidb-serverless/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");
let __logger_ts = require("../logger.cjs");
let __cache_core_index_ts = require("../cache/core/index.cjs");
let __mysql_core_session_ts = require("../mysql-core/session.cjs");

//#region src/tidb-serverless/session.ts
var TiDBServerlessSession = class TiDBServerlessSession extends __mysql_core_session_ts.MySqlSession {
	static [__entity_ts.entityKind] = "TiDBServerlessSession";
	logger;
	client;
	cache;
	constructor(baseClient, dialect, tx, relations, options = {}) {
		super(dialect);
		this.baseClient = baseClient;
		this.relations = relations;
		this.options = options;
		this.client = tx ?? baseClient;
		this.logger = options.logger ?? new __logger_ts.NoopLogger();
		this.cache = options.cache ?? new __cache_core_index_ts.NoopCache();
	}
	prepareQuery(query, mode, mapper, queryMetadata, cacheConfig) {
		const { client } = this;
		const queryConfig = mode === "arrays" ? { arrayMode: true } : { fullResult: true };
		const executor = async (params = []) => {
			const raw = client.execute(query.sql, params, queryConfig);
			if (mode === "arrays") return raw;
			if (mode === "objects") return raw.then((res) => res.rows);
			if (!mapper) return raw;
			return raw.then((res) => ({
				insertId: res.lastInsertId ?? 0,
				affectedRows: res.rowsAffected ?? 0
			}));
		};
		return new __mysql_core_session_ts.MySqlPreparedQuery(executor, void 0, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	async transaction(transaction) {
		const nativeTx = await this.baseClient.begin();
		try {
			const session = new TiDBServerlessSession(this.baseClient, this.dialect, nativeTx, this.relations, this.options);
			const result = await transaction(new TiDBServerlessTransaction(this.dialect, session, this.relations));
			await nativeTx.commit();
			return result;
		} catch (err) {
			await nativeTx.rollback();
			throw err;
		}
	}
};
var TiDBServerlessTransaction = class TiDBServerlessTransaction extends __mysql_core_session_ts.MySqlTransaction {
	static [__entity_ts.entityKind] = "TiDBServerlessTransaction";
	constructor(dialect, session, relations, nestedIndex = 0) {
		super(dialect, session, relations, nestedIndex);
	}
	async transaction(transaction) {
		const savepointName = `sp${this.nestedIndex + 1}`;
		const tx = new TiDBServerlessTransaction(this.dialect, this.session, this.relations, this.nestedIndex + 1);
		await tx.execute(__sql_sql_ts.sql.raw(`savepoint ${savepointName}`));
		try {
			const result = await transaction(tx);
			await tx.execute(__sql_sql_ts.sql.raw(`release savepoint ${savepointName}`));
			return result;
		} catch (err) {
			await tx.execute(__sql_sql_ts.sql.raw(`rollback to savepoint ${savepointName}`));
			throw err;
		}
	}
};

//#endregion
exports.TiDBServerlessSession = TiDBServerlessSession;
exports.TiDBServerlessTransaction = TiDBServerlessTransaction;
