// drizzle-orm/bun-sql/mysql/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../../entity.cjs");
let __logger_ts = require("../../logger.cjs");
let __cache_core_index_ts = require("../../cache/core/index.cjs");
let __mysql_core_session_ts = require("../../mysql-core/session.cjs");

//#region src/bun-sql/mysql/session.ts
var BunMySqlSession = class BunMySqlSession extends __mysql_core_session_ts.MySqlSession {
	static [__entity_ts.entityKind] = "BunMySqlSession";
	logger;
	cache;
	constructor(client, dialect, relations, options) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
		this.logger = options.logger ?? new __logger_ts.NoopLogger();
		this.cache = options.cache ?? new __cache_core_index_ts.NoopCache();
	}
	prepareQuery(query, mode, mapper, queryMetadata, cacheConfig) {
		const { client } = this;
		const executor = async (params = []) => {
			const raw = client.unsafe(query.sql, params);
			if (mode === "arrays") return raw.values();
			if (mode === "objects") return raw;
			if (!mapper) return raw;
			return raw.then(({ lastInsertRowid, affectedRows }) => ({
				insertId: lastInsertRowid,
				affectedRows
			}));
		};
		return new __mysql_core_session_ts.MySqlPreparedQuery(executor, void 0, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	async transaction(transaction, config) {
		const startTransactionSql = config ? this.getStartTransactionSQL(config)?.inlineParams().toQuery(this.dialect).sql.slice(18) ?? "" : "";
		if (config?.isolationLevel) throw new Error("Driver doesn't support setting isolation level on transaction");
		return this.client.begin(startTransactionSql, async (client) => {
			const session = new BunMySqlSession(client, this.dialect, this.relations, this.options);
			return transaction(new BunMySqlTransaction(this.dialect, session, this.relations, 0));
		});
	}
};
var BunMySqlTransaction = class BunMySqlTransaction extends __mysql_core_session_ts.MySqlTransaction {
	static [__entity_ts.entityKind] = "BunMySqlTransaction";
	async transaction(transaction) {
		return this.session.client.savepoint((client) => {
			const session = new BunMySqlSession(client, this.dialect, this.relations, this.session.options);
			return transaction(new BunMySqlTransaction(this.dialect, session, this.relations, this.nestedIndex + 1));
		});
	}
};

//#endregion
exports.BunMySqlSession = BunMySqlSession;
exports.BunMySqlTransaction = BunMySqlTransaction;
