// drizzle-orm/bun-sql/mysql/session.js
import { entityKind } from "../../entity.js";
import { NoopLogger } from "../../logger.js";
import { NoopCache } from "../../cache/core/index.js";
import { MySqlPreparedQuery, MySqlSession, MySqlTransaction } from "../../mysql-core/session.js";

//#region src/bun-sql/mysql/session.ts
var BunMySqlSession = class BunMySqlSession extends MySqlSession {
	static [entityKind] = "BunMySqlSession";
	logger;
	cache;
	constructor(client, dialect, relations, options) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
		this.logger = options.logger ?? new NoopLogger();
		this.cache = options.cache ?? new NoopCache();
	}
	prepareQuery(query, mode, mapper, queryMetadata, cacheConfig) {
		const { client } = this;
		const executor = async (params = []) => {
			const raw = client.unsafe(query.sql, params);
			if (mode === "arrays") return raw.values();
			if (mode === "objects") return raw;
			if (!mapper) return raw;
			return raw.then(({ lastInsertRowid, affectedRows }) => ({
				insertId: lastInsertRowid,
				affectedRows
			}));
		};
		return new MySqlPreparedQuery(executor, void 0, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	async transaction(transaction, config) {
		const startTransactionSql = config ? this.getStartTransactionSQL(config)?.inlineParams().toQuery(this.dialect).sql.slice(18) ?? "" : "";
		if (config?.isolationLevel) throw new Error("Driver doesn't support setting isolation level on transaction");
		return this.client.begin(startTransactionSql, async (client) => {
			const session = new BunMySqlSession(client, this.dialect, this.relations, this.options);
			return transaction(new BunMySqlTransaction(this.dialect, session, this.relations, 0));
		});
	}
};
var BunMySqlTransaction = class BunMySqlTransaction extends MySqlTransaction {
	static [entityKind] = "BunMySqlTransaction";
	async transaction(transaction) {
		return this.session.client.savepoint((client) => {
			const session = new BunMySqlSession(client, this.dialect, this.relations, this.session.options);
			return transaction(new BunMySqlTransaction(this.dialect, session, this.relations, this.nestedIndex + 1));
		});
	}
};

//#endregion
export { BunMySqlSession, BunMySqlTransaction };
