// waddler/mysql/mysql-core/dialect.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dialect_exports = {};
__export(dialect_exports, {
  MySQLDialect: () => MySQLDialect,
  SQLFunctions: () => SQLFunctions
});
module.exports = __toCommonJS(dialect_exports);
var import_sql_template_params = require("../../sql-template-params.cjs");
class MySQLDialect extends import_sql_template_params.Dialect {
  escapeParam() {
    return `?`;
  }
  escapeIdentifier(identifier) {
    return `\`${identifier}\``;
  }
  checkIdentifierObject(object) {
    if (Object.values(object).includes(void 0)) {
      throw new Error(
        `you can't specify undefined parameters. maybe you want to omit it?`
      );
    }
    if (Object.keys(object).length === 0) {
      throw new Error(`you need to specify at least one parameter.`);
    }
    if (Object.keys(object).length === 1 && object.as !== void 0) {
      throw new Error(
        `you can't specify only "as" property. you have to specify "column" or "table" property along with "as".`
      );
    }
    if (!["string", "undefined"].includes(typeof object.table) || !["string", "undefined"].includes(typeof object.column) || !["string", "undefined"].includes(typeof object.as)) {
      throw new Error(
        "object properties 'table', 'column', 'as' should be of string type or omitted."
      );
    }
  }
  // SQLValues
  valueToSQL({ value, params }) {
    if (value instanceof import_sql_template_params.SQLDefault) {
      return value.generateSQL().sql;
    }
    if (typeof value === "number" || typeof value === "bigint" || typeof value === "boolean" || typeof value === "string" || value === null || value instanceof Date || Buffer.isBuffer(value)) {
      params.push(value);
      return this.escapeParam();
    }
    if (typeof value === "object") {
      params.push(JSON.stringify(value));
      return this.escapeParam();
    }
    if (value === void 0) {
      throw new Error("value can't be undefined, maybe you mean sql.default?");
    }
    throw new Error(`you can't specify ${typeof value} as value.`);
  }
}
const SQLFunctions = {
  identifier: (value) => {
    return new import_sql_template_params.SQLIdentifier(value);
  },
  values: (value) => {
    return new import_sql_template_params.SQLValues(value);
  },
  raw: (value) => {
    return new import_sql_template_params.SQLRaw(value);
  },
  default: new import_sql_template_params.SQLDefault()
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MySQLDialect,
  SQLFunctions
});
