// drizzle-orm/postgres-js/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let __logger_ts = require("../logger.cjs");
let __cache_core_index_ts = require("../cache/core/index.cjs");
let __pg_core_async_session_ts = require("../pg-core/async/session.cjs");

//#region src/postgres-js/session.ts
var PostgresJsSession = class PostgresJsSession extends __pg_core_async_session_ts.PgAsyncSession {
	static [__entity_ts.entityKind] = "PostgresJsSession";
	logger;
	cache;
	constructor(client, dialect, relations, options = {}) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
		this.logger = options.logger ?? new __logger_ts.NoopLogger();
		this.cache = options.cache ?? new __cache_core_index_ts.NoopCache();
	}
	prepareQuery(query, mode, name, mapper, queryMetadata, cacheConfig) {
		const executor = async (params) => {
			if (mode === "objects") return this.client.unsafe(query.sql, params ?? [], { prepare: name !== false }).then((rows) => Object.values(rows));
			if (mode === "raw") return this.client.unsafe(query.sql, params ?? [], { prepare: name !== false });
			return this.client.unsafe(query.sql, params ?? [], { prepare: name !== false }).values();
		};
		return new __pg_core_async_session_ts.PgAsyncPreparedQuery(executor, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	transaction(transaction, config) {
		return this.client.begin(async (client) => {
			const session = new PostgresJsSession(client, this.dialect, this.relations, this.options);
			const tx = new PostgresJsTransaction(this.dialect, session, this.relations);
			if (config) await tx.setTransaction(config);
			return transaction(tx);
		});
	}
};
var PostgresJsTransaction = class PostgresJsTransaction extends __pg_core_async_session_ts.PgAsyncTransaction {
	static [__entity_ts.entityKind] = "PostgresJsTransaction";
	constructor(dialect, session, relations, nestedIndex = 0) {
		super(dialect, session, relations, nestedIndex, false);
		this.session = session;
	}
	transaction(transaction) {
		return this.session.client.savepoint((client) => {
			const session = new PostgresJsSession(client, this.dialect, this._.relations, this.session.options);
			return transaction(new PostgresJsTransaction(this.dialect, session, this._.relations));
		});
	}
};

//#endregion
exports.PostgresJsSession = PostgresJsSession;
exports.PostgresJsTransaction = PostgresJsTransaction;
