// drizzle-orm/postgres-js/session.js
import { entityKind } from "../entity.js";
import { NoopLogger } from "../logger.js";
import { NoopCache } from "../cache/core/index.js";
import { PgAsyncPreparedQuery, PgAsyncSession, PgAsyncTransaction } from "../pg-core/async/session.js";

//#region src/postgres-js/session.ts
var PostgresJsSession = class PostgresJsSession extends PgAsyncSession {
	static [entityKind] = "PostgresJsSession";
	logger;
	cache;
	constructor(client, dialect, relations, options = {}) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
		this.logger = options.logger ?? new NoopLogger();
		this.cache = options.cache ?? new NoopCache();
	}
	prepareQuery(query, mode, name, mapper, queryMetadata, cacheConfig) {
		const executor = async (params) => {
			if (mode === "objects") return this.client.unsafe(query.sql, params ?? [], { prepare: name !== false }).then((rows) => Object.values(rows));
			if (mode === "raw") return this.client.unsafe(query.sql, params ?? [], { prepare: name !== false });
			return this.client.unsafe(query.sql, params ?? [], { prepare: name !== false }).values();
		};
		return new PgAsyncPreparedQuery(executor, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	transaction(transaction, config) {
		return this.client.begin(async (client) => {
			const session = new PostgresJsSession(client, this.dialect, this.relations, this.options);
			const tx = new PostgresJsTransaction(this.dialect, session, this.relations);
			if (config) await tx.setTransaction(config);
			return transaction(tx);
		});
	}
};
var PostgresJsTransaction = class PostgresJsTransaction extends PgAsyncTransaction {
	static [entityKind] = "PostgresJsTransaction";
	constructor(dialect, session, relations, nestedIndex = 0) {
		super(dialect, session, relations, nestedIndex, false);
		this.session = session;
	}
	transaction(transaction) {
		return this.session.client.savepoint((client) => {
			const session = new PostgresJsSession(client, this.dialect, this._.relations, this.session.options);
			return transaction(new PostgresJsTransaction(this.dialect, session, this._.relations));
		});
	}
};

//#endregion
export { PostgresJsSession, PostgresJsTransaction };
