// drizzle-orm/expo-sqlite/migrator.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __migrator_utils_ts = require("../migrator.utils.cjs");
let react = require("react");

//#region src/expo-sqlite/migrator.ts
async function readMigrationFiles({ migrations }) {
	const migrationQueries = [];
	const sortedMigrations = Object.keys(migrations).sort();
	for (const key of sortedMigrations) {
		const query = migrations[key];
		if (!query) throw new Error(`Missing migration: ${key}`);
		try {
			const result = query.split("--> statement-breakpoint").map((it) => {
				return it;
			});
			const migrationDate = (0, __migrator_utils_ts.formatToMillis)(key.slice(0, 14));
			migrationQueries.push({
				sql: result,
				bps: true,
				folderMillis: migrationDate,
				hash: "",
				name: key
			});
		} catch {
			throw new Error(`Failed to parse migration: ${key}`);
		}
	}
	return migrationQueries;
}
async function migrate(db, config) {
	const migrations = await readMigrationFiles(config);
	return db.dialect.migrate(migrations, db.session);
}
const useMigrations = (db, migrations) => {
	const initialState = {
		success: false,
		error: void 0
	};
	const fetchReducer = (state, action) => {
		switch (action.type) {
			case "migrating": return { ...initialState };
			case "migrated": return {
				...initialState,
				success: action.payload
			};
			case "error": return {
				...initialState,
				error: action.payload
			};
			default: return state;
		}
	};
	const [state, dispatch] = (0, react.useReducer)(fetchReducer, initialState);
	(0, react.useEffect)(() => {
		dispatch({ type: "migrating" });
		migrate(db, migrations).then(() => {
			dispatch({
				type: "migrated",
				payload: true
			});
		}).catch((error) => {
			dispatch({
				type: "error",
				payload: error
			});
		});
	}, []);
	return state;
};

//#endregion
exports.migrate = migrate;
exports.useMigrations = useMigrations;
