// drizzle-orm/pg-core/async/select.js
import { PgSelectBase } from "../query-builders/select.js";
import { entityKind } from "../../entity.js";
import { tracer } from "../../tracing.js";
import { applyMixins, orderSelectedFields } from "../../utils.js";
import { QueryPromise } from "../../query-promise.js";

//#region src/pg-core/async/select.ts
var PgAsyncSelectBase = class extends PgSelectBase {
	static [entityKind] = "PgAsyncSelectQueryBuilder";
	/** @internal */
	_prepare(name, generateName = false) {
		const { session, config, dialect, joinsNotNullableMap, cacheConfig, usedTables } = this;
		const { fields } = config;
		return tracer.startActiveSpan("drizzle.prepareQuery", () => {
			const query = this.config.tagged ? dialect._sqlToQuery(this.getSQL()) : dialect.sqlToQuery(this.getSQL());
			const fieldsList = orderSelectedFields(fields, void 0, this.dialect.codecs);
			const mapper = this.dialect.mapperGenerators.rows(fieldsList, joinsNotNullableMap);
			return session.prepareQuery(query, "arrays", name ?? generateName, mapper, {
				type: "select",
				tables: [...usedTables]
			}, cacheConfig);
		});
	}
	/**
	* Create a prepared statement for this query. This allows
	* the database to remember this query for the given session
	* and call it by name, rather than specifying the full query.
	*
	* {@link https://www.postgresql.org/docs/current/sql-prepare.html | Postgres prepare documentation}
	*/
	prepare(name) {
		return this._prepare(name, true);
	}
	execute(placeholderValues) {
		return tracer.startActiveSpan("drizzle.operation", () => {
			return this._prepare().execute(placeholderValues);
		});
	}
};
applyMixins(PgAsyncSelectBase, [QueryPromise]);

//#endregion
export { PgAsyncSelectBase };
