// drizzle-orm/pg-core/async/session.js
import { PgAsyncDatabase } from "./db.js";
import { PgBasePreparedQuery, PgSession } from "../session.js";
import { entityKind, is } from "../../entity.js";
import { hasTelemetry, tracer } from "../../tracing.js";
import { assertUnreachable } from "../../utils.js";
import { fillPlaceholders, sql } from "../../sql/sql.js";
import { getMigrationsToRun } from "../../migrator.utils.js";
import { upgradeIfNeeded } from "../../up-migrations/pg.js";
import { DrizzleQueryError, TransactionRollbackError } from "../../errors.js";
import { NoopCache, strategyFor } from "../../cache/core/cache.js";

//#region src/pg-core/async/session.ts
var PgAsyncPreparedQuery = class extends PgBasePreparedQuery {
	static [entityKind] = "PgAsyncPreparedQuery";
	/** @internal */
	mapper;
	fastPath;
	constructor(executor, query, mapper, mode, logger, cache, queryMetadata, cacheConfig) {
		super(query);
		this.executor = executor;
		this.mode = mode;
		this.logger = logger;
		this.cache = cache;
		this.queryMetadata = queryMetadata;
		this.cacheConfig = cacheConfig;
		this.mapper = mapper;
		if (cache && cache.strategy() === "all" && cacheConfig === void 0) this.cacheConfig = {
			enabled: true,
			autoInvalidate: true
		};
		if (!this.cacheConfig?.enabled) this.cacheConfig = void 0;
		this.fastPath = cacheConfig === void 0 && (cache === void 0 || is(cache, NoopCache)) && !hasTelemetry;
	}
	async execute(placeholderValues = {}) {
		const { query, logger, executor, mapper, fastPath } = this;
		if (fastPath) {
			const sql = query._sql ? query._sql.join(" ") : query.sql;
			const params = query.params.length === 0 ? query.params : fillPlaceholders(query.params, placeholderValues);
			logger.logQuery(sql, params);
			const res = executor(params).catch((e) => {
				throw new DrizzleQueryError(sql, params, e);
			});
			if (!mapper) return res;
			return res.then((rows) => mapper(rows));
		}
		return tracer.startActiveSpan("drizzle.execute", async (span) => {
			const params = fillPlaceholders(this.query.params, placeholderValues);
			const sql = this.query._sql ? this.query._sql.join(" ") : this.query.sql;
			const { mapper } = this;
			span?.setAttributes({
				"drizzle.query.text": sql,
				"drizzle.query.params": JSON.stringify(params)
			});
			this.logger.logQuery(sql, params);
			const query = tracer.startActiveSpan("drizzle.driver.execute", async (span) => {
				span?.setAttributes({
					"drizzle.query.text": sql,
					"drizzle.query.params": JSON.stringify(params)
				});
				return await this.queryWithCache(sql, params, () => this.executor(params));
			});
			if (!mapper) return query;
			return query.then((rows) => tracer.startActiveSpan("drizzle.mapResponse", () => mapper(rows)));
		});
	}
	/** @internal */
	async queryWithCache(queryString, params, query) {
		const cacheStrat = this.cache !== void 0 && !is(this.cache, NoopCache) ? await strategyFor(queryString, params, this.queryMetadata, this.cacheConfig) : { type: "skip" };
		if (cacheStrat.type === "skip") return query().catch((e) => {
			throw new DrizzleQueryError(queryString, params, e);
		});
		const cache = this.cache;
		if (cacheStrat.type === "invalidate") return Promise.all([query(), cache.onMutate({ tables: cacheStrat.tables })]).then((res) => res[0]).catch((e) => {
			throw new DrizzleQueryError(queryString, params, e);
		});
		if (cacheStrat.type === "try") {
			const { tables, key, isTag, autoInvalidate, config } = cacheStrat;
			const fromCache = await cache.get(key, tables, isTag, autoInvalidate);
			if (fromCache === void 0) {
				const result = await query().catch((e) => {
					throw new DrizzleQueryError(queryString, params, e);
				});
				await cache.put(key, result, autoInvalidate ? tables : [], isTag, config);
				return result;
			}
			return fromCache;
		}
		assertUnreachable(cacheStrat);
	}
};
var PgAsyncSession = class extends PgSession {
	static [entityKind] = "PgAsyncSession";
	execute(query) {
		return tracer.startActiveSpan("drizzle.operation", () => {
			return tracer.startActiveSpan("drizzle.prepareQuery", () => {
				return this.prepareQuery(this.dialect.sqlToQuery(query), "raw", false);
			}).execute();
		});
	}
	arrays(query) {
		return tracer.startActiveSpan("drizzle.operation", () => {
			return tracer.startActiveSpan("drizzle.prepareQuery", () => {
				return this.prepareQuery(this.dialect.sqlToQuery(query), "arrays", false);
			}).execute();
		});
	}
	objects(query) {
		return tracer.startActiveSpan("drizzle.operation", () => {
			return tracer.startActiveSpan("drizzle.prepareQuery", () => {
				return this.prepareQuery(this.dialect.sqlToQuery(query), "objects", false);
			}).execute();
		});
	}
};
var PgAsyncTransaction = class extends PgAsyncDatabase {
	static [entityKind] = "PgAsyncTransaction";
	constructor(dialect, session, relations, nestedIndex = 0, parseRqbJson) {
		super(dialect, session, relations, parseRqbJson);
		this.nestedIndex = nestedIndex;
	}
	rollback() {
		throw new TransactionRollbackError();
	}
	/** @internal */
	getTransactionConfigSQL(config) {
		const chunks = [];
		if (config.isolationLevel) chunks.push(`isolation level ${config.isolationLevel}`);
		if (config.accessMode) chunks.push(config.accessMode);
		if (typeof config.deferrable === "boolean") chunks.push(config.deferrable ? "deferrable" : "not deferrable");
		return sql.raw(chunks.join(" "));
	}
	setTransaction(config) {
		return this.session.execute(sql`set transaction ${this.getTransactionConfigSQL(config)}`);
	}
};
async function migrate(migrations, db, config) {
	const migrationsTable = typeof config === "string" ? "__drizzle_migrations" : config.migrationsTable ?? "__drizzle_migrations";
	const migrationsSchema = typeof config === "string" ? "drizzle" : config.migrationsSchema ?? "drizzle";
	await db.execute(sql`CREATE SCHEMA IF NOT EXISTS ${sql.identifier(migrationsSchema)}`);
	const { newDb } = await upgradeIfNeeded(migrationsSchema, migrationsTable, db, migrations);
	if (newDb) {
		const migrationTableCreate = sql`
			CREATE TABLE IF NOT EXISTS ${sql.identifier(migrationsSchema)}.${sql.identifier(migrationsTable)} (
				id SERIAL PRIMARY KEY,
				hash text NOT NULL,
				created_at bigint,
				name text,
				applied_at timestamp with time zone DEFAULT now()
			)
		`;
		await db.execute(migrationTableCreate);
	}
	const dbMigrations = await db.session.objects(sql`select id, hash, created_at, name from ${sql.identifier(migrationsSchema)}.${sql.identifier(migrationsTable)}`);
	if (typeof config === "object" && config.init) {
		if (dbMigrations.length) return { exitCode: "databaseMigrations" };
		if (migrations.length > 1) return { exitCode: "localMigrations" };
		const [migration] = migrations;
		if (!migration) return;
		await db.execute(sql`insert into ${sql.identifier(migrationsSchema)}.${sql.identifier(migrationsTable)} ("hash", "created_at", "name") values(${migration.hash}, ${migration.folderMillis}, ${migration.name ?? null})`);
		return;
	}
	const migrationsToRun = getMigrationsToRun({
		localMigrations: migrations,
		dbMigrations
	});
	await db.transaction(async (tx) => {
		for (const migration of migrationsToRun) {
			for (const stmt of migration.sql) await tx.execute(sql.raw(stmt));
			await tx.execute(sql`insert into ${sql.identifier(migrationsSchema)}.${sql.identifier(migrationsTable)} ("hash", "created_at", "name") values(${migration.hash}, ${migration.folderMillis}, ${migration.name ?? null})`);
		}
	});
}

//#endregion
export { PgAsyncPreparedQuery, PgAsyncSession, PgAsyncTransaction, migrate };
