// drizzle-orm/pg-core/async/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_pg_core_async_db = require('./db.cjs');
const require_pg_core_session = require('../session.cjs');
let __entity_ts = require("../../entity.cjs");
let __tracing_ts = require("../../tracing.cjs");
let __utils_ts = require("../../utils.cjs");
let __sql_sql_ts = require("../../sql/sql.cjs");
let __migrator_utils_ts = require("../../migrator.utils.cjs");
let __up_migrations_pg_ts = require("../../up-migrations/pg.cjs");
let __errors_ts = require("../../errors.cjs");
let __cache_core_cache_ts = require("../../cache/core/cache.cjs");

//#region src/pg-core/async/session.ts
var PgAsyncPreparedQuery = class extends require_pg_core_session.PgBasePreparedQuery {
	static [__entity_ts.entityKind] = "PgAsyncPreparedQuery";
	/** @internal */
	mapper;
	fastPath;
	constructor(executor, query, mapper, mode, logger, cache, queryMetadata, cacheConfig) {
		super(query);
		this.executor = executor;
		this.mode = mode;
		this.logger = logger;
		this.cache = cache;
		this.queryMetadata = queryMetadata;
		this.cacheConfig = cacheConfig;
		this.mapper = mapper;
		if (cache && cache.strategy() === "all" && cacheConfig === void 0) this.cacheConfig = {
			enabled: true,
			autoInvalidate: true
		};
		if (!this.cacheConfig?.enabled) this.cacheConfig = void 0;
		this.fastPath = cacheConfig === void 0 && (cache === void 0 || (0, __entity_ts.is)(cache, __cache_core_cache_ts.NoopCache)) && !__tracing_ts.hasTelemetry;
	}
	async execute(placeholderValues = {}) {
		const { query, logger, executor, mapper, fastPath } = this;
		if (fastPath) {
			const sql = query._sql ? query._sql.join(" ") : query.sql;
			const params = query.params.length === 0 ? query.params : (0, __sql_sql_ts.fillPlaceholders)(query.params, placeholderValues);
			logger.logQuery(sql, params);
			const res = executor(params).catch((e) => {
				throw new __errors_ts.DrizzleQueryError(sql, params, e);
			});
			if (!mapper) return res;
			return res.then((rows) => mapper(rows));
		}
		return __tracing_ts.tracer.startActiveSpan("drizzle.execute", async (span) => {
			const params = (0, __sql_sql_ts.fillPlaceholders)(this.query.params, placeholderValues);
			const sql = this.query._sql ? this.query._sql.join(" ") : this.query.sql;
			const { mapper } = this;
			span?.setAttributes({
				"drizzle.query.text": sql,
				"drizzle.query.params": JSON.stringify(params)
			});
			this.logger.logQuery(sql, params);
			const query = __tracing_ts.tracer.startActiveSpan("drizzle.driver.execute", async (span) => {
				span?.setAttributes({
					"drizzle.query.text": sql,
					"drizzle.query.params": JSON.stringify(params)
				});
				return await this.queryWithCache(sql, params, () => this.executor(params));
			});
			if (!mapper) return query;
			return query.then((rows) => __tracing_ts.tracer.startActiveSpan("drizzle.mapResponse", () => mapper(rows)));
		});
	}
	/** @internal */
	async queryWithCache(queryString, params, query) {
		const cacheStrat = this.cache !== void 0 && !(0, __entity_ts.is)(this.cache, __cache_core_cache_ts.NoopCache) ? await (0, __cache_core_cache_ts.strategyFor)(queryString, params, this.queryMetadata, this.cacheConfig) : { type: "skip" };
		if (cacheStrat.type === "skip") return query().catch((e) => {
			throw new __errors_ts.DrizzleQueryError(queryString, params, e);
		});
		const cache = this.cache;
		if (cacheStrat.type === "invalidate") return Promise.all([query(), cache.onMutate({ tables: cacheStrat.tables })]).then((res) => res[0]).catch((e) => {
			throw new __errors_ts.DrizzleQueryError(queryString, params, e);
		});
		if (cacheStrat.type === "try") {
			const { tables, key, isTag, autoInvalidate, config } = cacheStrat;
			const fromCache = await cache.get(key, tables, isTag, autoInvalidate);
			if (fromCache === void 0) {
				const result = await query().catch((e) => {
					throw new __errors_ts.DrizzleQueryError(queryString, params, e);
				});
				await cache.put(key, result, autoInvalidate ? tables : [], isTag, config);
				return result;
			}
			return fromCache;
		}
		(0, __utils_ts.assertUnreachable)(cacheStrat);
	}
};
var PgAsyncSession = class extends require_pg_core_session.PgSession {
	static [__entity_ts.entityKind] = "PgAsyncSession";
	execute(query) {
		return __tracing_ts.tracer.startActiveSpan("drizzle.operation", () => {
			return __tracing_ts.tracer.startActiveSpan("drizzle.prepareQuery", () => {
				return this.prepareQuery(this.dialect.sqlToQuery(query), "raw", false);
			}).execute();
		});
	}
	arrays(query) {
		return __tracing_ts.tracer.startActiveSpan("drizzle.operation", () => {
			return __tracing_ts.tracer.startActiveSpan("drizzle.prepareQuery", () => {
				return this.prepareQuery(this.dialect.sqlToQuery(query), "arrays", false);
			}).execute();
		});
	}
	objects(query) {
		return __tracing_ts.tracer.startActiveSpan("drizzle.operation", () => {
			return __tracing_ts.tracer.startActiveSpan("drizzle.prepareQuery", () => {
				return this.prepareQuery(this.dialect.sqlToQuery(query), "objects", false);
			}).execute();
		});
	}
};
var PgAsyncTransaction = class extends require_pg_core_async_db.PgAsyncDatabase {
	static [__entity_ts.entityKind] = "PgAsyncTransaction";
	constructor(dialect, session, relations, nestedIndex = 0, parseRqbJson) {
		super(dialect, session, relations, parseRqbJson);
		this.nestedIndex = nestedIndex;
	}
	rollback() {
		throw new __errors_ts.TransactionRollbackError();
	}
	/** @internal */
	getTransactionConfigSQL(config) {
		const chunks = [];
		if (config.isolationLevel) chunks.push(`isolation level ${config.isolationLevel}`);
		if (config.accessMode) chunks.push(config.accessMode);
		if (typeof config.deferrable === "boolean") chunks.push(config.deferrable ? "deferrable" : "not deferrable");
		return __sql_sql_ts.sql.raw(chunks.join(" "));
	}
	setTransaction(config) {
		return this.session.execute(__sql_sql_ts.sql`set transaction ${this.getTransactionConfigSQL(config)}`);
	}
};
async function migrate(migrations, db, config) {
	const migrationsTable = typeof config === "string" ? "__drizzle_migrations" : config.migrationsTable ?? "__drizzle_migrations";
	const migrationsSchema = typeof config === "string" ? "drizzle" : config.migrationsSchema ?? "drizzle";
	await db.execute(__sql_sql_ts.sql`CREATE SCHEMA IF NOT EXISTS ${__sql_sql_ts.sql.identifier(migrationsSchema)}`);
	const { newDb } = await (0, __up_migrations_pg_ts.upgradeIfNeeded)(migrationsSchema, migrationsTable, db, migrations);
	if (newDb) {
		const migrationTableCreate = __sql_sql_ts.sql`
			CREATE TABLE IF NOT EXISTS ${__sql_sql_ts.sql.identifier(migrationsSchema)}.${__sql_sql_ts.sql.identifier(migrationsTable)} (
				id SERIAL PRIMARY KEY,
				hash text NOT NULL,
				created_at bigint,
				name text,
				applied_at timestamp with time zone DEFAULT now()
			)
		`;
		await db.execute(migrationTableCreate);
	}
	const dbMigrations = await db.session.objects(__sql_sql_ts.sql`select id, hash, created_at, name from ${__sql_sql_ts.sql.identifier(migrationsSchema)}.${__sql_sql_ts.sql.identifier(migrationsTable)}`);
	if (typeof config === "object" && config.init) {
		if (dbMigrations.length) return { exitCode: "databaseMigrations" };
		if (migrations.length > 1) return { exitCode: "localMigrations" };
		const [migration] = migrations;
		if (!migration) return;
		await db.execute(__sql_sql_ts.sql`insert into ${__sql_sql_ts.sql.identifier(migrationsSchema)}.${__sql_sql_ts.sql.identifier(migrationsTable)} ("hash", "created_at", "name") values(${migration.hash}, ${migration.folderMillis}, ${migration.name ?? null})`);
		return;
	}
	const migrationsToRun = (0, __migrator_utils_ts.getMigrationsToRun)({
		localMigrations: migrations,
		dbMigrations
	});
	await db.transaction(async (tx) => {
		for (const migration of migrationsToRun) {
			for (const stmt of migration.sql) await tx.execute(__sql_sql_ts.sql.raw(stmt));
			await tx.execute(__sql_sql_ts.sql`insert into ${__sql_sql_ts.sql.identifier(migrationsSchema)}.${__sql_sql_ts.sql.identifier(migrationsTable)} ("hash", "created_at", "name") values(${migration.hash}, ${migration.folderMillis}, ${migration.name ?? null})`);
		}
	});
}

//#endregion
exports.PgAsyncPreparedQuery = PgAsyncPreparedQuery;
exports.PgAsyncSession = PgAsyncSession;
exports.PgAsyncTransaction = PgAsyncTransaction;
exports.migrate = migrate;
