// waddler/sqlite/durable-sqlite/session.js
import { WaddlerQueryError } from "../../errors/index.js";
import { SQLTemplate } from "../../sql-template.js";
class DurableSqliteSQLTemplate extends SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
  }
  returningData = true;
  all() {
    this.returningData = true;
    return this;
  }
  run() {
    this.returningData = false;
    return this;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      if (this.returningData) {
        if (this.options.rowMode === "array") {
          const res = params.length > 0 ? this.client.sql.exec(query, ...params) : this.client.sql.exec(query);
          finalMetadata = { columnNames: res.columnNames, rowsRead: res.rowsRead, rowsWritten: res.rowsWritten };
          finalRes = res.raw().toArray();
        } else {
          const res = params.length > 0 ? this.client.sql.exec(query, ...params) : this.client.sql.exec(query);
          finalMetadata = { columnNames: res.columnNames, rowsRead: res.rowsRead, rowsWritten: res.rowsWritten };
          finalRes = res.toArray();
        }
      } else {
        const res = params.length > 0 ? this.client.sql.exec(query, ...params) : this.client.sql.exec(query);
        finalMetadata = { columnNames: res.columnNames, rowsRead: res.rowsRead, rowsWritten: res.rowsWritten };
        finalRes = res;
      }
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  async *stream() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    try {
      if (this.options.rowMode === "array") {
        const res = params.length > 0 ? this.client.sql.exec(query, ...params) : this.client.sql.exec(query);
        const stream2 = res.raw();
        for await (const row of stream2) {
          yield row;
        }
        return;
      }
      const stream = params.length > 0 ? this.client.sql.exec(query, ...params) : this.client.sql.exec(query);
      for await (const row of stream) {
        yield row;
      }
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    }
  }
}
export {
  DurableSqliteSQLTemplate
};
