// @netlify/dev/dist/main.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/main.ts
var main_exports = {};
__export(main_exports, {
  NetlifyDev: () => NetlifyDev,
  applyMigrations: () => import_database_dev2.applyMigrations,
  initializeTrackingTable: () => import_database_dev2.initializeTrackingTable,
  resetDatabase: () => import_database_dev2.resetDatabase
});
module.exports = __toCommonJS(main_exports);
var import_node_fs2 = require("fs");
var import_node_path2 = __toESM(require("path"), 1);
var import_node_process2 = __toESM(require("process"), 1);
var import_bootstrap = require("@netlify/ai/bootstrap");
var import_config = require("@netlify/config");
var import_dev_utils = require("@netlify/dev-utils");
var import_edge_functions_dev = require("@netlify/edge-functions-dev");
var import_functions_dev = require("@netlify/functions-dev");
var import_headers = require("@netlify/headers");
var import_images = require("@netlify/images");
var import_redirects = require("@netlify/redirects");
var import_static = require("@netlify/static");
var import_database_dev = require("@netlify/database-dev");
var import_database_dev2 = require("@netlify/database-dev");

// src/lib/env.ts
var SUPPORTED_CONTEXTS = ["all", "production", "deploy-preview", "branch-deploy", "dev", "dev-server"];
var injectEnvVariables = async ({
  accountSlug,
  baseVariables = {},
  envAPI,
  netlifyAPI,
  siteID
}) => {
  const results = {};
  let variables = baseVariables;
  if (netlifyAPI && siteID && accountSlug) {
    variables = await getEnvelopeEnv({
      accountId: accountSlug,
      api: netlifyAPI,
      env: baseVariables,
      siteId: siteID
    });
  }
  for (const [key, variable] of Object.entries(variables)) {
    const existsInProcess = envAPI.has(key);
    const [usedSource, ...overriddenSources] = existsInProcess ? ["process", ...variable.sources] : variable.sources;
    const isInternal = variable.sources.includes("internal");
    const result = {
      isInternal,
      originalValue: envAPI.get(key),
      overriddenSources,
      usedSource,
      value: variable.value
    };
    if (!existsInProcess || isInternal) {
      envAPI.set(key, variable.value);
    }
    results[key] = result;
  }
  return results;
};
var getValueForContext = (values, contextOrBranch) => {
  const isSupportedContext = SUPPORTED_CONTEXTS.includes(contextOrBranch);
  if (!isSupportedContext) {
    const valueMatchingAsBranch = values.find((val) => val.context_parameter === contextOrBranch);
    if (valueMatchingAsBranch != null) {
      return valueMatchingAsBranch;
    }
    const valueMatchingContext = values.find((val) => val.context === "all" || val.context === "branch-deploy");
    return valueMatchingContext ?? void 0;
  }
  const valueMatchingAsContext = values.find((val) => val.context === "all" || val.context === contextOrBranch);
  return valueMatchingAsContext ?? void 0;
};
var filterEnvBySource = (env, source) => Object.fromEntries(Object.entries(env).filter(([, variable]) => variable.sources[0] === source));
var fetchEnvelopeItems = async function({
  accountId,
  api,
  key,
  siteId
}) {
  if (accountId === void 0) {
    return [];
  }
  try {
    if (key) {
      const envelopeItem = await api.getEnvVar({ accountId, key, siteId });
      return [envelopeItem];
    }
    const envelopeItems = await api.getEnvVars({ accountId, siteId });
    return envelopeItems;
  } catch {
    return [];
  }
};
var formatEnvelopeData = ({
  context = "dev",
  envelopeItems = [],
  scope = "any",
  source
}) => envelopeItems.filter(({ values }) => Boolean(getValueForContext(values, context))).filter(({ scopes }) => scope === "any" ? true : scopes.includes(scope)).sort((left, right) => left.key.toLowerCase() < right.key.toLowerCase() ? -1 : 1).reduce((acc, cur) => {
  const val = getValueForContext(cur.values, context);
  if (val === void 0) {
    throw new TypeError(`failed to locate environment variable value for ${context} context`);
  }
  const { context: itemContext, context_parameter: branch, value } = val;
  return {
    ...acc,
    [cur.key]: {
      context: itemContext,
      branch,
      scopes: cur.scopes,
      sources: [source],
      value
    }
  };
}, {});
var getEnvelopeEnv = async ({
  accountId,
  api,
  context = "dev",
  env,
  key = "",
  raw = false,
  scope = "any",
  siteId
}) => {
  const [accountEnvelopeItems, siteEnvelopeItems] = await Promise.all([
    fetchEnvelopeItems({ api, accountId, key }),
    fetchEnvelopeItems({ api, accountId, key, siteId })
  ]);
  const accountEnv = formatEnvelopeData({ context, envelopeItems: accountEnvelopeItems, scope, source: "account" });
  const siteEnv = formatEnvelopeData({ context, envelopeItems: siteEnvelopeItems, scope, source: "ui" });
  if (raw) {
    const entries = Object.entries({ ...accountEnv, ...siteEnv });
    return entries.reduce(
      (obj, [envVarKey, metadata]) => ({
        ...obj,
        [envVarKey]: metadata.value
      }),
      {}
    );
  }
  const generalEnv = filterEnvBySource(env, "general");
  const internalEnv = filterEnvBySource(env, "internal");
  const addonsEnv = filterEnvBySource(env, "addons");
  const configFileEnv = filterEnvBySource(env, "configFile");
  const includeConfigEnvVars = /any|builds|post[-_]processing/.test(scope);
  return {
    ...generalEnv,
    ...accountEnv,
    ...includeConfigEnvVars ? addonsEnv : {},
    ...siteEnv,
    ...includeConfigEnvVars ? configFileEnv : {},
    ...internalEnv
  };
};

// src/lib/fs.ts
var import_node_fs = require("fs");
var isDirectory = async (path3) => {
  try {
    const stat = await import_node_fs.promises.stat(path3);
    return stat.isDirectory();
  } catch {
  }
  return false;
};
var isFile = async (path3) => {
  try {
    const stat = await import_node_fs.promises.stat(path3);
    return stat.isFile();
  } catch {
  }
  return false;
};

// src/lib/reqres.ts
var import_node_stream = require("stream");
var normalizeHeaders = (request) => {
  const result = [];
  let headers = request.headers;
  if (request.httpVersionMajor >= 2) {
    headers = { ...headers };
    delete headers[":authority"];
    delete headers[":method"];
    delete headers[":path"];
    delete headers[":scheme"];
  }
  for (const [key, value] of Object.entries(headers)) {
    if (Array.isArray(value)) {
      result.push([key, value.join(",")]);
    } else if (typeof value === "string") {
      result.push([key, value]);
    }
  }
  return result;
};
var getNormalizedRequest = (input, requestID, removeBody) => {
  const method = input.method.toUpperCase();
  const headers = input.headers;
  headers.set("x-nf-request-id", requestID);
  return new Request(input.url, {
    body: method === "GET" || method === "HEAD" || removeBody ? null : input.body,
    // @ts-expect-error Not typed!
    duplex: "half",
    headers,
    method
  });
};
var getNormalizedRequestFromNodeRequest = (input, requestID, removeBody) => {
  const { headers, url = "" } = input;
  const origin = `http://${headers.host ?? "localhost"}`;
  const fullUrl = new URL(url, origin);
  const method = input.method?.toUpperCase() ?? "GET";
  const body = input.method === "GET" || input.method === "HEAD" || removeBody ? null : import_node_stream.Readable.toWeb(input);
  const normalizedHeaders = normalizeHeaders(input);
  normalizedHeaders.push(["x-nf-request-id", requestID]);
  return new Request(fullUrl, {
    body,
    // @ts-expect-error Not typed!
    duplex: "half",
    headers: normalizedHeaders,
    method
  });
};

// src/lib/request_id.ts
var import_ulid = require("ulid");
var generateRequestID = () => (0, import_ulid.ulid)();

// src/lib/runtime.ts
var import_node_path = __toESM(require("path"), 1);
var import_node_process = __toESM(require("process"), 1);
var import_server = require("@netlify/blobs/server");
var import_runtime = require("@netlify/runtime");
var restoreEnvironment = (snapshot) => {
  for (const key in snapshot) {
    if (snapshot[key] === void 0) {
      delete import_node_process.default.env[key];
    } else {
      import_node_process.default.env[key] = snapshot[key];
    }
  }
};
var getRuntime = async ({ blobs, deployID, projectRoot, siteID }) => {
  const blobsToken = Math.random().toString().slice(2);
  const blobsServer = blobs ? new import_server.BlobsServer({
    directory: import_node_path.default.join(projectRoot, ".netlify", "blobs-serve"),
    token: blobsToken
  }) : null;
  const blobsServerDetails = await blobsServer?.start();
  const envSnapshot = {};
  const env = {
    delete: (key) => {
      envSnapshot[key] = envSnapshot[key] || import_node_process.default.env[key];
      delete import_node_process.default.env[key];
    },
    get: (key) => import_node_process.default.env[key],
    has: (key) => Boolean(import_node_process.default.env[key]),
    set: (key, value) => {
      envSnapshot[key] = envSnapshot[key] || import_node_process.default.env[key];
      import_node_process.default.env[key] = value;
    },
    toObject: () => import_node_process.default.env
  };
  (0, import_runtime.startRuntime)({
    blobs: blobsServerDetails ? {
      edgeURL: `http://localhost:${blobsServerDetails.port}`,
      uncachedEdgeURL: `http://localhost:${blobsServerDetails.port}`,
      primaryRegion: "us-east-2",
      token: blobsToken
    } : void 0,
    cache: {
      getCacheAPIContext: () => null,
      purgeToken: ""
    },
    deployID,
    env,
    getRequestContext: () => null,
    siteID
  });
  return {
    env,
    envSnapshot,
    stop: async () => {
      restoreEnvironment(envSnapshot);
      await blobsServer?.stop();
    }
  };
};

// src/main.ts
var notFoundHandler = async () => new Response("Not found", { status: 404 });
var NetlifyDev = class {
  #apiHost;
  #apiScheme;
  #apiToken;
  #cleanupJobs;
  #edgeFunctionsHandler;
  #functionsHandler;
  #geolocationConfig;
  #functionsServePath;
  #config;
  #features;
  #db;
  #headersHandler;
  #imageRemoteURLPatterns;
  #imageHandler;
  #logger;
  #projectRoot;
  #redirectsHandler;
  #serverAddress;
  #siteID;
  #staticHandler;
  #staticHandlerAdditionalDirectories;
  #skipGitignore;
  constructor(options) {
    if (options.apiURL) {
      const apiURL = new URL(options.apiURL);
      this.#apiHost = apiURL.host;
      this.#apiScheme = apiURL.protocol.slice(0, -1);
    }
    const projectRoot = options.projectRoot ?? import_node_process2.default.cwd();
    this.#apiToken = options.apiToken;
    this.#cleanupJobs = [];
    this.#geolocationConfig = options.geolocation;
    this.#features = {
      aiGateway: options.aiGateway?.enabled !== false,
      blobs: options.blobs?.enabled !== false,
      database: options.database?.enabled !== false,
      edgeFunctions: options.edgeFunctions?.enabled !== false,
      environmentVariables: options.environmentVariables?.enabled !== false,
      functions: options.functions?.enabled !== false,
      geolocation: options.geolocation?.enabled !== false,
      headers: options.headers?.enabled !== false,
      images: options.images?.enabled !== false,
      redirects: options.redirects?.enabled !== false,
      static: options.staticFiles?.enabled !== false
    };
    this.#functionsServePath = import_node_path2.default.join(projectRoot, ".netlify", "functions-serve");
    this.#imageRemoteURLPatterns = options.images?.remoteURLPatterns ?? [];
    this.#logger = options.logger ?? globalThis.console;
    this.#serverAddress = options.serverAddress;
    this.#projectRoot = projectRoot;
    this.#staticHandlerAdditionalDirectories = options.staticFiles?.directories ?? [];
    this.#skipGitignore = options.skipGitignore ?? false;
  }
  getServerAddress(requestServerAddress) {
    if (requestServerAddress) {
      return requestServerAddress;
    }
    if (typeof this.#serverAddress === "string") {
      return this.#serverAddress;
    }
    throw new Error("Server address is not defined");
  }
  /**
   * Runs a request through the Netlify request chain and returns a `Response`
   * if there's a match. We must not disturb the incoming request unless we
   * know we will be returning a response, so this method takes a read-only
   * request that is safe to access (used for matching) and a getter for the
   * actual request (used for handling matches).
   *
   * @param readRequest Read-only version of the request (without a body)
   * @param getWriteRequest Getter for the actual request (with a body)
   * @param destPath Destination directory for compiled files
   * @param options Options object
   * @returns
   */
  async handleInEphemeralDirectory(readRequest, getWriteRequest, destPath, options = {}) {
    const serverAddress = this.getServerAddress(options.serverAddress);
    const edgeFunctionMatch = await this.#edgeFunctionsHandler?.match(readRequest);
    if (edgeFunctionMatch) {
      return {
        response: await edgeFunctionMatch.handle(getWriteRequest(), serverAddress),
        type: "edge-function"
      };
    }
    const imageMatch = this.#imageHandler?.match(readRequest);
    if (imageMatch) {
      const response = await imageMatch.handle(serverAddress);
      return { response, type: "image" };
    }
    const functionMatch = await this.#functionsHandler?.match(readRequest, destPath);
    if (functionMatch) {
      if (functionMatch.preferStatic) {
        const staticMatch2 = await this.#staticHandler?.match(readRequest);
        if (staticMatch2) {
          const response = await staticMatch2.handle();
          await this.#headersHandler?.apply(readRequest, response, options.headersCollector);
          return { response, type: "static" };
        }
      }
      return { response: await functionMatch.handle(getWriteRequest()), type: "function" };
    }
    const redirectMatch = await this.#redirectsHandler?.match(readRequest);
    if (redirectMatch) {
      const redirectRequest = new Request(redirectMatch.target);
      const imageMatch2 = this.#imageHandler?.match(redirectRequest);
      if (imageMatch2) {
        const response2 = await imageMatch2.handle(serverAddress);
        return { response: response2, type: "image" };
      }
      const functionMatch2 = await this.#functionsHandler?.match(redirectRequest, destPath);
      if (functionMatch2 && !functionMatch2.preferStatic) {
        return {
          response: await functionMatch2.handle(getWriteRequest()),
          type: "function"
        };
      }
      const response = await this.#redirectsHandler?.handle(
        getWriteRequest(),
        redirectMatch,
        async (maybeStaticFile) => {
          const staticMatch2 = await this.#staticHandler?.match(maybeStaticFile);
          if (!staticMatch2) {
            return;
          }
          return async () => {
            const response2 = await staticMatch2.handle();
            await this.#headersHandler?.apply(new Request(redirectMatch.target), response2, options.headersCollector);
            return response2;
          };
        }
      );
      if (response) {
        return { response, type: "redirect" };
      }
    }
    const staticMatch = await this.#staticHandler?.match(readRequest);
    if (staticMatch) {
      const response = await staticMatch.handle();
      await this.#headersHandler?.apply(readRequest, response, options.headersCollector);
      return { response, type: "static" };
    }
  }
  async getConfig() {
    const configFilePath = import_node_path2.default.resolve(this.#projectRoot, "netlify.toml");
    const configFileExists = await isFile(configFilePath);
    const config = await (0, import_config.resolveConfig)({
      config: configFileExists ? configFilePath : void 0,
      context: "dev",
      cwd: import_node_process2.default.cwd(),
      host: this.#apiHost,
      offline: !this.#siteID,
      mode: "cli",
      repositoryRoot: this.#projectRoot,
      scheme: this.#apiScheme,
      siteId: this.#siteID,
      token: this.#apiToken
    });
    return config;
  }
  /**
   * Runs a `Request` through the Netlify request chain. If there is a match,
   * it returns the resulting `Response` object; if not, it returns `undefined`.
   */
  async handle(request, options = {}) {
    const result = await this.handleAndIntrospect(request, options);
    return result?.response;
  }
  /**
   * Runs a `Request` through the Netlify request chain. If there is a match,
   * it returns an object with the resulting `Response` object and information
   * about the match; if not, it returns `undefined`.
   */
  async handleAndIntrospect(request, options = {}) {
    await import_node_fs2.promises.mkdir(this.#functionsServePath, { recursive: true });
    const destPath = await import_node_fs2.promises.mkdtemp(import_node_path2.default.join(this.#functionsServePath, `_`));
    const requestID = generateRequestID();
    const matchRequest = getNormalizedRequest(request, requestID, true);
    const getHandleRequest = () => getNormalizedRequest(request, requestID, false);
    try {
      return await this.handleInEphemeralDirectory(matchRequest, getHandleRequest, destPath, options);
    } finally {
      try {
        await import_node_fs2.promises.rm(destPath, { force: true, recursive: true });
      } catch {
      }
    }
  }
  /**
   * Runs a Node.js `IncomingMessage` through the Netlify request chain. If
   * there is a match, it returns an object with the resulting `Response`
   * object and information about the match; if not, it returns `undefined`.
   */
  async handleAndIntrospectNodeRequest(request, options = {}) {
    await import_node_fs2.promises.mkdir(this.#functionsServePath, { recursive: true });
    const destPath = await import_node_fs2.promises.mkdtemp(import_node_path2.default.join(this.#functionsServePath, `_`));
    const requestID = generateRequestID();
    const matchRequest = getNormalizedRequestFromNodeRequest(request, requestID, true);
    const getHandleRequest = () => getNormalizedRequestFromNodeRequest(request, requestID, false);
    try {
      return await this.handleInEphemeralDirectory(matchRequest, getHandleRequest, destPath, options);
    } finally {
      try {
        await import_node_fs2.promises.rm(destPath, { force: true, recursive: true });
      } catch {
      }
    }
  }
  get siteIsLinked() {
    return Boolean(this.#siteID);
  }
  async start() {
    if (!this.#skipGitignore) {
      await (0, import_dev_utils.ensureNetlifyIgnore)(this.#projectRoot, this.#logger);
    }
    this.#apiToken = this.#apiToken ?? await (0, import_dev_utils.getAPIToken)();
    const state = new import_dev_utils.LocalState(this.#projectRoot);
    const siteID = state.get("siteId");
    this.#siteID = siteID;
    const config = await this.getConfig();
    this.#config = config;
    const reactiveConfig = new import_dev_utils.Reactive(config);
    const fileWatcher = new import_dev_utils.FileWatcher();
    this.#cleanupJobs.push(() => fileWatcher.close());
    if (config.configPath) {
      const reloadConfig = () => {
        void (async () => {
          try {
            const newConfig = await this.getConfig();
            this.#config = newConfig;
            reactiveConfig.set(newConfig);
          } catch (error) {
            this.#logger.warn(`Failed to reload config: ${String(error)}`);
          }
        })();
      };
      fileWatcher.subscribe({
        id: "netlify-config",
        paths: config.configPath,
        onChange: reloadConfig,
        onAdd: reloadConfig,
        onUnlink: reloadConfig
      });
    }
    const runtime = await getRuntime({
      blobs: this.#features.blobs,
      deployID: "0",
      projectRoot: this.#projectRoot,
      siteID: siteID ?? "0"
    });
    this.#cleanupJobs.push(() => runtime.stop());
    if (this.#features.database) {
      try {
        const dbDirectory = import_node_path2.default.join(this.#projectRoot, ".netlify", "db");
        const db = new import_database_dev.NetlifyDB({ directory: dbDirectory });
        const connectionString = await db.start();
        runtime.env.set("NETLIFY_DB_URL", connectionString);
        runtime.env.set("NETLIFY_DB_DRIVER", "server");
        state.set("dbConnectionString", connectionString);
        this.#db = db;
        this.#cleanupJobs.push(async () => {
          await db.stop();
          state.delete("dbConnectionString");
        });
      } catch (error) {
        this.#db = void 0;
        this.#logger.warn(`Failed to start Netlify Database locally: ${String(error)}`);
      }
    }
    if (this.#features.aiGateway && config?.siteInfo?.capabilities?.ai_gateway_disabled) {
      this.#features.aiGateway = false;
    }
    if (this.#features.aiGateway && this.#features.environmentVariables && config?.api) {
      await (0, import_bootstrap.setupAIGateway)({
        api: config.api,
        env: config.env || {},
        siteID,
        siteURL: config.siteInfo?.url,
        accountID: config.siteInfo?.account_id,
        siteHasDeploy: !!config.siteInfo?.published_deploy
      });
      if (config.env.AI_GATEWAY) {
        runtime.env.set("AI_GATEWAY", config.env.AI_GATEWAY.value);
        const aiGatewayContext = (0, import_bootstrap.parseAIGatewayContext)(config.env.AI_GATEWAY.value);
        if (aiGatewayContext) {
          runtime.env.set("NETLIFY_AI_GATEWAY_KEY", aiGatewayContext.token);
          runtime.env.set("NETLIFY_AI_GATEWAY_URL", aiGatewayContext.url);
          if (aiGatewayContext.envVars) {
            for (const envVar of aiGatewayContext.envVars) {
              runtime.env.set(envVar.key, aiGatewayContext.token);
              runtime.env.set(envVar.url, aiGatewayContext.url);
            }
          }
        }
      }
    }
    let serverAddress;
    if (typeof this.#serverAddress === "string") {
      serverAddress = this.#serverAddress;
    } else if (this.#serverAddress !== null && (this.#features.edgeFunctions || this.#features.images)) {
      const passthroughServer = new import_dev_utils.HTTPServer(async (req) => {
        const res = await this.handle(req);
        return res ?? new Response(null, { status: 404 });
      });
      this.#cleanupJobs.push(() => passthroughServer.stop());
      serverAddress = await passthroughServer.start();
      this.#serverAddress = serverAddress;
    }
    let envVariables = {};
    if (this.#features.environmentVariables) {
      envVariables = await injectEnvVariables({
        accountSlug: config?.siteInfo?.account_slug,
        baseVariables: config?.env || {},
        envAPI: runtime.env,
        netlifyAPI: config?.api,
        siteID
      });
    }
    let geolocation;
    if (this.#features.edgeFunctions) {
      const edgeFunctionsEnv = {
        // User-defined env vars + documented runtime env vars
        ...Object.entries(envVariables).reduce(
          (acc, [key, variable]) => ({
            ...acc,
            [key]: variable.value
          }),
          {}
        ),
        // Add runtime env vars that we've set ourselves so far. These are "internal" env vars,
        // part of the runtime emulation. They've already been populated on this process's env, which
        // is needed to make other dev features work. These are different than the "documented" runtime
        // env vars, in that they are implementation details, needed to make our features work.
        ...Object.keys(runtime.envSnapshot).reduce(
          (acc, key) => ({
            ...acc,
            [key]: runtime.env.get(key) ?? ""
          }),
          {}
        )
      };
      geolocation ??= await (0, import_dev_utils.getGeoLocation)({
        enabled: this.#features.geolocation,
        cache: this.#geolocationConfig?.cache ?? true,
        state
      });
      const edgeFunctionsHandler = new import_edge_functions_dev.EdgeFunctionsHandler({
        configDeclarations: this.#config?.config.edge_functions ?? [],
        directories: [this.#config?.config.build.edge_functions].filter(Boolean),
        env: edgeFunctionsEnv,
        geolocation,
        logger: this.#logger,
        siteID,
        siteName: config?.siteInfo.name
      });
      this.#edgeFunctionsHandler = edgeFunctionsHandler;
      this.#cleanupJobs.push(() => edgeFunctionsHandler.stop());
    }
    if (this.#features.functions) {
      const userFunctionsPath = this.#config?.config.functionsDirectory ?? import_node_path2.default.join(this.#projectRoot, "netlify/functions");
      const userFunctionsPathExists = await isDirectory(userFunctionsPath);
      geolocation ??= await (0, import_dev_utils.getGeoLocation)({
        enabled: this.#features.geolocation,
        cache: this.#geolocationConfig?.cache ?? true,
        state
      });
      this.#functionsHandler = new import_functions_dev.FunctionsHandler({
        config: reactiveConfig,
        destPath: this.#functionsServePath,
        fileWatcher,
        geolocation,
        projectRoot: this.#projectRoot,
        settings: {},
        siteId: this.#siteID,
        userFunctionsPath: userFunctionsPathExists ? userFunctionsPath : void 0
      });
    }
    if (this.#features.headers) {
      this.#headersHandler = new import_headers.HeadersHandler({
        configPath: this.#config?.configPath,
        configHeaders: this.#config?.config.headers,
        projectDir: this.#projectRoot,
        publishDir: this.#config?.config.build.publish ?? void 0,
        logger: this.#logger
      });
    }
    if (this.#features.redirects) {
      this.#redirectsHandler = new import_redirects.RedirectsHandler({
        configPath: this.#config?.configPath,
        configRedirects: this.#config?.config.redirects,
        jwtRoleClaim: "",
        jwtSecret: "",
        notFoundHandler,
        projectDir: this.#projectRoot
      });
    }
    if (this.#features.static) {
      this.#staticHandler = new import_static.StaticHandler({
        directory: [
          this.#config?.config.build.publish ?? this.#projectRoot,
          ...this.#staticHandlerAdditionalDirectories
        ]
      });
    }
    if (this.#features.images) {
      const remoteImages = [...this.#config?.config?.images?.remote_images ?? [], ...this.#imageRemoteURLPatterns];
      this.#imageHandler = new import_images.ImageHandler({
        imagesConfig: {
          remote_images: remoteImages
        },
        logger: this.#logger
      });
    }
    return {
      serverAddress
    };
  }
  get db() {
    return this.#db;
  }
  async stop() {
    await Promise.allSettled(this.#cleanupJobs.map((task) => task()));
  }
  getEnabledFeatures() {
    return Object.entries(this.#features).filter(([_, enabled]) => enabled).map(([feature]) => feature);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  NetlifyDev,
  applyMigrations,
  initializeTrackingTable,
  resetDatabase
});
