// waddler/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  ClickHouseDialect: () => import_dialect.ClickHouseDialect,
  DuckdbDialect: () => import_dialect2.DuckdbDialect,
  MySQLDialect: () => import_dialect3.MySQLDialect,
  PgDialect: () => import_dialect4.PgDialect,
  SqliteDialect: () => import_dialect5.SqliteDialect,
  createRecyclingPool: () => import_recycling_pool.createRecyclingPool
});
module.exports = __toCommonJS(index_exports);
var import_dialect = require("./clickhouse-core/dialect.cjs");
var import_dialect2 = require("./duckdb-core/dialect.cjs");
var import_dialect3 = require("./mysql/mysql-core/dialect.cjs");
var import_dialect4 = require("./pg/pg-core/dialect.cjs");
var import_recycling_pool = require("./recycling-pool.cjs");
__reExport(index_exports, require("./sql-template-params.cjs"), module.exports);
__reExport(index_exports, require("./sql-template.cjs"), module.exports);
__reExport(index_exports, require("./sql.cjs"), module.exports);
var import_dialect5 = require("./sqlite/sqlite-core/dialect.cjs");
__reExport(index_exports, require("./types.cjs"), module.exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ClickHouseDialect,
  DuckdbDialect,
  MySQLDialect,
  PgDialect,
  SqliteDialect,
  createRecyclingPool,
  ...require("./sql-template-params.cjs"),
  ...require("./sql-template.cjs"),
  ...require("./sql.cjs"),
  ...require("./types.cjs")
});
