// waddler/logger.js
class ConsoleLogWriter {
  write(message) {
    console.log(message);
  }
}
class DefaultLogger {
  writer;
  constructor(config) {
    this.writer = config?.writer ?? new ConsoleLogWriter();
  }
  logQuery(query, params, metadata) {
    let paramsStr = "";
    if (Array.isArray(params)) {
      const stringifiedParams = params.map((p) => {
        try {
          return JSON.stringify(p);
        } catch {
          return String(p);
        }
      });
      paramsStr = stringifiedParams.length ? ` -- params: [${stringifiedParams.join(", ")}]` : "";
    } else {
      const stringifiedParams = {};
      let paramsCount = 0;
      for (const [key, value] of Object.entries(params)) {
        try {
          stringifiedParams[key] = JSON.stringify(value);
          paramsCount++;
        } catch {
          stringifiedParams[key] = String(value);
          paramsCount++;
        }
      }
      paramsStr = paramsCount ? ` -- params: ${JSON.stringify(stringifiedParams)}` : "";
    }
    const metadataStr = metadata === void 0 ? "" : ` --metadata: ${JSON.stringify(metadata)}`;
    this.writer.write(`Query: ${query}${paramsStr}${metadataStr}`);
  }
}
class NoopLogger {
  logQuery() {
  }
}
export {
  ConsoleLogWriter,
  DefaultLogger,
  NoopLogger
};
