// waddler/sql-template.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var sql_template_exports = {};
__export(sql_template_exports, {
  SQLTemplate: () => SQLTemplate
});
module.exports = __toCommonJS(sql_template_exports);
var import_logger = require("./logger.cjs");
class SQLTemplate {
  constructor(sqlWrapper, dialect, configOptions = {}) {
    this.sqlWrapper = sqlWrapper;
    this.dialect = dialect;
    this.configOptions = configOptions;
    this.logger = configOptions.logger ?? new import_logger.NoopLogger();
  }
  logger;
  append(other) {
    this.sqlWrapper.append(other.sqlWrapper);
    this.sqlWrapper.recalculateQuery(this.dialect);
  }
  toSQL() {
    return this.sqlWrapper.getQuery(this.dialect);
  }
  // Allow it to be awaited (like a Promise)
  then(onfulfilled, onrejected) {
    const result = this.execute();
    return Promise.resolve(result).then(onfulfilled, onrejected);
  }
  async *chunked(chunkSize = 1) {
    let rows = [];
    let row;
    const asyncIterator = this.stream();
    let iterResult = await asyncIterator.next();
    while (!iterResult.done) {
      row = iterResult.value;
      rows.push(row);
      if (rows.length % chunkSize === 0) {
        yield rows;
        rows = [];
      }
      iterResult = await asyncIterator.next();
    }
    if (rows.length !== 0) {
      yield rows;
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SQLTemplate
});
