// waddler/sql-template-params.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var sql_template_params_exports = {};
__export(sql_template_params_exports, {
  Dialect: () => Dialect,
  SQLChunk: () => SQLChunk,
  SQLCommonParam: () => SQLCommonParam,
  SQLDefault: () => SQLDefault,
  SQLIdentifier: () => SQLIdentifier,
  SQLQuery: () => SQLQuery,
  SQLRaw: () => SQLRaw,
  SQLString: () => SQLString,
  SQLValues: () => SQLValues
});
module.exports = __toCommonJS(sql_template_params_exports);
class Dialect {
  createEmptyParams() {
    return [];
  }
  pushParams(params, param, _lastParamIdx, mode = "bulk") {
    if (mode === "bulk") params.push(...param);
    else params.push(param);
  }
}
class SQLChunk {
}
class SQLQuery extends SQLChunk {
  constructor(sqlWrapper, dialect) {
    super();
    this.sqlWrapper = sqlWrapper;
    this.dialect = dialect;
  }
  generateSQL() {
    this.sqlWrapper.prepareQuery(this.dialect);
    return this.sqlWrapper.getQuery(this.dialect);
  }
  append(other) {
    this.sqlWrapper.append(other.sqlWrapper);
  }
  toSQL() {
    return this.generateSQL();
  }
}
class SQLCommonParam extends SQLChunk {
  constructor(value, type = "String") {
    super();
    this.value = value;
    this.type = type;
  }
  INT32_MAX = 2147483647;
  INT32_MIN = -2147483648;
  generateSQL({ dialect, lastParamIdx }) {
    if (typeof this.value === "bigint") this.type = "Int64";
    if (typeof this.value === "number" && this.value % 1 === 0) {
      this.type = "Int32";
      if (this.value > this.INT32_MAX || this.value < this.INT32_MIN) {
        this.type = "Int64";
      }
    }
    if (Array.isArray(this.value)) {
      const nodeType = typeof this.value[0];
      if (nodeType === "string") this.type = "Array(String)";
    }
    const params = dialect.createEmptyParams();
    dialect.pushParams(params, this.value, lastParamIdx + 1, "single");
    return {
      sql: dialect.escapeParam(lastParamIdx + 1, this.type),
      params,
      paramsCount: 1
    };
  }
}
class SQLString extends SQLChunk {
  constructor(value) {
    super();
    this.value = value;
  }
  generateSQL() {
    return { sql: this.value };
  }
}
class SQLIdentifier extends SQLChunk {
  constructor(value) {
    super();
    this.value = value;
  }
  objectToSQL(object, dialect) {
    dialect.checkIdentifierObject(object);
    const chunks = [];
    if (object.schema !== void 0) chunks.push(`${dialect.escapeIdentifier(object.schema)}`);
    if (object.table !== void 0) chunks.push(`${dialect.escapeIdentifier(object.table)}`);
    if (object.column !== void 0) chunks.push(`${dialect.escapeIdentifier(object.column)}`);
    const as = object.as === void 0 ? "" : ` as ${dialect.escapeIdentifier(object.as)}`;
    return `${chunks.join(".")}${as}`;
  }
  generateSQL({ dialect }) {
    if (typeof this.value === "string") {
      return {
        sql: `${dialect.escapeIdentifier(this.value)}`
      };
    }
    if (Array.isArray(this.value)) {
      if (this.value.length === 0) {
        throw new Error(
          `you can't specify empty array as parameter for sql.identifier.`
        );
      }
      const chunks = [];
      for (const val of this.value) {
        if (Array.isArray(val)) {
          throw new Error(
            `you can't specify array of arrays as parameter for sql.identifier.`
          );
        }
        if (typeof val === "string") {
          chunks.push(dialect.escapeIdentifier(val));
        } else if (typeof val === "object" && val !== null) {
          chunks.push(this.objectToSQL(val, dialect));
        } else {
          throw new Error(
            `you can't specify array of (null or undefined or number or bigint or boolean or symbol or function) as parameter for sql.identifier.`
          );
        }
      }
      return { sql: chunks.join(", ") };
    }
    if (typeof this.value === "object" && this.value !== null) {
      return { sql: this.objectToSQL(this.value, dialect) };
    }
    if (this.value === null) {
      throw new Error(
        `you can't specify null as parameter for sql.identifier.`
      );
    }
    throw new Error(
      `you can't specify ${typeof this.value} as parameter for sql.identifier.`
    );
  }
}
class SQLValues extends SQLChunk {
  constructor(value, types = []) {
    super();
    this.value = value;
    this.types = types;
  }
  params = [];
  paramsCount = 0;
  generateSQL({ dialect, lastParamIdx }) {
    this.params = dialect.createEmptyParams();
    this.paramsCount = 0;
    if (!Array.isArray(this.value)) {
      if (this.value === null) {
        throw new Error(`you can't specify null as parameter for sql.values.`);
      }
      throw new Error(
        `you can't specify ${typeof this.value} as parameter for sql.values.`
      );
    }
    if (this.value.length === 0) {
      throw new Error(
        `you can't specify empty array as parameter for sql.values.`
      );
    }
    const sql = this.value.map((rowValues) => this.rowValuesToSQL(rowValues, dialect, lastParamIdx)).join(", ");
    return {
      sql,
      params: this.params,
      paramsCount: this.paramsCount
    };
  }
  rowValuesToSQL(rowValues, dialect, lastParamIdx) {
    if (Array.isArray(rowValues)) {
      if (rowValues.length === 0) {
        throw new Error(`array of values can't be empty.`);
      }
      return `(${rowValues.map((value, index) => {
        const sql = dialect.valueToSQL({
          value,
          lastParamIdx,
          params: this.params,
          types: this.types,
          colIdx: index,
          paramsCount: this.paramsCount
        });
        this.paramsCount++;
        return sql;
      }).join(", ")})`;
    }
    if (rowValues === null) {
      throw new Error(
        `you can't specify array of null as parameter for sql.values.`
      );
    }
    throw new Error(
      `you can't specify array of ${typeof rowValues} as parameter for sql.values.`
    );
  }
}
class SQLRaw extends SQLChunk {
  constructor(value) {
    super();
    this.value = value;
  }
  generateSQL() {
    if (typeof this.value === "number" || typeof this.value === "bigint" || typeof this.value === "string" || typeof this.value === "boolean") {
      return { sql: `${this.value}` };
    }
    if (typeof this.value === "object") {
      throw new Error(
        "you can't specify array, object or null as parameter for sql.raw."
      );
    }
    if (this.value === void 0) {
      throw new Error(
        "you can't specify undefined as parameter for sql.raw, maybe you mean using sql.default?"
      );
    }
    if (typeof this.value === "symbol") {
      throw new Error("you can't specify symbol as parameter for sql.raw.");
    }
    throw new Error("you can't specify function as parameter for sql.raw.");
  }
}
class SQLDefault extends SQLChunk {
  generateSQL() {
    return { sql: "default" };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Dialect,
  SQLChunk,
  SQLCommonParam,
  SQLDefault,
  SQLIdentifier,
  SQLQuery,
  SQLRaw,
  SQLString,
  SQLValues
});
