// waddler/sql-template.js
import { NoopLogger } from "./logger.js";
class SQLTemplate {
  constructor(sqlWrapper, dialect, configOptions = {}) {
    this.sqlWrapper = sqlWrapper;
    this.dialect = dialect;
    this.configOptions = configOptions;
    this.logger = configOptions.logger ?? new NoopLogger();
  }
  logger;
  append(other) {
    this.sqlWrapper.append(other.sqlWrapper);
    this.sqlWrapper.recalculateQuery(this.dialect);
  }
  toSQL() {
    return this.sqlWrapper.getQuery(this.dialect);
  }
  // Allow it to be awaited (like a Promise)
  then(onfulfilled, onrejected) {
    const result = this.execute();
    return Promise.resolve(result).then(onfulfilled, onrejected);
  }
  async *chunked(chunkSize = 1) {
    let rows = [];
    let row;
    const asyncIterator = this.stream();
    let iterResult = await asyncIterator.next();
    while (!iterResult.done) {
      row = iterResult.value;
      rows.push(row);
      if (rows.length % chunkSize === 0) {
        yield rows;
        rows = [];
      }
      iterResult = await asyncIterator.next();
    }
    if (rows.length !== 0) {
      yield rows;
    }
  }
}
export {
  SQLTemplate
};
