// waddler/sql.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var sql_exports = {};
__export(sql_exports, {
  SQLWrapper: () => SQLWrapper,
  paramsCheck: () => paramsCheck
});
module.exports = __toCommonJS(sql_exports);
var import_sql_template_params = require("./sql-template-params.cjs");
var import_sql_template = require("./sql-template.cjs");
class SQLWrapper {
  constructor(queryChunks = [], sql, params) {
    this.queryChunks = queryChunks;
    this.sql = sql;
    this.params = params;
  }
  with({ templateParams, rawParams }) {
    if (templateParams) {
      const { strings, params } = templateParams;
      if (strings === void 0) return this;
      if (params.length > 0 || strings.length > 0 && strings[0] !== "") {
        this.queryChunks.push(new import_sql_template_params.SQLString(strings[0]));
      }
      for (const [paramIndex, param] of params.entries()) {
        if (param instanceof import_sql_template_params.SQLQuery || param instanceof import_sql_template.SQLTemplate) {
          this.append(param.sqlWrapper);
          this.queryChunks.push(new import_sql_template_params.SQLString(strings[paramIndex + 1]));
        } else if (param instanceof import_sql_template_params.SQLChunk) this.queryChunks.push(param, new import_sql_template_params.SQLString(strings[paramIndex + 1]));
        else {
          paramsCheck(param);
          this.queryChunks.push(new import_sql_template_params.SQLCommonParam(param), new import_sql_template_params.SQLString(strings[paramIndex + 1]));
        }
      }
    } else if (rawParams) {
      this.sql = rawParams.sql;
      this.params = rawParams.params;
    }
    return this;
  }
  getQuery(dialect) {
    return {
      sql: this.sql ?? "",
      params: this.params ?? dialect.createEmptyParams()
    };
  }
  // This alias is provided to improve clarity when recalculating queries for functions such as "append."
  recalculateQuery(dialect) {
    return this.prepareQuery(dialect);
  }
  prepareQuery(dialect) {
    if (this.queryChunks.length === 1 && this.queryChunks[0] instanceof import_sql_template_params.SQLString) {
      this.sql = this.queryChunks[0].generateSQL().sql;
      this.params = dialect.createEmptyParams();
    }
    const params4driver = dialect.createEmptyParams();
    let paramsCount = 0;
    let query = "";
    for (const chunk of this.queryChunks) {
      if (chunk instanceof import_sql_template_params.SQLString || chunk instanceof import_sql_template_params.SQLRaw || chunk instanceof import_sql_template_params.SQLDefault) {
        query += chunk.generateSQL().sql;
      }
      if (chunk instanceof import_sql_template_params.SQLIdentifier) {
        query += chunk.generateSQL({ dialect }).sql;
      }
      if (chunk instanceof import_sql_template_params.SQLValues || chunk instanceof import_sql_template_params.SQLCommonParam) {
        const { sql, params, paramsCount: newParamsCount } = chunk.generateSQL({ dialect, lastParamIdx: paramsCount });
        query += sql;
        dialect.pushParams(params4driver, params, paramsCount + 1, "bulk");
        paramsCount += newParamsCount;
      }
    }
    this.sql = query;
    this.params = params4driver;
    return this;
  }
  append(other) {
    const thisLastChunk = this.queryChunks.at(-1), otherFirstChunk = other.queryChunks.at(0);
    if (thisLastChunk instanceof import_sql_template_params.SQLString && otherFirstChunk instanceof import_sql_template_params.SQLString) {
      const middleChunk = new import_sql_template_params.SQLString(
        `${thisLastChunk.generateSQL().sql}${otherFirstChunk.generateSQL().sql}`
      );
      this.queryChunks = [
        ...this.queryChunks.slice(0, -1),
        middleChunk,
        ...other.queryChunks.slice(1)
      ];
      return;
    }
    this.queryChunks.push(...other.queryChunks);
  }
}
function paramsCheck(param) {
  if (param === void 0) {
    throw new Error("you can't specify undefined as parameter");
  }
  if (typeof param === "symbol") {
    throw new Error("you can't specify symbol as parameter");
  }
  if (typeof param === "function") {
    throw new Error("you can't specify function as parameter");
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SQLWrapper,
  paramsCheck
});
