// waddler/sql.js
import {
  SQLChunk,
  SQLCommonParam,
  SQLDefault,
  SQLIdentifier,
  SQLQuery,
  SQLRaw,
  SQLString,
  SQLValues
} from "./sql-template-params.js";
import { SQLTemplate } from "./sql-template.js";
class SQLWrapper {
  constructor(queryChunks = [], sql, params) {
    this.queryChunks = queryChunks;
    this.sql = sql;
    this.params = params;
  }
  with({ templateParams, rawParams }) {
    if (templateParams) {
      const { strings, params } = templateParams;
      if (strings === void 0) return this;
      if (params.length > 0 || strings.length > 0 && strings[0] !== "") {
        this.queryChunks.push(new SQLString(strings[0]));
      }
      for (const [paramIndex, param] of params.entries()) {
        if (param instanceof SQLQuery || param instanceof SQLTemplate) {
          this.append(param.sqlWrapper);
          this.queryChunks.push(new SQLString(strings[paramIndex + 1]));
        } else if (param instanceof SQLChunk) this.queryChunks.push(param, new SQLString(strings[paramIndex + 1]));
        else {
          paramsCheck(param);
          this.queryChunks.push(new SQLCommonParam(param), new SQLString(strings[paramIndex + 1]));
        }
      }
    } else if (rawParams) {
      this.sql = rawParams.sql;
      this.params = rawParams.params;
    }
    return this;
  }
  getQuery(dialect) {
    return {
      sql: this.sql ?? "",
      params: this.params ?? dialect.createEmptyParams()
    };
  }
  // This alias is provided to improve clarity when recalculating queries for functions such as "append."
  recalculateQuery(dialect) {
    return this.prepareQuery(dialect);
  }
  prepareQuery(dialect) {
    if (this.queryChunks.length === 1 && this.queryChunks[0] instanceof SQLString) {
      this.sql = this.queryChunks[0].generateSQL().sql;
      this.params = dialect.createEmptyParams();
    }
    const params4driver = dialect.createEmptyParams();
    let paramsCount = 0;
    let query = "";
    for (const chunk of this.queryChunks) {
      if (chunk instanceof SQLString || chunk instanceof SQLRaw || chunk instanceof SQLDefault) {
        query += chunk.generateSQL().sql;
      }
      if (chunk instanceof SQLIdentifier) {
        query += chunk.generateSQL({ dialect }).sql;
      }
      if (chunk instanceof SQLValues || chunk instanceof SQLCommonParam) {
        const { sql, params, paramsCount: newParamsCount } = chunk.generateSQL({ dialect, lastParamIdx: paramsCount });
        query += sql;
        dialect.pushParams(params4driver, params, paramsCount + 1, "bulk");
        paramsCount += newParamsCount;
      }
    }
    this.sql = query;
    this.params = params4driver;
    return this;
  }
  append(other) {
    const thisLastChunk = this.queryChunks.at(-1), otherFirstChunk = other.queryChunks.at(0);
    if (thisLastChunk instanceof SQLString && otherFirstChunk instanceof SQLString) {
      const middleChunk = new SQLString(
        `${thisLastChunk.generateSQL().sql}${otherFirstChunk.generateSQL().sql}`
      );
      this.queryChunks = [
        ...this.queryChunks.slice(0, -1),
        middleChunk,
        ...other.queryChunks.slice(1)
      ];
      return;
    }
    this.queryChunks.push(...other.queryChunks);
  }
}
function paramsCheck(param) {
  if (param === void 0) {
    throw new Error("you can't specify undefined as parameter");
  }
  if (typeof param === "symbol") {
    throw new Error("you can't specify symbol as parameter");
  }
  if (typeof param === "function") {
    throw new Error("you can't specify function as parameter");
  }
}
export {
  SQLWrapper,
  paramsCheck
};
