// @netlify/database/dist/main.js
// src/main.ts
import { neon, neonConfig, Pool as NeonPool } from "@neondatabase/serverless";
import ws from "ws";
import pg from "pg";
import { waddler as waddlerNeonHttp } from "waddler/neon-http";
import { waddler as waddlerNodePostgres } from "waddler/node-postgres";
import { getEnvironment } from "@netlify/runtime-utils";

// src/environment.ts
var MissingDatabaseConnectionError = class extends Error {
  constructor() {
    super(
      "The environment has not been configured to use Netlify Database. You must supply the `connectionString` option when calling `getDatabase()`. See https://ntl.fyi/database-environment for details."
    );
    this.name = "MissingDatabaseConnectionError";
  }
};

// src/main.ts
function getConnectionString() {
  const env = getEnvironment();
  const connectionString = env.get("NETLIFY_DB_URL");
  if (!connectionString) {
    throw new MissingDatabaseConnectionError();
  }
  return connectionString;
}
function getDatabase(options = {}) {
  const env = getEnvironment();
  const connectionString = options.connectionString ?? env.get("NETLIFY_DB_URL");
  if (!connectionString) {
    throw new MissingDatabaseConnectionError();
  }
  const driver = env.get("NETLIFY_DB_DRIVER");
  if (driver === "serverless") {
    if (!neonConfig.webSocketConstructor && typeof WebSocket === "undefined") {
      neonConfig.webSocketConstructor = ws;
    }
    const httpClient = neon(connectionString);
    return {
      driver: "serverless",
      sql: waddlerNeonHttp({ client: httpClient }),
      pool: new NeonPool({ connectionString }),
      httpClient,
      connectionString
    };
  }
  const pool = new pg.Pool({ connectionString });
  return {
    driver: "server",
    sql: waddlerNodePostgres({ client: pool }),
    pool,
    connectionString
  };
}
export {
  MissingDatabaseConnectionError,
  getConnectionString,
  getDatabase
};
