// @netlify/ai/dist/bootstrap/main.js
// src/bootstrap/main.ts
var fetchAIProviders = async ({ api }) => {
  try {
    if (!api.accessToken) {
      return [];
    }
    const data = await api.getAIGatewayProviders();
    if (!data.providers) {
      return [];
    }
    const envVars = [];
    for (const provider of Object.values(data.providers)) {
      if (provider.token_env_var && provider.url_env_var) {
        envVars.push({
          key: provider.token_env_var,
          url: provider.url_env_var
        });
      }
    }
    return envVars;
  } catch (error) {
    console.warn(`Failed to fetch AI providers: ${error instanceof Error ? error.message : String(error)}`);
    return [];
  }
};
var fetchAIGatewayToken = async ({
  api,
  siteId
}) => {
  try {
    if (!api.accessToken) {
      return null;
    }
    const data = await api.getAIGatewayToken({ site_id: siteId });
    if (!data.token || !data.url) {
      return null;
    }
    return {
      token: data.token,
      url: data.url
    };
  } catch (error) {
    console.warn(
      `Failed to fetch AI Gateway token for site ${siteId}: ${error instanceof Error ? error.message : String(error)}`
    );
    return null;
  }
};
var fetchAccountAIGatewayToken = async ({
  api,
  accountId
}) => {
  try {
    if (!api.accessToken) {
      return null;
    }
    const data = await api.getAccountAIGatewayToken({ account_id: accountId });
    if (!data.token || !data.url) {
      return null;
    }
    return {
      token: data.token,
      url: data.url
    };
  } catch (error) {
    console.warn(
      `Failed to fetch AI Gateway token for account ${accountId}: ${error instanceof Error ? error.message : String(error)}`
    );
    return null;
  }
};
var setupAIGateway = async (config) => {
  const { api, env, siteID, siteURL, accountID, siteHasDeploy = true } = config;
  let aiGatewayToken = null;
  let gatewayUrl = null;
  const providersPromise = fetchAIProviders({ api });
  if (siteID && siteID !== "unlinked" && siteURL && siteHasDeploy) {
    aiGatewayToken = await fetchAIGatewayToken({ api, siteId: siteID });
    if (aiGatewayToken) gatewayUrl = `${siteURL}/.netlify/ai`;
  }
  if (!aiGatewayToken && accountID) {
    aiGatewayToken = await fetchAccountAIGatewayToken({ api, accountId: accountID });
    if (aiGatewayToken) gatewayUrl = aiGatewayToken.url;
  }
  if (aiGatewayToken && gatewayUrl) {
    const envVars = await providersPromise;
    const aiGatewayContext = JSON.stringify({
      token: aiGatewayToken.token,
      url: gatewayUrl,
      envVars
    });
    const base64Context = Buffer.from(aiGatewayContext).toString("base64");
    env.AI_GATEWAY = { sources: ["internal"], value: base64Context };
  }
};
var parseAIGatewayContext = (aiGatewayValue) => {
  try {
    if (aiGatewayValue) {
      const decodedContext = Buffer.from(aiGatewayValue, "base64").toString("utf8");
      const aiGatewayContext = JSON.parse(decodedContext);
      return aiGatewayContext;
    }
  } catch {
  }
  return void 0;
};
export {
  fetchAIGatewayToken,
  fetchAIProviders,
  fetchAccountAIGatewayToken,
  parseAIGatewayContext,
  setupAIGateway
};
