// waddler/gel/gel-core/dialect.js
import { Dialect, SQLDefault, SQLIdentifier, SQLRaw, SQLValues } from "../../sql-template-params.js";
class GelDialect extends Dialect {
  escapeParam(lastParamIdx) {
    return `$${lastParamIdx}`;
  }
  escapeIdentifier(identifier) {
    return `"${identifier}"`;
  }
  checkIdentifierObject(object) {
    if (Object.values(object).includes(void 0)) {
      throw new Error(
        `you can't specify undefined parameters. maybe you want to omit it?`
      );
    }
    if (Object.keys(object).length === 0) {
      throw new Error(`you need to specify at least one parameter.`);
    }
    if (object.schema !== void 0 && object.table === void 0 && object.column !== void 0) {
      throw new Error(
        `you can't specify only "schema" and "column" properties, you need also specify "table".`
      );
    }
    if (Object.keys(object).length === 1 && object.as !== void 0) {
      throw new Error(`you can't specify only "as" property.`);
    }
    if (object.as !== void 0 && object.column === void 0 && object.table === void 0) {
      throw new Error(
        `you have to specify "column" or "table" property along with "as".`
      );
    }
    if (!["string", "undefined"].includes(typeof object.schema) || !["string", "undefined"].includes(typeof object.table) || !["string", "undefined"].includes(typeof object.column) || !["string", "undefined"].includes(typeof object.as)) {
      throw new Error(
        "object properties 'schema', 'table', 'column', 'as' should be of string type or omitted."
      );
    }
  }
  // SQLValues
  valueToSQL({ value, lastParamIdx, params }) {
    if (value instanceof SQLDefault) {
      return value.generateSQL().sql;
    }
    if (value === void 0) {
      throw new Error("value can't be undefined, maybe you mean sql.default?");
    }
    params.push(value);
    return this.escapeParam(lastParamIdx + params.length);
  }
}
const SQLFunctions = {
  identifier: (value) => {
    return new SQLIdentifier(value);
  },
  values: (value) => {
    return new SQLValues(value);
  },
  raw: (value) => {
    return new SQLRaw(value);
  },
  default: new SQLDefault()
};
export {
  GelDialect,
  SQLFunctions
};
