// waddler/sqlite/libsql/session.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var session_exports = {};
__export(session_exports, {
  LibsqlSQLTemplate: () => LibsqlSQLTemplate
});
module.exports = __toCommonJS(session_exports);
var import_errors = require("../../errors/index.cjs");
var import_sql_template = require("../../sql-template.cjs");
class LibsqlSQLTemplate extends import_sql_template.SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
  }
  returningData = true;
  all() {
    this.returningData = true;
    return this;
  }
  run() {
    this.returningData = false;
    return this;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      const stmt = { sql: query, args: params };
      if (this.returningData) {
        if (this.options.rowMode === "array") {
          const queryResult = await this.client.execute(stmt);
          finalRes = queryResult.rows.map((row) => queryResult.columns.map((col) => row[col]));
          finalMetadata = {
            columnTypes: queryResult.columnTypes,
            columns: queryResult.columns,
            lastInsertRowid: queryResult.lastInsertRowid,
            rowsAffected: queryResult.rowsAffected
          };
        } else {
          const queryResult = await this.client.execute(stmt);
          finalRes = queryResult.rows;
          finalMetadata = {
            columnTypes: queryResult.columnTypes,
            columns: queryResult.columns,
            lastInsertRowid: queryResult.lastInsertRowid,
            rowsAffected: queryResult.rowsAffected
          };
        }
      } else {
        const queryResult = await this.client.execute(stmt);
        finalMetadata = {
          columnTypes: queryResult.columnTypes,
          columns: queryResult.columns,
          lastInsertRowid: queryResult.lastInsertRowid,
          rowsAffected: queryResult.rowsAffected
        };
        finalRes = queryResult;
      }
    } catch (error) {
      throw new import_errors.WaddlerQueryError(query, params, error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  /**
   * For now, throws the Error.
   * Current implementation (a placeholder) will be replaced once cloudflare d1 driver acquires streaming support or when a suitable third-party solution is found.
   */
  // eslint-disable-next-line require-yield
  async *stream() {
    throw new Error("stream is not implemented for d1 yet.");
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LibsqlSQLTemplate
});
