// waddler/sqlite/libsql/session.js
import { WaddlerQueryError } from "../../errors/index.js";
import { SQLTemplate } from "../../sql-template.js";
class LibsqlSQLTemplate extends SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
  }
  returningData = true;
  all() {
    this.returningData = true;
    return this;
  }
  run() {
    this.returningData = false;
    return this;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      const stmt = { sql: query, args: params };
      if (this.returningData) {
        if (this.options.rowMode === "array") {
          const queryResult = await this.client.execute(stmt);
          finalRes = queryResult.rows.map((row) => queryResult.columns.map((col) => row[col]));
          finalMetadata = {
            columnTypes: queryResult.columnTypes,
            columns: queryResult.columns,
            lastInsertRowid: queryResult.lastInsertRowid,
            rowsAffected: queryResult.rowsAffected
          };
        } else {
          const queryResult = await this.client.execute(stmt);
          finalRes = queryResult.rows;
          finalMetadata = {
            columnTypes: queryResult.columnTypes,
            columns: queryResult.columns,
            lastInsertRowid: queryResult.lastInsertRowid,
            rowsAffected: queryResult.rowsAffected
          };
        }
      } else {
        const queryResult = await this.client.execute(stmt);
        finalMetadata = {
          columnTypes: queryResult.columnTypes,
          columns: queryResult.columns,
          lastInsertRowid: queryResult.lastInsertRowid,
          rowsAffected: queryResult.rowsAffected
        };
        finalRes = queryResult;
      }
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  /**
   * For now, throws the Error.
   * Current implementation (a placeholder) will be replaced once cloudflare d1 driver acquires streaming support or when a suitable third-party solution is found.
   */
  // eslint-disable-next-line require-yield
  async *stream() {
    throw new Error("stream is not implemented for d1 yet.");
  }
}
export {
  LibsqlSQLTemplate
};
