// drizzle-orm/zod/schema.js
import { isWithEnum } from "../utils.js";
import { columnToSchema } from "./column.js";
import { is } from "../entity.js";
import { isTable } from "../table.js";
import { Column } from "../column.js";
import { getColumns } from "../utils.js";
import { SQL, isView } from "../sql/sql.js";
import { z } from "zod/v4";

//#region src/zod/schema.ts
function handleColumns(columns, refinements, conditions, factory) {
	const columnSchemas = {};
	for (const [key, selected] of Object.entries(columns)) {
		if (!is(selected, Column) && !is(selected, SQL) && !is(selected, SQL.Aliased) && typeof selected === "object") {
			columnSchemas[key] = handleColumns(isTable(selected) || isView(selected) ? getColumns(selected) : selected, refinements[key] ?? {}, conditions, factory);
			continue;
		}
		const refinement = refinements[key];
		if (refinement !== void 0 && typeof refinement !== "function") {
			columnSchemas[key] = refinement;
			continue;
		}
		const column = is(selected, Column) ? selected : void 0;
		const schema = column ? columnToSchema(column, factory) : z.any();
		const refined = typeof refinement === "function" ? refinement(schema) : schema;
		if (conditions.never(column)) continue;
		else columnSchemas[key] = refined;
		if (column) {
			if (conditions.nullable(column)) columnSchemas[key] = columnSchemas[key].nullable();
			if (conditions.optional(column)) columnSchemas[key] = columnSchemas[key].optional();
		}
	}
	return z.object(columnSchemas);
}
function handleEnum(enum_, factory) {
	return (factory?.zodInstance ?? z).enum(enum_.enumValues);
}
const selectConditions = {
	never: () => false,
	optional: () => false,
	nullable: (column) => !column.notNull
};
const insertConditions = {
	never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
	optional: (column) => !column.notNull || column.notNull && column.hasDefault,
	nullable: (column) => !column.notNull
};
const updateConditions = {
	never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
	optional: () => true,
	nullable: (column) => !column.notNull
};
const createSelectSchema = (entity, refine) => {
	if (isWithEnum(entity)) return handleEnum(entity);
	return handleColumns(getColumns(entity), refine ?? {}, selectConditions);
};
const createInsertSchema = (entity, refine) => {
	return handleColumns(getColumns(entity), refine ?? {}, insertConditions);
};
const createUpdateSchema = (entity, refine) => {
	return handleColumns(getColumns(entity), refine ?? {}, updateConditions);
};
function createSchemaFactory(options) {
	const createSelectSchema = (entity, refine) => {
		if (isWithEnum(entity)) return handleEnum(entity, options);
		return handleColumns(getColumns(entity), refine ?? {}, selectConditions, options);
	};
	const createInsertSchema = (entity, refine) => {
		return handleColumns(getColumns(entity), refine ?? {}, insertConditions, options);
	};
	const createUpdateSchema = (entity, refine) => {
		return handleColumns(getColumns(entity), refine ?? {}, updateConditions, options);
	};
	return {
		createSelectSchema,
		createInsertSchema,
		createUpdateSchema
	};
}

//#endregion
export { createInsertSchema, createSchemaFactory, createSelectSchema, createUpdateSchema };
