// drizzle-orm/zod/schema.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
const require_utils = require('../utils.cjs');
const require_zod_column = require('./column.cjs');
let __entity_ts = require("../entity.cjs");
let __table_ts = require("../table.cjs");
let __column_ts = require("../column.cjs");
let __utils_ts = require("../utils.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");
let zod_v4 = require("zod/v4");

//#region src/zod/schema.ts
function handleColumns(columns, refinements, conditions, factory) {
	const columnSchemas = {};
	for (const [key, selected] of Object.entries(columns)) {
		if (!(0, __entity_ts.is)(selected, __column_ts.Column) && !(0, __entity_ts.is)(selected, __sql_sql_ts.SQL) && !(0, __entity_ts.is)(selected, __sql_sql_ts.SQL.Aliased) && typeof selected === "object") {
			columnSchemas[key] = handleColumns((0, __table_ts.isTable)(selected) || (0, __sql_sql_ts.isView)(selected) ? (0, __utils_ts.getColumns)(selected) : selected, refinements[key] ?? {}, conditions, factory);
			continue;
		}
		const refinement = refinements[key];
		if (refinement !== void 0 && typeof refinement !== "function") {
			columnSchemas[key] = refinement;
			continue;
		}
		const column = (0, __entity_ts.is)(selected, __column_ts.Column) ? selected : void 0;
		const schema = column ? require_zod_column.columnToSchema(column, factory) : zod_v4.z.any();
		const refined = typeof refinement === "function" ? refinement(schema) : schema;
		if (conditions.never(column)) continue;
		else columnSchemas[key] = refined;
		if (column) {
			if (conditions.nullable(column)) columnSchemas[key] = columnSchemas[key].nullable();
			if (conditions.optional(column)) columnSchemas[key] = columnSchemas[key].optional();
		}
	}
	return zod_v4.z.object(columnSchemas);
}
function handleEnum(enum_, factory) {
	return (factory?.zodInstance ?? zod_v4.z).enum(enum_.enumValues);
}
const selectConditions = {
	never: () => false,
	optional: () => false,
	nullable: (column) => !column.notNull
};
const insertConditions = {
	never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
	optional: (column) => !column.notNull || column.notNull && column.hasDefault,
	nullable: (column) => !column.notNull
};
const updateConditions = {
	never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
	optional: () => true,
	nullable: (column) => !column.notNull
};
const createSelectSchema = (entity, refine) => {
	if (require_utils.isWithEnum(entity)) return handleEnum(entity);
	return handleColumns((0, __utils_ts.getColumns)(entity), refine ?? {}, selectConditions);
};
const createInsertSchema = (entity, refine) => {
	return handleColumns((0, __utils_ts.getColumns)(entity), refine ?? {}, insertConditions);
};
const createUpdateSchema = (entity, refine) => {
	return handleColumns((0, __utils_ts.getColumns)(entity), refine ?? {}, updateConditions);
};
function createSchemaFactory(options) {
	const createSelectSchema = (entity, refine) => {
		if (require_utils.isWithEnum(entity)) return handleEnum(entity, options);
		return handleColumns((0, __utils_ts.getColumns)(entity), refine ?? {}, selectConditions, options);
	};
	const createInsertSchema = (entity, refine) => {
		return handleColumns((0, __utils_ts.getColumns)(entity), refine ?? {}, insertConditions, options);
	};
	const createUpdateSchema = (entity, refine) => {
		return handleColumns((0, __utils_ts.getColumns)(entity), refine ?? {}, updateConditions, options);
	};
	return {
		createSelectSchema,
		createInsertSchema,
		createUpdateSchema
	};
}

//#endregion
exports.createInsertSchema = createInsertSchema;
exports.createSchemaFactory = createSchemaFactory;
exports.createSelectSchema = createSelectSchema;
exports.createUpdateSchema = createUpdateSchema;
