// waddler/duckdb-neo/driver.js
import { DuckDBInstance } from "@duckdb/node-api";
import { DuckdbDialect, SQLFunctions } from "../duckdb-core/dialect.js";
import { DefaultLogger } from "../logger.js";
import { RecyclingPool } from "../recycling-pool.js";
import { SQLDefault, SQLIdentifier, SQLQuery, SQLRaw, SQLValues } from "../sql-template-params.js";
import { SQLWrapper } from "../sql.js";
import { DuckdbNeoSQLTemplate } from "./session.js";
const sql = (strings, ...params) => {
  const sqlWrapper = new SQLWrapper();
  sqlWrapper.with({ templateParams: { strings, params } });
  const dialect = new DuckdbDialect();
  return new SQLQuery(sqlWrapper, dialect);
};
Object.assign(sql, SQLFunctions);
const createSqlTemplate = (pool, configOptions) => {
  const dialect = new DuckdbDialect();
  let logger;
  if (configOptions.logger === true) {
    logger = new DefaultLogger();
  } else if (configOptions.logger !== false) {
    logger = configOptions.logger;
  }
  const fn = (strings, ...params) => {
    const sql2 = new SQLWrapper();
    sql2.with({ templateParams: { strings, params } }).prepareQuery(dialect);
    return new DuckdbNeoSQLTemplate(sql2, pool, dialect, { logger });
  };
  Object.assign(fn, {
    identifier: (value) => {
      return new SQLIdentifier(value);
    },
    values: (value) => {
      return new SQLValues(value);
    },
    raw: (value) => {
      return new SQLRaw(value);
    },
    unsafe: async (query, params, options) => {
      params = params ?? [];
      options = options ?? { rowMode: "default" };
      const sqlWrapper = new SQLWrapper();
      sqlWrapper.with({ rawParams: { sql: query, params } });
      const unsafeDriver = new DuckdbNeoSQLTemplate(sqlWrapper, pool, dialect, { logger }, options);
      return await unsafeDriver.execute();
    },
    default: new SQLDefault()
  });
  return fn;
};
const createFactory = ({
  url,
  accessMode = "read_write",
  maxMemory = "512MB",
  threads = "4"
}) => {
  const factory = {
    create: async function() {
      try {
        const instance = await DuckDBInstance.create(url, {
          access_mode: accessMode,
          max_memory: maxMemory,
          threads
        });
        const conn = await instance.connect();
        const connObj = { instance, connection: conn };
        return connObj;
      } catch (error) {
        const newError = new Error(error.message);
        throw newError;
      }
    },
    destroy: async function() {
    }
  };
  return factory;
};
function waddler({
  dbUrl,
  url,
  min = 1,
  max = 1,
  accessMode = "read_write",
  maxMemory = "512MB",
  threads = "4",
  logger
}) {
  url = url === void 0 && dbUrl !== void 0 ? dbUrl : url;
  const factory = createFactory({
    url,
    accessMode,
    maxMemory,
    threads
  });
  const options = {
    max,
    // maximum size of the pool
    min
    // minimum size of the pool
  };
  const pool = new RecyclingPool(factory, options);
  return createSqlTemplate(pool, { logger });
}
export {
  sql,
  waddler
};
