// waddler/duckdb-neo/driver.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var driver_exports = {};
__export(driver_exports, {
  sql: () => sql,
  waddler: () => waddler
});
module.exports = __toCommonJS(driver_exports);
var import_node_api = require("@duckdb/node-api");
var import_dialect = require("../duckdb-core/dialect.cjs");
var import_logger = require("../logger.cjs");
var import_recycling_pool = require("../recycling-pool.cjs");
var import_sql_template_params = require("../sql-template-params.cjs");
var import_sql = require("../sql.cjs");
var import_session = require("./session.cjs");
const sql = (strings, ...params) => {
  const sqlWrapper = new import_sql.SQLWrapper();
  sqlWrapper.with({ templateParams: { strings, params } });
  const dialect = new import_dialect.DuckdbDialect();
  return new import_sql_template_params.SQLQuery(sqlWrapper, dialect);
};
Object.assign(sql, import_dialect.SQLFunctions);
const createSqlTemplate = (pool, configOptions) => {
  const dialect = new import_dialect.DuckdbDialect();
  let logger;
  if (configOptions.logger === true) {
    logger = new import_logger.DefaultLogger();
  } else if (configOptions.logger !== false) {
    logger = configOptions.logger;
  }
  const fn = (strings, ...params) => {
    const sql2 = new import_sql.SQLWrapper();
    sql2.with({ templateParams: { strings, params } }).prepareQuery(dialect);
    return new import_session.DuckdbNeoSQLTemplate(sql2, pool, dialect, { logger });
  };
  Object.assign(fn, {
    identifier: (value) => {
      return new import_sql_template_params.SQLIdentifier(value);
    },
    values: (value) => {
      return new import_sql_template_params.SQLValues(value);
    },
    raw: (value) => {
      return new import_sql_template_params.SQLRaw(value);
    },
    unsafe: async (query, params, options) => {
      params = params ?? [];
      options = options ?? { rowMode: "default" };
      const sqlWrapper = new import_sql.SQLWrapper();
      sqlWrapper.with({ rawParams: { sql: query, params } });
      const unsafeDriver = new import_session.DuckdbNeoSQLTemplate(sqlWrapper, pool, dialect, { logger }, options);
      return await unsafeDriver.execute();
    },
    default: new import_sql_template_params.SQLDefault()
  });
  return fn;
};
const createFactory = ({
  url,
  accessMode = "read_write",
  maxMemory = "512MB",
  threads = "4"
}) => {
  const factory = {
    create: async function() {
      try {
        const instance = await import_node_api.DuckDBInstance.create(url, {
          access_mode: accessMode,
          max_memory: maxMemory,
          threads
        });
        const conn = await instance.connect();
        const connObj = { instance, connection: conn };
        return connObj;
      } catch (error) {
        const newError = new Error(error.message);
        throw newError;
      }
    },
    destroy: async function() {
    }
  };
  return factory;
};
function waddler({
  dbUrl,
  url,
  min = 1,
  max = 1,
  accessMode = "read_write",
  maxMemory = "512MB",
  threads = "4",
  logger
}) {
  url = url === void 0 && dbUrl !== void 0 ? dbUrl : url;
  const factory = createFactory({
    url,
    accessMode,
    maxMemory,
    threads
  });
  const options = {
    max,
    // maximum size of the pool
    min
    // minimum size of the pool
  };
  const pool = new import_recycling_pool.RecyclingPool(factory, options);
  return createSqlTemplate(pool, { logger });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  sql,
  waddler
});
