// waddler/duckdb-neo/session.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var session_exports = {};
__export(session_exports, {
  DuckdbNeoSQLTemplate: () => DuckdbNeoSQLTemplate
});
module.exports = __toCommonJS(session_exports);
var import_errors = require("../errors/index.cjs");
var import_sql_template = require("../sql-template.cjs");
var import_result_transformers = require("./result-transformers.cjs");
var import_utils = require("./utils.cjs");
class DuckdbNeoSQLTemplate extends import_sql_template.SQLTemplate {
  constructor(sql, pool, dialect, configOptions, options = { rowMode: "default" }) {
    super(sql, dialect, configOptions);
    this.pool = pool;
    this.options = options;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes;
    let finalMetadata;
    const connObj = await this.pool.acquire();
    try {
      const prepared = await connObj.connection.prepare(query);
      (0, import_utils.bindParams)(prepared, params);
      const duckDbResult = await prepared.run().finally();
      finalMetadata = {
        columnCount: duckDbResult.columnCount,
        returnType: duckDbResult.returnType,
        rowsChanged: duckDbResult.rowsChanged,
        statementType: duckDbResult.statementType
      };
      finalRes = this.options.rowMode === "default" ? await (0, import_result_transformers.transformResultToObjects)(duckDbResult) : await (0, import_result_transformers.transformResultToArrays)(duckDbResult);
    } catch (error) {
      await this.pool.release(connObj);
      throw new import_errors.WaddlerQueryError(query, params, error);
    }
    await this.pool.release(connObj);
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  async *stream() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    const connObj = await this.pool.acquire();
    try {
      const prepared = await connObj.connection.prepare(query);
      (0, import_utils.bindParams)(prepared, params);
      const duckDbResult = await prepared.run();
      for (; ; ) {
        const chunk = await duckDbResult.fetchChunk();
        if (chunk.rowCount === 0) {
          break;
        }
        const columnVectors = (0, import_result_transformers.getColumnVectors)(chunk);
        for (let rowIndex = 0; rowIndex < chunk.rowCount; rowIndex++) {
          const row = (0, import_result_transformers.transformResultRowToObject)(duckDbResult, columnVectors, rowIndex);
          yield row;
        }
      }
    } catch (error) {
      await this.pool.release(connObj);
      throw new import_errors.WaddlerQueryError(query, params, error);
    }
    await this.pool.release(connObj);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DuckdbNeoSQLTemplate
});
