// waddler/duckdb-neo/session.js
import { WaddlerQueryError } from "../errors/index.js";
import { SQLTemplate } from "../sql-template.js";
import {
  getColumnVectors,
  transformResultRowToObject,
  transformResultToArrays,
  transformResultToObjects
} from "./result-transformers.js";
import { bindParams } from "./utils.js";
class DuckdbNeoSQLTemplate extends SQLTemplate {
  constructor(sql, pool, dialect, configOptions, options = { rowMode: "default" }) {
    super(sql, dialect, configOptions);
    this.pool = pool;
    this.options = options;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes;
    let finalMetadata;
    const connObj = await this.pool.acquire();
    try {
      const prepared = await connObj.connection.prepare(query);
      bindParams(prepared, params);
      const duckDbResult = await prepared.run().finally();
      finalMetadata = {
        columnCount: duckDbResult.columnCount,
        returnType: duckDbResult.returnType,
        rowsChanged: duckDbResult.rowsChanged,
        statementType: duckDbResult.statementType
      };
      finalRes = this.options.rowMode === "default" ? await transformResultToObjects(duckDbResult) : await transformResultToArrays(duckDbResult);
    } catch (error) {
      await this.pool.release(connObj);
      throw new WaddlerQueryError(query, params, error);
    }
    await this.pool.release(connObj);
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  async *stream() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    const connObj = await this.pool.acquire();
    try {
      const prepared = await connObj.connection.prepare(query);
      bindParams(prepared, params);
      const duckDbResult = await prepared.run();
      for (; ; ) {
        const chunk = await duckDbResult.fetchChunk();
        if (chunk.rowCount === 0) {
          break;
        }
        const columnVectors = getColumnVectors(chunk);
        for (let rowIndex = 0; rowIndex < chunk.rowCount; rowIndex++) {
          const row = transformResultRowToObject(duckDbResult, columnVectors, rowIndex);
          yield row;
        }
      }
    } catch (error) {
      await this.pool.release(connObj);
      throw new WaddlerQueryError(query, params, error);
    }
    await this.pool.release(connObj);
  }
}
export {
  DuckdbNeoSQLTemplate
};
