// drizzle-orm/vercel-postgres/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");
let __logger_ts = require("../logger.cjs");
let __pg_core_async_session_ts = require("../pg-core/async/session.cjs");
let __query_name_generator_ts = require("../query-name-generator.cjs");
let __cache_core_cache_ts = require("../cache/core/cache.cjs");
let _vercel_postgres = require("@vercel/postgres");

//#region src/vercel-postgres/session.ts
const noop = (val) => val;
const typeConfig = { getTypeParser: ((typeId, format) => {
	switch (typeId) {
		case _vercel_postgres.types.builtins.TIMESTAMPTZ:
		case _vercel_postgres.types.builtins.TIMESTAMP:
		case _vercel_postgres.types.builtins.DATE:
		case _vercel_postgres.types.builtins.INTERVAL:
		case 1231:
		case 1115:
		case 1185:
		case 1187:
		case 1182: return noop;
		default: return _vercel_postgres.types.getTypeParser(typeId, format);
	}
}) };
var VercelPgSession = class VercelPgSession extends __pg_core_async_session_ts.PgAsyncSession {
	static [__entity_ts.entityKind] = "VercelPgSession";
	logger;
	cache;
	constructor(client, dialect, relations, options = {}) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
		this.logger = options.logger ?? new __logger_ts.NoopLogger();
		this.cache = options.cache ?? new __cache_core_cache_ts.NoopCache();
	}
	prepareQuery(query, mode, name, mapper, queryMetadata, cacheConfig) {
		const queryName = typeof name === "string" ? name : name === true ? (0, __query_name_generator_ts.preparedStatementName)(query.sql, query.params) : void 0;
		const executor = async (params) => {
			return this.client.query({
				name: queryName,
				rowMode: mode === "arrays" ? "array" : void 0,
				text: query.sql,
				types: typeConfig
			}, params).then((r) => mode === "raw" ? r : r.rows);
		};
		return new __pg_core_async_session_ts.PgAsyncPreparedQuery(executor, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	async transaction(transaction, config) {
		const session = typeof this.client === "function" || this.client instanceof _vercel_postgres.VercelPool ? new VercelPgSession(await this.client.connect(), this.dialect, this.relations, this.options) : this;
		const tx = new VercelPgTransaction(this.dialect, session, this.relations, void 0, false);
		await tx.execute(__sql_sql_ts.sql`begin${config ? __sql_sql_ts.sql` ${tx.getTransactionConfigSQL(config)}` : void 0}`);
		try {
			const result = await transaction(tx);
			await tx.execute(__sql_sql_ts.sql`commit`);
			return result;
		} catch (error) {
			await tx.execute(__sql_sql_ts.sql`rollback`);
			throw error;
		} finally {
			if (typeof this.client === "function" || this.client instanceof _vercel_postgres.VercelPool) session.client.release();
		}
	}
};
var VercelPgTransaction = class VercelPgTransaction extends __pg_core_async_session_ts.PgAsyncTransaction {
	static [__entity_ts.entityKind] = "VercelPgTransaction";
	async transaction(transaction) {
		const savepointName = `sp${this.nestedIndex + 1}`;
		const tx = new VercelPgTransaction(this.dialect, this.session, this._.relations, this.nestedIndex + 1, false);
		await tx.execute(__sql_sql_ts.sql.raw(`savepoint ${savepointName}`));
		try {
			const result = await transaction(tx);
			await tx.execute(__sql_sql_ts.sql.raw(`release savepoint ${savepointName}`));
			return result;
		} catch (err) {
			await tx.execute(__sql_sql_ts.sql.raw(`rollback to savepoint ${savepointName}`));
			throw err;
		}
	}
};

//#endregion
exports.VercelPgSession = VercelPgSession;
exports.VercelPgTransaction = VercelPgTransaction;
