// waddler/pg/vercel-postgres/session.js
import { types, VercelPool } from "@vercel/postgres";
import { WaddlerQueryError } from "../../errors/index.js";
import { SQLTemplate } from "../../sql-template.js";
const pgTypeConfig = {
  // @ts-expect-error
  getTypeParser: (typeId, format) => {
    if (typeId === types.builtins.INTERVAL) return (val) => val;
    if (typeId === 1187) return (val) => val;
    return types.getTypeParser(typeId, format);
  }
};
class VercelPgSQLTemplate extends SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
    const query = this.sqlWrapper.getQuery(this.dialect).sql;
    this.queryConfig = {
      rowMode: "array",
      text: query,
      types: pgTypeConfig
    };
    this.rawQueryConfig = {
      text: query,
      types: pgTypeConfig
    };
  }
  rawQueryConfig;
  queryConfig;
  async execute() {
    const { params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      const queryResult = await (this.options.rowMode === "array" ? this.client.query(this.queryConfig, params) : this.client.query(this.rawQueryConfig, params));
      ({ rows: finalRes, ...finalMetadata } = queryResult);
    } catch (error) {
      throw new WaddlerQueryError(this.queryConfig.text, params, error);
    }
    this.logger.logQuery(this.queryConfig.text, params, finalMetadata);
    return finalRes;
  }
  // TODO: revise: maybe I should override chunked method because we can use QueryStream with option 'batchSize' in QueryStreamConfig
  async *stream() {
    let conn;
    let query = "", params = [];
    try {
      conn = this.client instanceof VercelPool ? await this.client.connect() : this.client;
      const queryStreamObj = this.configOptions?.extensions?.find((it) => it.name === "WaddlerPgQueryStream");
      if (!queryStreamObj) {
        throw new Error(
          'To use stream feature, you would need to provide queryStream() function to waddler extensions, example: waddler("", { extensions: [queryStream()] })'
        );
      }
      ({ sql: query, params } = this.sqlWrapper.getQuery(this.dialect));
      this.logger.logQuery(this.queryConfig.text, params);
      const queryStream = new queryStreamObj.constructor(query, params, { types: this.queryConfig.types });
      const stream = conn.query(queryStream);
      for await (const row of stream) {
        yield row;
      }
      if (this.client instanceof VercelPool) {
        conn.release();
      }
    } catch (error) {
      if (this.client instanceof VercelPool) {
        conn?.release();
      }
      throw new WaddlerQueryError(query, params, error);
    }
  }
}
export {
  VercelPgSQLTemplate
};
