// drizzle-orm/netlify-db/session.js
import { entityKind } from "../entity.js";
import { sql } from "../sql/sql.js";
import { NoopLogger } from "../logger.js";
import { PgDialect } from "../pg-core/dialect.js";
import { PgAsyncPreparedQuery, PgAsyncSession, PgAsyncTransaction } from "../pg-core/async/session.js";
import { neonConfig, types } from "@neondatabase/serverless";
import { preparedStatementName } from "../query-name-generator.js";
import { NoopCache } from "../cache/core/cache.js";

//#region src/netlify-db/session.ts
/**
* Ensures a WebSocket implementation is available for Neon's
* serverless driver. The Neon driver checks for a global WebSocket
* automatically, but this sets it on neonConfig explicitly so the
* check only happens once.
*/
function ensureWebSocket() {
	if (neonConfig.webSocketConstructor) return;
	if (typeof WebSocket !== "undefined") neonConfig.webSocketConstructor = WebSocket;
}
var NetlifyDbSession = class extends PgAsyncSession {
	static [entityKind] = "NetlifyDbSession";
	clientQuery;
	logger;
	cache;
	constructor(httpClient, pool, dialect, relations, options) {
		super(dialect);
		this.httpClient = httpClient;
		this.pool = pool;
		this.relations = relations;
		this.options = options;
		this.clientQuery = httpClient.query ?? httpClient;
		this.logger = options.logger ?? new NoopLogger();
		this.cache = options.cache ?? new NoopCache();
	}
	prepareQuery(query, mode, _name, mapper, queryMetadata, cacheConfig) {
		const executor = (params) => {
			if (mode === "raw") return (async () => this.httpClient(query.sql, params, {
				arrayMode: false,
				fullResults: true,
				authToken: this.options.authToken
			}))();
			return this.httpClient(query.sql, params, {
				arrayMode: mode === "arrays",
				fullResults: true,
				authToken: this.options.authToken
			}).then((it) => it.rows);
		};
		return new PgAsyncPreparedQuery(executor, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	async batch(queries) {
		const preparedQueries = [];
		const builtQueries = [];
		const q = this.httpClient;
		for (const query of queries) {
			const preparedQuery = query._prepare();
			const builtQuery = preparedQuery.getQuery();
			preparedQueries.push(preparedQuery);
			builtQueries.push(q(builtQuery.sql, builtQuery.params, {
				fullResults: true,
				arrayMode: preparedQuery.mode === "arrays"
			}));
		}
		return (await this.httpClient.transaction(builtQueries, {
			authToken: this.options.authToken,
			fullResults: true,
			arrayMode: true
		})).map((result, i) => preparedQueries[i].mapper ? preparedQueries[i].mapper(result.rows) : result);
	}
	async query(query, params) {
		this.logger.logQuery(query, params);
		return await this.clientQuery(query, params, {
			arrayMode: true,
			fullResults: true
		});
	}
	async queryObjects(query, params) {
		return this.clientQuery(query, params, {
			arrayMode: false,
			fullResults: true
		});
	}
	async transaction(transaction, config = {}) {
		ensureWebSocket();
		const poolClient = await this.pool.connect();
		const dialect = new PgDialect({
			useJitMappers: this.options.useJitMappers,
			codecs: this.options.transactionCodecs
		});
		const tx = new NetlifyDbTransaction(dialect, new NetlifyDbWsSession(poolClient, dialect, this.relations, this.options), this.relations, void 0, false);
		await tx.execute(sql`begin ${tx.getTransactionConfigSQL(config)}`);
		try {
			const result = await transaction(tx);
			await tx.execute(sql`commit`);
			return result;
		} catch (error) {
			await tx.execute(sql`rollback`);
			throw error;
		} finally {
			poolClient.release();
		}
	}
};
const noop = (val) => val;
const typeConfig = { getTypeParser: ((typeId, format) => {
	switch (typeId) {
		case types.builtins.TIMESTAMPTZ:
		case types.builtins.TIMESTAMP:
		case types.builtins.DATE:
		case types.builtins.INTERVAL:
		case 1231:
		case 1115:
		case 1185:
		case 1187:
		case 1182: return noop;
		default: return types.getTypeParser(typeId, format);
	}
}) };
/**
* Internal WebSocket-based session used only within transactions.
* Delegates all queries to a PoolClient over WebSocket, using
* NeonPreparedQuery from the neon-serverless adapter.
*/
var NetlifyDbWsSession = class extends PgAsyncSession {
	static [entityKind] = "NetlifyDbWsSession";
	logger;
	cache;
	constructor(client, dialect, relations, options = {}) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.logger = options.logger ?? new NoopLogger();
		this.cache = options.cache ?? new NoopCache();
	}
	prepareQuery(query, mode, name, mapper, queryMetadata, cacheConfig) {
		const queryName = typeof name === "string" ? name : name === true ? preparedStatementName(query.sql, query.params) : void 0;
		const executor = async (params) => {
			return this.client.query({
				name: queryName,
				rowMode: mode === "arrays" ? "array" : void 0,
				text: query.sql,
				types: typeConfig
			}, params).then((r) => mode === "raw" ? r : r.rows);
		};
		return new PgAsyncPreparedQuery(executor, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	async transaction(_transaction, _config) {
		throw new Error("Nested transactions are handled by NetlifyDbTransaction via savepoints");
	}
};
var NetlifyDbTransaction = class NetlifyDbTransaction extends PgAsyncTransaction {
	static [entityKind] = "NetlifyDbTransaction";
	async transaction(transaction) {
		const savepointName = `sp${this.nestedIndex + 1}`;
		const tx = new NetlifyDbTransaction(this.dialect, this.session, this._.relations, this.nestedIndex + 1, false);
		await tx.execute(sql.raw(`savepoint ${savepointName}`));
		try {
			const result = await transaction(tx);
			await tx.execute(sql.raw(`release savepoint ${savepointName}`));
			return result;
		} catch (e) {
			await tx.execute(sql.raw(`rollback to savepoint ${savepointName}`));
			throw e;
		}
	}
};

//#endregion
export { NetlifyDbSession, NetlifyDbTransaction, NetlifyDbWsSession };
