// drizzle-orm/cache/core/cache-effect.js
import { NoopCache } from "./cache.js";
import { entityKind } from "../../entity.js";
import * as Effect from "effect/Effect";
import * as Layer from "effect/Layer";
import * as Context from "effect/Context";
import * as Schema from "effect/Schema";

//#region src/cache/core/cache-effect.ts
/**
* Effect service for caching query results in Drizzle ORM.
*
* By default, this service uses a no-op cache (no caching occurs). Use
* `EffectCache.fromDrizzle` to adapt a standard Drizzle cache implementation.
*
* @example
* ```ts
* // Use default (no caching)
* const db = yield* PgDrizzle.make({ relations }).pipe(
*   Effect.provide(PgDrizzle.DefaultServices),
* );
*
* // Use a custom Drizzle cache
* const db = yield* PgDrizzle.make({ relations }).pipe(
*   Effect.provide(EffectCache.layerFromDrizzle(myCache)),
*   Effect.provide(PgDrizzle.DefaultServices),
* );
* ```
*/
var EffectCache = class EffectCache extends Context.Service()("drizzle-orm/EffectCache", { make: Effect.sync(() => make(new NoopCache())) }) {
	static [entityKind] = "drizzle-orm/EffectCache";
	/**
	* The default layer providing a no-op cache.
	*/
	static Default = Layer.effect(EffectCache, EffectCache.make);
	/**
	* Creates an EffectCacheShape from a standard Drizzle cache.
	*
	* @param cache - A Drizzle cache instance implementing the `Cache` interface.
	* @returns A new EffectCacheShape that delegates to the provided Drizzle cache.
	*
	* @example
	* ```ts
	* const drizzleCache = new MyCustomCache();
	* const effectCache = EffectCache.fromDrizzle(drizzleCache);
	* ```
	*/
	static fromDrizzle(cache) {
		return make(cache);
	}
	/**
	* Creates a Layer that provides an EffectCache from a standard Drizzle cache.
	*
	* @param cache - A Drizzle cache instance implementing the `Cache` interface.
	* @returns A Layer that provides the EffectCache service.
	*
	* @example
	* ```ts
	* const drizzleCache = new MyCustomCache();
	* const db = yield* PgDrizzle.make({ relations }).pipe(
	*   Effect.provide(EffectCache.layerFromDrizzle(drizzleCache)),
	*   Effect.provide(PgDrizzle.DefaultServices),
	* );
	* ```
	*/
	static layerFromDrizzle(cache) {
		return Layer.succeed(EffectCache, EffectCache.fromDrizzle(cache));
	}
};
function make(cache) {
	const strategy = () => cache.strategy();
	const get = (...args) => Effect.tryPromise({
		try: () => cache.get(...args),
		catch: (error) => new EffectCacheError({ cause: error })
	});
	const put = (...args) => Effect.tryPromise({
		try: () => cache.put(...args),
		catch: (error) => new EffectCacheError({ cause: error })
	});
	const onMutate = (params) => Effect.tryPromise({
		try: () => cache.onMutate(params),
		catch: (error) => new EffectCacheError({ cause: error })
	});
	return {
		strategy,
		get,
		put,
		onMutate,
		cache
	};
}
/**
* Error type for cache operation failures.
*
* This error is thrown when any cache operation (get, put, onMutate) fails.
* The original error is available in the `cause` property.
*/
var EffectCacheError = class extends Schema.TaggedErrorClass()("EffectCacheError", { cause: Schema.Unknown }) {
	static [entityKind] = "EffectCacheError";
};

//#endregion
export { EffectCache, EffectCacheError };
