// drizzle-orm/cache/core/cache-effect.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_cache_core_cache = require('./cache.cjs');
let __entity_ts = require("../../entity.cjs");
let effect_Effect = require("effect/Effect");
effect_Effect = require_runtime.__toESM(effect_Effect);
let effect_Layer = require("effect/Layer");
effect_Layer = require_runtime.__toESM(effect_Layer);
let effect_Context = require("effect/Context");
effect_Context = require_runtime.__toESM(effect_Context);
let effect_Schema = require("effect/Schema");
effect_Schema = require_runtime.__toESM(effect_Schema);

//#region src/cache/core/cache-effect.ts
/**
* Effect service for caching query results in Drizzle ORM.
*
* By default, this service uses a no-op cache (no caching occurs). Use
* `EffectCache.fromDrizzle` to adapt a standard Drizzle cache implementation.
*
* @example
* ```ts
* // Use default (no caching)
* const db = yield* PgDrizzle.make({ relations }).pipe(
*   Effect.provide(PgDrizzle.DefaultServices),
* );
*
* // Use a custom Drizzle cache
* const db = yield* PgDrizzle.make({ relations }).pipe(
*   Effect.provide(EffectCache.layerFromDrizzle(myCache)),
*   Effect.provide(PgDrizzle.DefaultServices),
* );
* ```
*/
var EffectCache = class EffectCache extends effect_Context.Service()("drizzle-orm/EffectCache", { make: effect_Effect.sync(() => make(new require_cache_core_cache.NoopCache())) }) {
	static [__entity_ts.entityKind] = "drizzle-orm/EffectCache";
	/**
	* The default layer providing a no-op cache.
	*/
	static Default = effect_Layer.effect(EffectCache, EffectCache.make);
	/**
	* Creates an EffectCacheShape from a standard Drizzle cache.
	*
	* @param cache - A Drizzle cache instance implementing the `Cache` interface.
	* @returns A new EffectCacheShape that delegates to the provided Drizzle cache.
	*
	* @example
	* ```ts
	* const drizzleCache = new MyCustomCache();
	* const effectCache = EffectCache.fromDrizzle(drizzleCache);
	* ```
	*/
	static fromDrizzle(cache) {
		return make(cache);
	}
	/**
	* Creates a Layer that provides an EffectCache from a standard Drizzle cache.
	*
	* @param cache - A Drizzle cache instance implementing the `Cache` interface.
	* @returns A Layer that provides the EffectCache service.
	*
	* @example
	* ```ts
	* const drizzleCache = new MyCustomCache();
	* const db = yield* PgDrizzle.make({ relations }).pipe(
	*   Effect.provide(EffectCache.layerFromDrizzle(drizzleCache)),
	*   Effect.provide(PgDrizzle.DefaultServices),
	* );
	* ```
	*/
	static layerFromDrizzle(cache) {
		return effect_Layer.succeed(EffectCache, EffectCache.fromDrizzle(cache));
	}
};
function make(cache) {
	const strategy = () => cache.strategy();
	const get = (...args) => effect_Effect.tryPromise({
		try: () => cache.get(...args),
		catch: (error) => new EffectCacheError({ cause: error })
	});
	const put = (...args) => effect_Effect.tryPromise({
		try: () => cache.put(...args),
		catch: (error) => new EffectCacheError({ cause: error })
	});
	const onMutate = (params) => effect_Effect.tryPromise({
		try: () => cache.onMutate(params),
		catch: (error) => new EffectCacheError({ cause: error })
	});
	return {
		strategy,
		get,
		put,
		onMutate,
		cache
	};
}
/**
* Error type for cache operation failures.
*
* This error is thrown when any cache operation (get, put, onMutate) fails.
* The original error is available in the `cause` property.
*/
var EffectCacheError = class extends effect_Schema.TaggedErrorClass()("EffectCacheError", { cause: effect_Schema.Unknown }) {
	static [__entity_ts.entityKind] = "EffectCacheError";
};

//#endregion
exports.EffectCache = EffectCache;
exports.EffectCacheError = EffectCacheError;
