// drizzle-orm/mysql-proxy/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let __logger_ts = require("../logger.cjs");
let __cache_core_index_ts = require("../cache/core/index.cjs");
let __mysql_core_session_ts = require("../mysql-core/session.cjs");

//#region src/mysql-proxy/session.ts
var MySqlRemoteSession = class extends __mysql_core_session_ts.MySqlSession {
	static [__entity_ts.entityKind] = "MySqlRemoteSession";
	logger;
	cache;
	constructor(client, dialect, relations, options) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
		this.logger = options.logger ?? new __logger_ts.NoopLogger();
		this.cache = options.cache ?? new __cache_core_index_ts.NoopCache();
	}
	prepareQuery(query, mode, mapper, queryMetadata, cacheConfig) {
		const executor = async (params = []) => {
			const raw = this.client(query.sql, params, mode === "arrays" ? "all" : "execute");
			if (mode === "objects") return raw.then(({ rows }) => rows[0]);
			if (mode === "arrays" || !mapper) return raw.then(({ rows }) => rows);
			return raw.then(({ rows, insertId, affectedRows }) => ({
				insertId: insertId ?? rows[0]?.insertId,
				affectedRows: affectedRows ?? rows[0]?.affectedRows
			}));
		};
		return new __mysql_core_session_ts.MySqlPreparedQuery(executor, void 0, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	async transaction(_transaction, _config) {
		throw new Error("Transactions are not supported by the MySql Proxy driver");
	}
};

//#endregion
exports.MySqlRemoteSession = MySqlRemoteSession;
