// drizzle-orm/mysql-proxy/session.js
import { entityKind } from "../entity.js";
import { NoopLogger } from "../logger.js";
import { NoopCache } from "../cache/core/index.js";
import { MySqlPreparedQuery, MySqlSession } from "../mysql-core/session.js";

//#region src/mysql-proxy/session.ts
var MySqlRemoteSession = class extends MySqlSession {
	static [entityKind] = "MySqlRemoteSession";
	logger;
	cache;
	constructor(client, dialect, relations, options) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
		this.logger = options.logger ?? new NoopLogger();
		this.cache = options.cache ?? new NoopCache();
	}
	prepareQuery(query, mode, mapper, queryMetadata, cacheConfig) {
		const executor = async (params = []) => {
			const raw = this.client(query.sql, params, mode === "arrays" ? "all" : "execute");
			if (mode === "objects") return raw.then(({ rows }) => rows[0]);
			if (mode === "arrays" || !mapper) return raw.then(({ rows }) => rows);
			return raw.then(({ rows, insertId, affectedRows }) => ({
				insertId: insertId ?? rows[0]?.insertId,
				affectedRows: affectedRows ?? rows[0]?.affectedRows
			}));
		};
		return new MySqlPreparedQuery(executor, void 0, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	async transaction(_transaction, _config) {
		throw new Error("Transactions are not supported by the MySql Proxy driver");
	}
};

//#endregion
export { MySqlRemoteSession };
