// drizzle-orm/singlestore-proxy/migrator.js
import { sql } from "../sql/sql.js";
import { readMigrationFiles } from "../migrator.js";
import { getMigrationsToRun } from "../migrator.utils.js";
import { upgradeIfNeeded } from "../up-migrations/singlestore-proxy.js";

//#region src/singlestore-proxy/migrator.ts
async function migrate(db, callback, config) {
	const migrations = readMigrationFiles(config);
	const migrationsTable = config.migrationsTable ?? "__drizzle_migrations";
	const { newDb } = await upgradeIfNeeded(migrationsTable, db, callback, migrations);
	if (newDb) {
		const migrationTableCreate = sql`
			CREATE TABLE IF NOT EXISTS ${sql.identifier(migrationsTable)} (
				id SERIAL PRIMARY KEY,
				hash TEXT NOT NULL,
				created_at BIGINT,
				name TEXT,
				applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
			)
		`;
		await db.session.execute(migrationTableCreate);
	}
	const dbMigrations = await db.select({
		id: sql.raw("id"),
		hash: sql.raw("hash"),
		created_at: sql.raw("created_at"),
		name: sql.raw("name")
	}).from(sql.identifier(migrationsTable).getSQL());
	if (typeof config === "object" && config.init) {
		if (dbMigrations.length) return { exitCode: "databaseMigrations" };
		if (migrations.length > 1) return { exitCode: "localMigrations" };
		const [migration] = migrations;
		if (!migration) return;
		await callback([db.dialect.sqlToQuery(sql`insert into ${sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`, \`name\`) values(${migration.hash}, '${migration.folderMillis}', ${migration.name})`.inlineParams()).sql]);
		return;
	}
	const migrationsToRun = getMigrationsToRun({
		localMigrations: migrations,
		dbMigrations
	});
	const queriesToRun = [];
	for (const migration of migrationsToRun) queriesToRun.push(...migration.sql, db.dialect.sqlToQuery(sql`insert into ${sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`, \`name\`) values(${migration.hash}, '${migration.folderMillis}', ${migration.name})`.inlineParams()).sql);
	await callback(queriesToRun);
}

//#endregion
export { migrate };
