// drizzle-orm/singlestore-proxy/migrator.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __sql_sql_ts = require("../sql/sql.cjs");
let __migrator_ts = require("../migrator.cjs");
let __migrator_utils_ts = require("../migrator.utils.cjs");
let __up_migrations_singlestore_proxy_ts = require("../up-migrations/singlestore-proxy.cjs");

//#region src/singlestore-proxy/migrator.ts
async function migrate(db, callback, config) {
	const migrations = (0, __migrator_ts.readMigrationFiles)(config);
	const migrationsTable = config.migrationsTable ?? "__drizzle_migrations";
	const { newDb } = await (0, __up_migrations_singlestore_proxy_ts.upgradeIfNeeded)(migrationsTable, db, callback, migrations);
	if (newDb) {
		const migrationTableCreate = __sql_sql_ts.sql`
			CREATE TABLE IF NOT EXISTS ${__sql_sql_ts.sql.identifier(migrationsTable)} (
				id SERIAL PRIMARY KEY,
				hash TEXT NOT NULL,
				created_at BIGINT,
				name TEXT,
				applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
			)
		`;
		await db.session.execute(migrationTableCreate);
	}
	const dbMigrations = await db.select({
		id: __sql_sql_ts.sql.raw("id"),
		hash: __sql_sql_ts.sql.raw("hash"),
		created_at: __sql_sql_ts.sql.raw("created_at"),
		name: __sql_sql_ts.sql.raw("name")
	}).from(__sql_sql_ts.sql.identifier(migrationsTable).getSQL());
	if (typeof config === "object" && config.init) {
		if (dbMigrations.length) return { exitCode: "databaseMigrations" };
		if (migrations.length > 1) return { exitCode: "localMigrations" };
		const [migration] = migrations;
		if (!migration) return;
		await callback([db.dialect.sqlToQuery(__sql_sql_ts.sql`insert into ${__sql_sql_ts.sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`, \`name\`) values(${migration.hash}, '${migration.folderMillis}', ${migration.name})`.inlineParams()).sql]);
		return;
	}
	const migrationsToRun = (0, __migrator_utils_ts.getMigrationsToRun)({
		localMigrations: migrations,
		dbMigrations
	});
	const queriesToRun = [];
	for (const migration of migrationsToRun) queriesToRun.push(...migration.sql, db.dialect.sqlToQuery(__sql_sql_ts.sql`insert into ${__sql_sql_ts.sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`, \`name\`) values(${migration.hash}, '${migration.folderMillis}', ${migration.name})`.inlineParams()).sql);
	await callback(queriesToRun);
}

//#endregion
exports.migrate = migrate;
