// drizzle-orm/planetscale-serverless/driver.js
import { PlanetscaleSession } from "./session.js";
import { entityKind } from "../entity.js";
import { jitCompatCheck } from "../utils.js";
import { DefaultLogger } from "../logger.js";
import { MySqlDatabase } from "../mysql-core/db.js";
import { MySqlDialect } from "../mysql-core/dialect.js";
import { Client } from "@planetscale/database";

//#region src/planetscale-serverless/driver.ts
var PlanetScaleDatabase = class extends MySqlDatabase {
	static [entityKind] = "PlanetScaleDatabase";
};
function construct(client, config = {}) {
	if (!(client instanceof Client)) throw new Error(`Warning: You need to pass an instance of Client:

import { Client } from "@planetscale/database";

const client = new Client({
  host: process.env["DATABASE_HOST"],
  username: process.env["DATABASE_USERNAME"],
  password: process.env["DATABASE_PASSWORD"],
});

const db = drizzle({ client });
		`);
	const dialect = new MySqlDialect({ useJitMappers: jitCompatCheck(config.jit) });
	let logger;
	if (config.logger === true) logger = new DefaultLogger();
	else if (config.logger !== false) logger = config.logger;
	const relations = config.relations ?? {};
	const db = new PlanetScaleDatabase(dialect, new PlanetscaleSession(client, dialect, void 0, relations, {
		logger,
		cache: config.cache
	}), relations);
	db.$client = client;
	db.$cache = config.cache;
	if (db.$cache) db.$cache["invalidate"] = config.cache?.onMutate;
	return db;
}
function drizzle(...params) {
	if (typeof params[0] === "string") return construct(new Client({ url: params[0] }), params[1]);
	const { connection, client, ...DrizzleMySqlConfig } = params[0];
	if (client) return construct(client, DrizzleMySqlConfig);
	return construct(typeof connection === "string" ? new Client({ url: connection }) : new Client(connection), DrizzleMySqlConfig);
}
(function(_drizzle) {
	function mock(config) {
		return construct({}, config);
	}
	_drizzle.mock = mock;
})(drizzle || (drizzle = {}));

//#endregion
export { PlanetScaleDatabase, drizzle };
