// drizzle-orm/planetscale-serverless/session.js
import { entityKind } from "../entity.js";
import { sql } from "../sql/sql.js";
import { NoopLogger } from "../logger.js";
import { NoopCache } from "../cache/core/index.js";
import { MySqlPreparedQuery, MySqlSession, MySqlTransaction } from "../mysql-core/session.js";

//#region src/planetscale-serverless/session.ts
var PlanetscaleSession = class PlanetscaleSession extends MySqlSession {
	static [entityKind] = "PlanetscaleSession";
	logger;
	client;
	cache;
	constructor(baseClient, dialect, tx, relations, options = {}) {
		super(dialect);
		this.baseClient = baseClient;
		this.relations = relations;
		this.options = options;
		this.client = tx ?? baseClient;
		this.logger = options.logger ?? new NoopLogger();
		this.cache = options.cache ?? new NoopCache();
	}
	prepareQuery(query, mode, mapper, queryMetadata, cacheConfig) {
		const { client } = this;
		const queryConfig = { as: mode === "arrays" ? "array" : "object" };
		const executor = async (params = []) => {
			const raw = client.execute(query.sql, params, queryConfig);
			if (mode !== "raw") return raw.then(({ rows }) => rows);
			if (!mapper) return raw;
			return raw.then(({ insertId, rowsAffected }) => ({
				insertId: Number.parseFloat(insertId),
				affectedRows: rowsAffected
			}));
		};
		return new MySqlPreparedQuery(executor, void 0, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	transaction(transaction) {
		return this.baseClient.transaction((pstx) => {
			const session = new PlanetscaleSession(this.baseClient, this.dialect, pstx, this.relations, this.options);
			return transaction(new PlanetScaleTransaction(this.dialect, session, this.relations));
		});
	}
};
var PlanetScaleTransaction = class PlanetScaleTransaction extends MySqlTransaction {
	static [entityKind] = "PlanetScaleTransaction";
	constructor(dialect, session, relations, nestedIndex = 0) {
		super(dialect, session, relations, nestedIndex);
	}
	async transaction(transaction) {
		const savepointName = `sp${this.nestedIndex + 1}`;
		const tx = new PlanetScaleTransaction(this.dialect, this.session, this.relations, this.nestedIndex + 1);
		await tx.execute(sql.raw(`savepoint ${savepointName}`));
		try {
			const result = await transaction(tx);
			await tx.execute(sql.raw(`release savepoint ${savepointName}`));
			return result;
		} catch (err) {
			await tx.execute(sql.raw(`rollback to savepoint ${savepointName}`));
			throw err;
		}
	}
};

//#endregion
export { PlanetScaleTransaction, PlanetscaleSession };
