// drizzle-orm/planetscale-serverless/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");
let __logger_ts = require("../logger.cjs");
let __cache_core_index_ts = require("../cache/core/index.cjs");
let __mysql_core_session_ts = require("../mysql-core/session.cjs");

//#region src/planetscale-serverless/session.ts
var PlanetscaleSession = class PlanetscaleSession extends __mysql_core_session_ts.MySqlSession {
	static [__entity_ts.entityKind] = "PlanetscaleSession";
	logger;
	client;
	cache;
	constructor(baseClient, dialect, tx, relations, options = {}) {
		super(dialect);
		this.baseClient = baseClient;
		this.relations = relations;
		this.options = options;
		this.client = tx ?? baseClient;
		this.logger = options.logger ?? new __logger_ts.NoopLogger();
		this.cache = options.cache ?? new __cache_core_index_ts.NoopCache();
	}
	prepareQuery(query, mode, mapper, queryMetadata, cacheConfig) {
		const { client } = this;
		const queryConfig = { as: mode === "arrays" ? "array" : "object" };
		const executor = async (params = []) => {
			const raw = client.execute(query.sql, params, queryConfig);
			if (mode !== "raw") return raw.then(({ rows }) => rows);
			if (!mapper) return raw;
			return raw.then(({ insertId, rowsAffected }) => ({
				insertId: Number.parseFloat(insertId),
				affectedRows: rowsAffected
			}));
		};
		return new __mysql_core_session_ts.MySqlPreparedQuery(executor, void 0, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	transaction(transaction) {
		return this.baseClient.transaction((pstx) => {
			const session = new PlanetscaleSession(this.baseClient, this.dialect, pstx, this.relations, this.options);
			return transaction(new PlanetScaleTransaction(this.dialect, session, this.relations));
		});
	}
};
var PlanetScaleTransaction = class PlanetScaleTransaction extends __mysql_core_session_ts.MySqlTransaction {
	static [__entity_ts.entityKind] = "PlanetScaleTransaction";
	constructor(dialect, session, relations, nestedIndex = 0) {
		super(dialect, session, relations, nestedIndex);
	}
	async transaction(transaction) {
		const savepointName = `sp${this.nestedIndex + 1}`;
		const tx = new PlanetScaleTransaction(this.dialect, this.session, this.relations, this.nestedIndex + 1);
		await tx.execute(__sql_sql_ts.sql.raw(`savepoint ${savepointName}`));
		try {
			const result = await transaction(tx);
			await tx.execute(__sql_sql_ts.sql.raw(`release savepoint ${savepointName}`));
			return result;
		} catch (err) {
			await tx.execute(__sql_sql_ts.sql.raw(`rollback to savepoint ${savepointName}`));
			throw err;
		}
	}
};

//#endregion
exports.PlanetScaleTransaction = PlanetScaleTransaction;
exports.PlanetscaleSession = PlanetscaleSession;
