// waddler/sqlite/expo-sqlite/session.js
import { WaddlerQueryError } from "../../errors/index.js";
import { SQLTemplate } from "../../sql-template.js";
class ExpoSqliteSQLTemplate extends SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
  }
  returningData = true;
  all() {
    this.returningData = true;
    return this;
  }
  run() {
    this.returningData = false;
    return this;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    const stmt = this.client.prepareSync(query);
    try {
      if (this.returningData === true) {
        if (this.options.rowMode === "array") {
          const queryResult = stmt.executeForRawResultSync(params);
          finalRes = queryResult.getAllSync();
          finalMetadata = { changes: queryResult.changes, lastInsertRowId: queryResult.lastInsertRowId };
        } else {
          const queryResult = stmt.executeSync(params);
          finalRes = queryResult.getAllSync();
          finalMetadata = { changes: queryResult.changes, lastInsertRowId: queryResult.lastInsertRowId };
        }
      } else {
        const queryResult = await stmt.executeAsync(params);
        finalMetadata = { changes: queryResult.changes, lastInsertRowId: queryResult.lastInsertRowId };
        finalRes = queryResult;
      }
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    } finally {
      await stmt.finalizeAsync();
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  async *stream() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    const stmt = this.client.prepareSync(query);
    try {
      const stream = await (this.options.rowMode === "array" ? stmt.executeForRawResultAsync(params) : stmt.executeAsync(params));
      for await (const row of stream) {
        yield row;
      }
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    } finally {
      await stmt.finalizeAsync();
    }
  }
}
export {
  ExpoSqliteSQLTemplate
};
