// waddler/sqlite/expo-sqlite/session.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var session_exports = {};
__export(session_exports, {
  ExpoSqliteSQLTemplate: () => ExpoSqliteSQLTemplate
});
module.exports = __toCommonJS(session_exports);
var import_errors = require("../../errors/index.cjs");
var import_sql_template = require("../../sql-template.cjs");
class ExpoSqliteSQLTemplate extends import_sql_template.SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
  }
  returningData = true;
  all() {
    this.returningData = true;
    return this;
  }
  run() {
    this.returningData = false;
    return this;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    const stmt = this.client.prepareSync(query);
    try {
      if (this.returningData === true) {
        if (this.options.rowMode === "array") {
          const queryResult = stmt.executeForRawResultSync(params);
          finalRes = queryResult.getAllSync();
          finalMetadata = { changes: queryResult.changes, lastInsertRowId: queryResult.lastInsertRowId };
        } else {
          const queryResult = stmt.executeSync(params);
          finalRes = queryResult.getAllSync();
          finalMetadata = { changes: queryResult.changes, lastInsertRowId: queryResult.lastInsertRowId };
        }
      } else {
        const queryResult = await stmt.executeAsync(params);
        finalMetadata = { changes: queryResult.changes, lastInsertRowId: queryResult.lastInsertRowId };
        finalRes = queryResult;
      }
    } catch (error) {
      throw new import_errors.WaddlerQueryError(query, params, error);
    } finally {
      await stmt.finalizeAsync();
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  async *stream() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    const stmt = this.client.prepareSync(query);
    try {
      const stream = await (this.options.rowMode === "array" ? stmt.executeForRawResultAsync(params) : stmt.executeAsync(params));
      for await (const row of stream) {
        yield row;
      }
    } catch (error) {
      throw new import_errors.WaddlerQueryError(query, params, error);
    } finally {
      await stmt.finalizeAsync();
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ExpoSqliteSQLTemplate
});
