// drizzle-orm/mysql2/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");
let __logger_ts = require("../logger.cjs");
let __cache_core_index_ts = require("../cache/core/index.cjs");
let __mysql_core_session_ts = require("../mysql-core/session.cjs");
let node_events = require("node:events");

//#region src/mysql2/session.ts
const typeCast = function(field, next) {
	if (field.type === "TIMESTAMP" || field.type === "DATETIME" || field.type === "DATE") return field.string();
	return next();
};
var MySql2Session = class MySql2Session extends __mysql_core_session_ts.MySqlSession {
	static [__entity_ts.entityKind] = "MySql2Session";
	logger;
	cache;
	constructor(client, dialect, relations, options) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
		this.logger = options.logger ?? new __logger_ts.NoopLogger();
		this.cache = options.cache ?? new __cache_core_index_ts.NoopCache();
	}
	prepareQuery(query, mode, mapper, queryMetadata, cacheConfig) {
		const { client } = this;
		const executor = async (params = []) => {
			const raw = client.query({
				sql: query.sql,
				typeCast,
				rowsAsArray: mode === "arrays"
			}, params);
			if (mode !== "raw") return raw.then((data) => data[0]);
			if (!mapper) return raw;
			return raw.then(([res]) => ({
				insertId: res.insertId,
				affectedRows: res.affectedRows
			}));
		};
		const iterator = async function* (params = []) {
			const conn = (isPool(client) ? await client.getConnection() : client).connection;
			const stream = conn.query({
				sql: query.sql,
				typeCast,
				rowsAsArray: mode === "arrays"
			}, params).stream();
			function dataListener() {
				stream.pause();
			}
			stream.on("data", dataListener);
			try {
				const onEnd = (0, node_events.once)(stream, "end");
				const onError = (0, node_events.once)(stream, "error");
				while (true) {
					stream.resume();
					const row = await Promise.race([
						onEnd,
						onError,
						new Promise((resolve) => stream.once("data", resolve))
					]);
					if (row === void 0 || Array.isArray(row) && row.length === 0) break;
					if (row instanceof Error) throw row;
					yield row;
				}
			} finally {
				stream.off("data", dataListener);
				if (isPool(client)) conn.end();
			}
		};
		return new __mysql_core_session_ts.MySqlPreparedQuery(executor, iterator, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	async transaction(transaction, config) {
		const session = isPool(this.client) ? new MySql2Session(await this.client.getConnection(), this.dialect, this.relations, this.options) : this;
		const tx = new MySql2Transaction(this.dialect, session, this.relations, 0);
		if (config) {
			const setTransactionConfigSql = this.getSetTransactionSQL(config);
			if (setTransactionConfigSql) await tx.execute(setTransactionConfigSql);
			const startTransactionSql = this.getStartTransactionSQL(config);
			await (startTransactionSql ? tx.execute(startTransactionSql) : tx.execute(__sql_sql_ts.sql`begin`));
		} else await tx.execute(__sql_sql_ts.sql`begin`);
		try {
			const result = await transaction(tx);
			await tx.execute(__sql_sql_ts.sql`commit`);
			return result;
		} catch (err) {
			await tx.execute(__sql_sql_ts.sql`rollback`);
			throw err;
		} finally {
			if (isPool(this.client)) session.client.release();
		}
	}
};
var MySql2Transaction = class MySql2Transaction extends __mysql_core_session_ts.MySqlTransaction {
	static [__entity_ts.entityKind] = "MySql2Transaction";
	async transaction(transaction) {
		const savepointName = `sp${this.nestedIndex + 1}`;
		const tx = new MySql2Transaction(this.dialect, this.session, this.relations, this.nestedIndex + 1);
		await tx.execute(__sql_sql_ts.sql.raw(`savepoint ${savepointName}`));
		try {
			const result = await transaction(tx);
			await tx.execute(__sql_sql_ts.sql.raw(`release savepoint ${savepointName}`));
			return result;
		} catch (err) {
			await tx.execute(__sql_sql_ts.sql.raw(`rollback to savepoint ${savepointName}`));
			throw err;
		}
	}
};
function isPool(client) {
	return "getConnection" in client;
}

//#endregion
exports.MySql2Session = MySql2Session;
exports.MySql2Transaction = MySql2Transaction;
