// drizzle-orm/codecs.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_entity = require('./entity.cjs');

//#region src/codecs.ts
const noopCodecs = {};
const arrayToItemTypeCodecNameMap = {
	cast: "cast",
	castArray: "cast",
	castInJson: "castInJson",
	castArrayInJson: "castInJson",
	castParam: "castParam",
	castArrayParam: "castParam",
	normalize: "normalize",
	normalizeArray: "normalize",
	normalizeInJson: "normalizeInJson",
	normalizeArrayInJson: "normalizeInJson",
	normalizeParam: "normalizeParam",
	normalizeParamArray: "normalizeParam"
};
const itemToArrayTypeCodecNameMap = {
	cast: "castArray",
	castArray: "castArray",
	castInJson: "castArrayInJson",
	castArrayInJson: "castArrayInJson",
	castParam: "castArrayParam",
	castArrayParam: "castArrayParam",
	normalize: "normalizeArray",
	normalizeArray: "normalizeArray",
	normalizeInJson: "normalizeArrayInJson",
	normalizeArrayInJson: "normalizeArrayInJson",
	normalizeParam: "normalizeParamArray",
	normalizeParamArray: "normalizeParamArray"
};
var CodecsCollection = class {
	static [require_entity.entityKind] = "CodecsCollection";
	constructor(resolveTypes, codecs = noopCodecs) {
		this.resolveTypes = resolveTypes;
		this.codecs = codecs;
	}
	get(column, type) {
		const sqlType = column.codec;
		if (!sqlType) return void 0;
		const codecType = column.dimensions ? itemToArrayTypeCodecNameMap[type] : arrayToItemTypeCodecNameMap[type];
		return this.codecs[sqlType]?.[codecType];
	}
	apply(column, type, value) {
		const sqlType = column.codec;
		if (!sqlType) return value;
		const codecType = column.dimensions ? itemToArrayTypeCodecNameMap[type] : arrayToItemTypeCodecNameMap[type];
		const codec = this.codecs[sqlType]?.[codecType];
		if (!codec) return value;
		if (codecType === "castParam" || codecType === "castArrayParam") return codec(value, column, column.dimensions);
		return codec(value, column.dimensions);
	}
};
function refineCodecs(source, extension = {}) {
	const keys = new Set([...Object.keys(source), ...Object.keys(extension)]).values();
	const result = {};
	for (const k of keys) {
		if (!(k in extension)) {
			result[k] = source[k] ? { ...source[k] } : void 0;
			continue;
		}
		if (!(k in source) || extension[k] === void 0) {
			result[k] = extension[k] ? { ...extension[k] } : void 0;
			continue;
		}
		const innerKeys = new Set([...Object.keys(extension[k]), ...Object.keys(source[k] ?? {})]).values();
		result[k] = {};
		for (const ik of innerKeys) result[k][ik] = ik in extension[k] ? extension[k][ik] : source[k]?.[ik];
	}
	return result;
}

//#endregion
exports.CodecsCollection = CodecsCollection;
exports.noopCodecs = noopCodecs;
exports.refineCodecs = refineCodecs;
