// waddler/sqlite/better-sqlite3/session.js
import { WaddlerQueryError } from "../../errors/index.js";
import { SQLTemplate } from "../../sql-template.js";
class BetterSqlite3SQLTemplate extends SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
  }
  returningData = true;
  all() {
    this.returningData = true;
    return this;
  }
  run() {
    this.returningData = false;
    return this;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      const stmt = this.client.prepare(query);
      if (this.returningData) {
        finalRes = this.options.rowMode === "array" ? stmt.raw().all(...params) : stmt.all(...params);
      } else {
        finalRes = stmt.run(...params);
        finalMetadata = finalRes;
      }
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  async *stream() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    try {
      const stmt = this.client.prepare(query);
      const stream = this.options.rowMode === "array" ? stmt.raw().iterate(...params) : stmt.iterate(...params);
      for await (const row of stream) {
        yield row;
      }
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    }
  }
}
export {
  BetterSqlite3SQLTemplate
};
