// drizzle-orm/arktype/schema.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
const require_utils = require('../utils.cjs');
const require_arktype_column = require('./column.cjs');
let __entity_ts = require("../entity.cjs");
let __table_ts = require("../table.cjs");
let __column_ts = require("../column.cjs");
let __utils_ts = require("../utils.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");
let arktype = require("arktype");

//#region src/arktype/schema.ts
function handleColumns(columns, refinements, conditions) {
	const columnSchemas = {};
	for (const [key, selected] of Object.entries(columns)) {
		if (!(0, __entity_ts.is)(selected, __column_ts.Column) && !(0, __entity_ts.is)(selected, __sql_sql_ts.SQL) && !(0, __entity_ts.is)(selected, __sql_sql_ts.SQL.Aliased) && typeof selected === "object") {
			columnSchemas[key] = handleColumns((0, __table_ts.isTable)(selected) || (0, __sql_sql_ts.isView)(selected) ? (0, __utils_ts.getColumns)(selected) : selected, refinements[key] ?? {}, conditions);
			continue;
		}
		const refinement = refinements[key];
		if (refinement !== void 0 && (typeof refinement !== "function" || typeof refinement === "function" && refinement.expression !== void 0)) {
			columnSchemas[key] = refinement;
			continue;
		}
		const column = (0, __entity_ts.is)(selected, __column_ts.Column) ? selected : void 0;
		const schema = column ? require_arktype_column.columnToSchema(column) : arktype.type.unknown;
		const refined = typeof refinement === "function" ? refinement(schema) : schema;
		if (conditions.never(column)) continue;
		else columnSchemas[key] = refined;
		if (column) {
			if (conditions.nullable(column)) columnSchemas[key] = columnSchemas[key].or(arktype.type.null);
			if (conditions.optional(column)) columnSchemas[key] = columnSchemas[key].optional();
		}
	}
	return (0, arktype.type)(columnSchemas);
}
const createSelectSchema = ((entity, refine) => {
	if (require_utils.isWithEnum(entity)) return arktype.type.enumerated(...entity.enumValues);
	return handleColumns((0, __utils_ts.getColumns)(entity), refine ?? {}, {
		never: () => false,
		optional: () => false,
		nullable: (column) => !column.notNull
	});
});
const createInsertSchema = ((entity, refine) => {
	return handleColumns((0, __utils_ts.getColumns)(entity), refine ?? {}, {
		never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
		optional: (column) => !column.notNull || column.notNull && column.hasDefault,
		nullable: (column) => !column.notNull
	});
});
const createUpdateSchema = ((entity, refine) => {
	return handleColumns((0, __utils_ts.getColumns)(entity), refine ?? {}, {
		never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
		optional: () => true,
		nullable: (column) => !column.notNull
	});
});

//#endregion
exports.createInsertSchema = createInsertSchema;
exports.createSelectSchema = createSelectSchema;
exports.createUpdateSchema = createUpdateSchema;
