// drizzle-orm/arktype/schema.js
import { isWithEnum } from "../utils.js";
import { columnToSchema } from "./column.js";
import { is } from "../entity.js";
import { isTable } from "../table.js";
import { Column } from "../column.js";
import { getColumns } from "../utils.js";
import { SQL, isView } from "../sql/sql.js";
import { type } from "arktype";

//#region src/arktype/schema.ts
function handleColumns(columns, refinements, conditions) {
	const columnSchemas = {};
	for (const [key, selected] of Object.entries(columns)) {
		if (!is(selected, Column) && !is(selected, SQL) && !is(selected, SQL.Aliased) && typeof selected === "object") {
			columnSchemas[key] = handleColumns(isTable(selected) || isView(selected) ? getColumns(selected) : selected, refinements[key] ?? {}, conditions);
			continue;
		}
		const refinement = refinements[key];
		if (refinement !== void 0 && (typeof refinement !== "function" || typeof refinement === "function" && refinement.expression !== void 0)) {
			columnSchemas[key] = refinement;
			continue;
		}
		const column = is(selected, Column) ? selected : void 0;
		const schema = column ? columnToSchema(column) : type.unknown;
		const refined = typeof refinement === "function" ? refinement(schema) : schema;
		if (conditions.never(column)) continue;
		else columnSchemas[key] = refined;
		if (column) {
			if (conditions.nullable(column)) columnSchemas[key] = columnSchemas[key].or(type.null);
			if (conditions.optional(column)) columnSchemas[key] = columnSchemas[key].optional();
		}
	}
	return type(columnSchemas);
}
const createSelectSchema = ((entity, refine) => {
	if (isWithEnum(entity)) return type.enumerated(...entity.enumValues);
	return handleColumns(getColumns(entity), refine ?? {}, {
		never: () => false,
		optional: () => false,
		nullable: (column) => !column.notNull
	});
});
const createInsertSchema = ((entity, refine) => {
	return handleColumns(getColumns(entity), refine ?? {}, {
		never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
		optional: (column) => !column.notNull || column.notNull && column.hasDefault,
		nullable: (column) => !column.notNull
	});
});
const createUpdateSchema = ((entity, refine) => {
	return handleColumns(getColumns(entity), refine ?? {}, {
		never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
		optional: () => true,
		nullable: (column) => !column.notNull
	});
});

//#endregion
export { createInsertSchema, createSelectSchema, createUpdateSchema };
