// waddler/pool-ts/Deque.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Deque_exports = {};
__export(Deque_exports, {
  Deque: () => Deque
});
module.exports = __toCommonJS(Deque_exports);
var import_DequeIterator = require("./DequeIterator.cjs");
var import_DoublyLinkedList = require("./DoublyLinkedList.cjs");
class Deque {
  _list;
  constructor() {
    this._list = new import_DoublyLinkedList.DoublyLinkedList();
  }
  /**
   * removes and returns the first element from the queue
   * @return [description]
   */
  shift() {
    if (this.length <= 0) {
      return null;
    }
    const node = this._list.head;
    this._list.remove(node);
    return node.data;
  }
  /**
   * adds one elemts to the beginning of the queue
   * @param  {any} element [description]
   * @return           [description]
   */
  unshift(element) {
    const node = new import_DoublyLinkedList.Node(element);
    this._list.insertBeginning(node);
  }
  /**
   * adds one to the end of the queue
   * @param  {any} element [description]
   * @return          [description]
   */
  push(element) {
    const node = new import_DoublyLinkedList.Node(element);
    this._list.insertEnd(node);
  }
  /**
   * removes and returns the last element from the queue
   */
  pop() {
    if (this.length <= 0) {
      return null;
    }
    const node = this._list.tail;
    this._list.remove(node);
    return node.data;
  }
  [Symbol.iterator]() {
    return new import_DequeIterator.DequeIterator(this._list);
  }
  iterator() {
    return new import_DequeIterator.DequeIterator(this._list);
  }
  reverseIterator() {
    return new import_DequeIterator.DequeIterator(this._list, true);
  }
  /**
   * get a reference to the item at the head of the queue
   * @return [description]
   */
  get head() {
    if (this.length === 0) {
      return null;
    }
    const node = this._list.head;
    return node.data;
  }
  /**
   * get a reference to the item at the tail of the queue
   * @return  [description]
   */
  get tail() {
    if (this.length === 0) {
      return null;
    }
    const node = this._list.tail;
    return node.data;
  }
  get length() {
    return this._list.length;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Deque
});
