// waddler/pool-ts/Deque.js
import { DequeIterator } from "./DequeIterator.js";
import { DoublyLinkedList, Node } from "./DoublyLinkedList.js";
class Deque {
  _list;
  constructor() {
    this._list = new DoublyLinkedList();
  }
  /**
   * removes and returns the first element from the queue
   * @return [description]
   */
  shift() {
    if (this.length <= 0) {
      return null;
    }
    const node = this._list.head;
    this._list.remove(node);
    return node.data;
  }
  /**
   * adds one elemts to the beginning of the queue
   * @param  {any} element [description]
   * @return           [description]
   */
  unshift(element) {
    const node = new Node(element);
    this._list.insertBeginning(node);
  }
  /**
   * adds one to the end of the queue
   * @param  {any} element [description]
   * @return          [description]
   */
  push(element) {
    const node = new Node(element);
    this._list.insertEnd(node);
  }
  /**
   * removes and returns the last element from the queue
   */
  pop() {
    if (this.length <= 0) {
      return null;
    }
    const node = this._list.tail;
    this._list.remove(node);
    return node.data;
  }
  [Symbol.iterator]() {
    return new DequeIterator(this._list);
  }
  iterator() {
    return new DequeIterator(this._list);
  }
  reverseIterator() {
    return new DequeIterator(this._list, true);
  }
  /**
   * get a reference to the item at the head of the queue
   * @return [description]
   */
  get head() {
    if (this.length === 0) {
      return null;
    }
    const node = this._list.head;
    return node.data;
  }
  /**
   * get a reference to the item at the tail of the queue
   * @return  [description]
   */
  get tail() {
    if (this.length === 0) {
      return null;
    }
    const node = this._list.tail;
    return node.data;
  }
  get length() {
    return this._list.length;
  }
}
export {
  Deque
};
