// waddler/pool-ts/PooledResource.js
var PooledResourceState = /* @__PURE__ */ ((PooledResourceState2) => {
  PooledResourceState2["ALLOCATED"] = "ALLOCATED";
  PooledResourceState2["IDLE"] = "IDLE";
  PooledResourceState2["INVALID"] = "INVALID";
  PooledResourceState2["RETURNING"] = "RETURNING";
  PooledResourceState2["VALIDATION"] = "VALIDATION";
  return PooledResourceState2;
})(PooledResourceState || {});
class PooledResource {
  creationTime;
  lastReturnTime;
  lastBorrowTime;
  lastIdleTime;
  obj;
  state;
  constructor(resource) {
    this.creationTime = Date.now();
    this.lastReturnTime = null;
    this.lastBorrowTime = null;
    this.lastIdleTime = null;
    this.obj = resource;
    this.state = "IDLE" /* IDLE */;
  }
  // mark the resource as "allocated"
  allocate() {
    this.lastBorrowTime = Date.now();
    this.state = "ALLOCATED" /* ALLOCATED */;
  }
  // mark the resource as "deallocated"
  deallocate() {
    this.lastReturnTime = Date.now();
    this.state = "IDLE" /* IDLE */;
  }
  invalidate() {
    this.state = "INVALID" /* INVALID */;
  }
  test() {
    this.state = "VALIDATION" /* VALIDATION */;
  }
  idle() {
    this.lastIdleTime = Date.now();
    this.state = "IDLE" /* IDLE */;
  }
  returning() {
    this.state = "RETURNING" /* RETURNING */;
  }
}
export {
  PooledResource
};
