// waddler/pool-ts/Pool.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Pool_exports = {};
__export(Pool_exports, {
  Pool: () => Pool
});
module.exports = __toCommonJS(Pool_exports);
var import_Deferred = require("./Deferred.cjs");
var import_factoryValidator = require("./factoryValidator.cjs");
var import_PooledResource = require("./PooledResource.cjs");
var import_PoolOptions = require("./PoolOptions.cjs");
var import_ResourceLoan = require("./ResourceLoan.cjs");
var import_ResourceRequest = require("./ResourceRequest.cjs");
var import_utils = require("./utils.cjs");
class Pool {
  _config;
  promiseConstructor;
  _factory;
  _draining;
  _started;
  _waitingClientsQueue;
  _factoryCreateOperations;
  _factoryDestroyOperations;
  _availableObjects;
  _testOnBorrowResources;
  _testOnReturnResources;
  _validationOperations;
  _allObjects;
  _resourceLoans;
  _evictionIterator;
  _evictor;
  _scheduledEviction;
  constructor(Evictor, dequeConstructor, priorityQueueConstructor, factory, options) {
    (0, import_factoryValidator.validateFactory)(factory);
    this._config = new import_PoolOptions.PoolOptions(options);
    this.promiseConstructor = this._config.promiseConstructor;
    this._factory = factory;
    this._draining = false;
    this._started = false;
    this._waitingClientsQueue = new priorityQueueConstructor(this._config.priorityRange);
    this._factoryCreateOperations = /* @__PURE__ */ new Set();
    this._factoryDestroyOperations = /* @__PURE__ */ new Set();
    this._availableObjects = new dequeConstructor();
    this._testOnBorrowResources = /* @__PURE__ */ new Set();
    this._testOnReturnResources = /* @__PURE__ */ new Set();
    this._validationOperations = /* @__PURE__ */ new Set();
    this._allObjects = /* @__PURE__ */ new Set();
    this._resourceLoans = /* @__PURE__ */ new Map();
    this._evictionIterator = this._availableObjects.iterator();
    this._evictor = new Evictor();
    this._scheduledEviction = null;
  }
  async _destroy(pooledResource, ensureMinimum = true) {
    pooledResource.invalidate();
    this._allObjects.delete(pooledResource);
    await this._factory.destroy(pooledResource.obj);
    if (ensureMinimum === true) await this._ensureMinimum();
  }
  // not in use for now
  _applyDestroyTimeout(promise) {
    const timeoutPromise = new this.promiseConstructor((resolve, reject) => {
      setTimeout(() => {
        reject(new Error("destroy timed out"));
      }, this._config.destroyTimeoutMillis).unref();
    });
    return this.promiseConstructor.race([timeoutPromise, promise]);
  }
  // not in use because this._config.testOnBorrow default equals false
  _testOnBorrow() {
    if (this._availableObjects.length < 1) {
      return false;
    }
    const pooledResource = this._availableObjects.shift();
    pooledResource.test();
    this._testOnBorrowResources.add(pooledResource);
    const validationPromise = this._factory.validate === void 0 ? Promise.resolve(true) : this._factory.validate(pooledResource.obj);
    const wrappedValidationPromise = this.promiseConstructor.resolve(validationPromise);
    this._trackOperation(
      wrappedValidationPromise,
      this._validationOperations
    ).then((isValid) => {
      this._testOnBorrowResources.delete(pooledResource);
      if (!isValid) {
        pooledResource.invalidate();
        this._destroy(pooledResource);
        this._dispense();
        return;
      }
      this._dispatchPooledResourceToNextWaitingClient(pooledResource);
    });
    return true;
  }
  _dispatchResource() {
    if (this._availableObjects.length < 1) {
      return false;
    }
    const pooledResource = this._availableObjects.shift();
    return this._dispatchPooledResourceToNextWaitingClient(pooledResource);
  }
  async _dispense() {
    const numWaitingClients = this._waitingClientsQueue.length;
    if (numWaitingClients < 1) {
      return;
    }
    const resourceShortfall = numWaitingClients - this._potentiallyAllocableResourceCount;
    const actualNumberOfResourcesToCreate = Math.min(
      this.spareResourceCapacity,
      resourceShortfall
    );
    const resourceCreationPromiseList = [];
    for (let i = 0; actualNumberOfResourcesToCreate > i; i++) {
      resourceCreationPromiseList.push(this._createResource());
    }
    await Promise.all(resourceCreationPromiseList);
    if (this._config.testOnBorrow) {
      const desiredNumberOfResourcesToMoveIntoTest = numWaitingClients - this._testOnBorrowResources.size;
      const actualNumberOfResourcesToMoveIntoTest = Math.min(
        this._availableObjects.length,
        desiredNumberOfResourcesToMoveIntoTest
      );
      for (let i = 0; actualNumberOfResourcesToMoveIntoTest > i; i++) {
        this._testOnBorrow();
      }
    }
    if (!this._config.testOnBorrow) {
      const actualNumberOfResourcesToDispatch = Math.min(
        this._availableObjects.length,
        numWaitingClients
      );
      for (let i = 0; actualNumberOfResourcesToDispatch > i; i++) {
        this._dispatchResource();
      }
    }
  }
  _dispatchPooledResourceToNextWaitingClient(pooledResource) {
    const clientResourceRequest = this._waitingClientsQueue.dequeue();
    if (clientResourceRequest === null || clientResourceRequest.state !== import_Deferred.Deferred.PENDING) {
      this._addPooledResourceToAvailableObjects(pooledResource);
      return false;
    }
    const loan = new import_ResourceLoan.ResourceLoan(pooledResource, this.promiseConstructor);
    this._resourceLoans.set(pooledResource.obj, loan);
    pooledResource.allocate();
    clientResourceRequest.resolve(pooledResource.obj);
    return true;
  }
  // not in use for now
  _trackOperation(operation, set) {
    set.add(operation);
    return operation.then(
      (v) => {
        set.delete(operation);
        return this.promiseConstructor.resolve(v);
      },
      (e) => {
        set.delete(operation);
        return this.promiseConstructor.reject(e);
      }
    );
  }
  async _createResource() {
    const factoryPromise = this._factory.create();
    try {
      this._factoryCreateOperations.add(factoryPromise);
      const resource = await factoryPromise;
      const pooledResource = new import_PooledResource.PooledResource(resource);
      this._allObjects.add(pooledResource);
      this._addPooledResourceToAvailableObjects(pooledResource);
      this._factoryCreateOperations.delete(factoryPromise);
      await this._dispense();
    } catch (error) {
      await this._dispense();
      this._factoryCreateOperations.delete(factoryPromise);
      throw error;
    }
  }
  async _ensureMinimum() {
    if (this._draining === true) {
      return;
    }
    const resourceCreationPromiseList = [];
    const minShortfall = this._config.min - this._count;
    for (let i = 0; i < minShortfall; i++) {
      resourceCreationPromiseList.push(this._createResource());
    }
    await Promise.all(resourceCreationPromiseList);
  }
  // not in use for now
  _evict() {
    const testsToRun = Math.min(
      this._config.numTestsPerEvictionRun,
      this._availableObjects.length
    );
    const evictionConfig = {
      softIdleTimeoutMillis: this._config.softIdleTimeoutMillis,
      idleTimeoutMillis: this._config.idleTimeoutMillis,
      min: this._config.min
    };
    for (let testsHaveRun = 0; testsHaveRun < testsToRun; ) {
      const iterationResult = this._evictionIterator.next();
      if (iterationResult.done === true && this._availableObjects.length < 1) {
        this._evictionIterator.reset();
        return;
      }
      if (iterationResult.done === true && this._availableObjects.length > 0) {
        this._evictionIterator.reset();
        continue;
      }
      const resource = iterationResult.value;
      const shouldEvict = this._evictor.evict(
        evictionConfig,
        resource,
        this._availableObjects.length
      );
      testsHaveRun++;
      if (shouldEvict === true) {
        this._evictionIterator.remove();
        this._destroy(resource);
      }
    }
  }
  // not in use because this._config.evictionRunIntervalMillis default equals 0
  _scheduleEvictorRun() {
    if (this._config.evictionRunIntervalMillis > 0) {
      this._scheduledEviction = setTimeout(() => {
        this._evict();
        this._scheduleEvictorRun();
      }, this._config.evictionRunIntervalMillis).unref();
    }
  }
  // not in use for now
  _descheduleEvictorRun() {
    if (this._scheduledEviction) {
      clearTimeout(this._scheduledEviction);
    }
    this._scheduledEviction = null;
  }
  async start() {
    if (this._draining === true || this._started === true) return;
    this._started = true;
    try {
      this._scheduleEvictorRun();
      await this._ensureMinimum();
    } catch (error) {
      this._started = false;
      throw error;
    }
  }
  async acquire(priority) {
    if (this._started === false) {
      await this.start();
    }
    if (this._draining) {
      throw new Error("pool is draining and cannot accept work");
    }
    if (this.spareResourceCapacity < 1 && this._availableObjects.length < 1 && this._config.maxWaitingClients !== void 0 && this._waitingClientsQueue.length >= this._config.maxWaitingClients) {
      throw new Error("max waitingClients count exceeded");
    }
    const resourceRequest = new import_ResourceRequest.ResourceRequest(
      this._config.acquireTimeoutMillis,
      this.promiseConstructor
    );
    this._waitingClientsQueue.enqueue(resourceRequest, priority);
    try {
      await this._dispense();
    } catch (error) {
      if (resourceRequest.state === import_Deferred.Deferred.PENDING) {
        resourceRequest.reject(error);
      }
      throw error;
    }
    return resourceRequest.promise;
  }
  // not in use for now
  async use(fn, priority) {
    const resource_1 = await this.acquire(priority);
    return await fn(resource_1).then(
      (result) => {
        this.release(resource_1);
        return result;
      },
      (err) => {
        this.destroy(resource_1);
        throw err;
      }
    );
  }
  // not in use for now
  isBorrowedResource(resource) {
    return this._resourceLoans.has(resource);
  }
  async release(resource) {
    const loan = this._resourceLoans.get(resource);
    if (loan === void 0) {
      throw new Error("Resource not currently part of this pool");
    }
    this._resourceLoans.delete(resource);
    loan.resolve(resource);
    const pooledResource = loan.pooledResource;
    pooledResource.deallocate();
    this._addPooledResourceToAvailableObjects(pooledResource);
    await this._dispense();
  }
  async destroy(resource) {
    const loan = this._resourceLoans.get(resource);
    if (loan === void 0) {
      throw new Error("Resource not currently part of this pool");
    }
    this._resourceLoans.delete(resource);
    loan.resolve(resource);
    const pooledResource = loan.pooledResource;
    pooledResource.deallocate();
    if (this._count - 1 >= this.min) {
      await this._destroy(pooledResource);
    } else {
      this._addPooledResourceToAvailableObjects(pooledResource);
    }
    await this._dispense();
  }
  _addPooledResourceToAvailableObjects(pooledResource) {
    pooledResource.idle();
    if (this._config.fifo === true) {
      this._availableObjects.push(pooledResource);
    } else {
      this._availableObjects.unshift(pooledResource);
    }
  }
  // not in use for now
  async drain() {
    this._draining = true;
    await this.__allResourceRequestsSettled();
    await this.__allResourcesReturned();
    this._descheduleEvictorRun();
    this._draining = false;
  }
  // not in use for now
  __allResourceRequestsSettled() {
    if (this._waitingClientsQueue.length > 0) {
      return (0, import_utils.reflector)(this._waitingClientsQueue.tail.promise);
    }
    return Promise.resolve();
  }
  // not in use due to not using drain method
  __allResourcesReturned() {
    const ps = [...this._resourceLoans.values()].map((loan) => loan.promise).map((element) => (0, import_utils.reflector)(element));
    return Promise.all(ps);
  }
  // not in use for now
  async clear(ensureMinimum = true) {
    const reflectedCreatePromises = [...this._factoryCreateOperations].map((element) => (0, import_utils.reflector)(element));
    await Promise.all(reflectedCreatePromises);
    for (const resource of this._availableObjects) {
      await this._destroy(resource, ensureMinimum);
    }
    const reflectedDestroyPromises = [...this._factoryDestroyOperations].map((element_1) => (0, import_utils.reflector)(element_1));
    return await (0, import_utils.reflector)(Promise.all(reflectedDestroyPromises));
  }
  async end() {
    await this.drain();
    await this.clear(false);
    this._started = false;
  }
  // not in use for now
  ready() {
    return new Promise((resolve) => {
      const isReady = () => {
        if (this.available >= this.min) {
          resolve();
        } else {
          setTimeout(isReady, 100);
        }
      };
      isReady();
    });
  }
  get _potentiallyAllocableResourceCount() {
    return this._availableObjects.length + this._testOnBorrowResources.size + this._testOnReturnResources.size + this._factoryCreateOperations.size;
  }
  get _count() {
    return this._allObjects.size + this._factoryCreateOperations.size;
  }
  get spareResourceCapacity() {
    return this._config.max - (this._allObjects.size + this._factoryCreateOperations.size);
  }
  get size() {
    return this._count;
  }
  get available() {
    return this._availableObjects.length;
  }
  get borrowed() {
    return this._resourceLoans.size;
  }
  get pending() {
    return this._waitingClientsQueue.length;
  }
  get max() {
    return this._config.max;
  }
  get min() {
    return this._config.min;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Pool
});
