// waddler/pool-ts/PoolDefaults.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PoolDefaults_exports = {};
__export(PoolDefaults_exports, {
  PoolDefaults: () => PoolDefaults
});
module.exports = __toCommonJS(PoolDefaults_exports);
class PoolDefaults {
  fifo;
  priorityRange;
  testOnBorrow;
  testOnReturn;
  evictionRunIntervalMillis;
  numTestsPerEvictionRun;
  softIdleTimeoutMillis;
  idleTimeoutMillis;
  acquireTimeoutMillis;
  destroyTimeoutMillis;
  maxWaitingClients;
  min;
  max;
  promiseConstructor;
  constructor() {
    this.fifo = true;
    this.priorityRange = 1;
    this.testOnBorrow = false;
    this.testOnReturn = false;
    this.evictionRunIntervalMillis = 0;
    this.numTestsPerEvictionRun = 3;
    this.softIdleTimeoutMillis = -1;
    this.idleTimeoutMillis = 3e4;
    this.acquireTimeoutMillis = void 0;
    this.destroyTimeoutMillis = void 0;
    this.maxWaitingClients = void 0;
    this.min = 0;
    this.max = void 0;
    this.promiseConstructor = Promise;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PoolDefaults
});
