// waddler/pool-ts/PoolOptions.js
import { PoolDefaults } from "./PoolDefaults.js";
class PoolOptions {
  max;
  min;
  maxWaitingClients;
  testOnBorrow;
  testOnReturn;
  acquireTimeoutMillis;
  destroyTimeoutMillis;
  priorityRange;
  fifo;
  evictionRunIntervalMillis;
  numTestsPerEvictionRun;
  softIdleTimeoutMillis;
  idleTimeoutMillis;
  promiseConstructor;
  constructor(opts = {}) {
    const poolDefaults = new PoolDefaults();
    this.fifo = opts.fifo ?? poolDefaults.fifo;
    this.priorityRange = opts.priorityRange || poolDefaults.priorityRange;
    this.testOnBorrow = opts.testOnBorrow ?? poolDefaults.testOnBorrow;
    this.testOnReturn = opts.testOnReturn ?? poolDefaults.testOnReturn;
    this.acquireTimeoutMillis = opts.acquireTimeoutMillis === void 0 ? void 0 : Number.parseInt(opts.acquireTimeoutMillis.toString(), 10);
    this.destroyTimeoutMillis = opts.destroyTimeoutMillis ? Number.parseInt(opts.destroyTimeoutMillis.toString(), 10) : void 0;
    this.maxWaitingClients = opts.maxWaitingClients === void 0 ? void 0 : Number.parseInt(opts.maxWaitingClients.toString(), 10);
    this.max = Number.parseInt(opts.max?.toString() || "1", 10);
    this.min = Number.parseInt(opts.min?.toString() || "1", 10);
    this.max = Math.max(Number.isNaN(this.max) ? 1 : this.max, 1);
    this.min = Math.min(Number.isNaN(this.min) ? 1 : this.min, this.max);
    this.evictionRunIntervalMillis = opts.evictionRunIntervalMillis || poolDefaults.evictionRunIntervalMillis;
    this.numTestsPerEvictionRun = opts.numTestsPerEvictionRun || poolDefaults.numTestsPerEvictionRun;
    this.softIdleTimeoutMillis = opts.softIdleTimeoutMillis || poolDefaults.softIdleTimeoutMillis;
    this.idleTimeoutMillis = opts.idleTimeoutMillis || poolDefaults.idleTimeoutMillis;
    this.promiseConstructor = opts.promiseConstructor === void 0 ? poolDefaults.promiseConstructor : opts.promiseConstructor;
  }
}
export {
  PoolOptions
};
