// waddler/pool-ts/PoolOptions.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PoolOptions_exports = {};
__export(PoolOptions_exports, {
  PoolOptions: () => PoolOptions
});
module.exports = __toCommonJS(PoolOptions_exports);
var import_PoolDefaults = require("./PoolDefaults.cjs");
class PoolOptions {
  max;
  min;
  maxWaitingClients;
  testOnBorrow;
  testOnReturn;
  acquireTimeoutMillis;
  destroyTimeoutMillis;
  priorityRange;
  fifo;
  evictionRunIntervalMillis;
  numTestsPerEvictionRun;
  softIdleTimeoutMillis;
  idleTimeoutMillis;
  promiseConstructor;
  constructor(opts = {}) {
    const poolDefaults = new import_PoolDefaults.PoolDefaults();
    this.fifo = opts.fifo ?? poolDefaults.fifo;
    this.priorityRange = opts.priorityRange || poolDefaults.priorityRange;
    this.testOnBorrow = opts.testOnBorrow ?? poolDefaults.testOnBorrow;
    this.testOnReturn = opts.testOnReturn ?? poolDefaults.testOnReturn;
    this.acquireTimeoutMillis = opts.acquireTimeoutMillis === void 0 ? void 0 : Number.parseInt(opts.acquireTimeoutMillis.toString(), 10);
    this.destroyTimeoutMillis = opts.destroyTimeoutMillis ? Number.parseInt(opts.destroyTimeoutMillis.toString(), 10) : void 0;
    this.maxWaitingClients = opts.maxWaitingClients === void 0 ? void 0 : Number.parseInt(opts.maxWaitingClients.toString(), 10);
    this.max = Number.parseInt(opts.max?.toString() || "1", 10);
    this.min = Number.parseInt(opts.min?.toString() || "1", 10);
    this.max = Math.max(Number.isNaN(this.max) ? 1 : this.max, 1);
    this.min = Math.min(Number.isNaN(this.min) ? 1 : this.min, this.max);
    this.evictionRunIntervalMillis = opts.evictionRunIntervalMillis || poolDefaults.evictionRunIntervalMillis;
    this.numTestsPerEvictionRun = opts.numTestsPerEvictionRun || poolDefaults.numTestsPerEvictionRun;
    this.softIdleTimeoutMillis = opts.softIdleTimeoutMillis || poolDefaults.softIdleTimeoutMillis;
    this.idleTimeoutMillis = opts.idleTimeoutMillis || poolDefaults.idleTimeoutMillis;
    this.promiseConstructor = opts.promiseConstructor === void 0 ? poolDefaults.promiseConstructor : opts.promiseConstructor;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PoolOptions
});
