// waddler/pool-ts/PooledResource.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PooledResource_exports = {};
__export(PooledResource_exports, {
  PooledResource: () => PooledResource
});
module.exports = __toCommonJS(PooledResource_exports);
var PooledResourceState = /* @__PURE__ */ ((PooledResourceState2) => {
  PooledResourceState2["ALLOCATED"] = "ALLOCATED";
  PooledResourceState2["IDLE"] = "IDLE";
  PooledResourceState2["INVALID"] = "INVALID";
  PooledResourceState2["RETURNING"] = "RETURNING";
  PooledResourceState2["VALIDATION"] = "VALIDATION";
  return PooledResourceState2;
})(PooledResourceState || {});
class PooledResource {
  creationTime;
  lastReturnTime;
  lastBorrowTime;
  lastIdleTime;
  obj;
  state;
  constructor(resource) {
    this.creationTime = Date.now();
    this.lastReturnTime = null;
    this.lastBorrowTime = null;
    this.lastIdleTime = null;
    this.obj = resource;
    this.state = "IDLE" /* IDLE */;
  }
  // mark the resource as "allocated"
  allocate() {
    this.lastBorrowTime = Date.now();
    this.state = "ALLOCATED" /* ALLOCATED */;
  }
  // mark the resource as "deallocated"
  deallocate() {
    this.lastReturnTime = Date.now();
    this.state = "IDLE" /* IDLE */;
  }
  invalidate() {
    this.state = "INVALID" /* INVALID */;
  }
  test() {
    this.state = "VALIDATION" /* VALIDATION */;
  }
  idle() {
    this.lastIdleTime = Date.now();
    this.state = "IDLE" /* IDLE */;
  }
  returning() {
    this.state = "RETURNING" /* RETURNING */;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PooledResource
});
