// waddler/pool-ts/PriorityQueue.js
import { Queue } from "./Queue.js";
class PriorityQueue {
  _size;
  _slots;
  constructor(size) {
    this._size = Math.max(Math.trunc(+size), 1);
    this._slots = [];
    for (let i = 0; i < this._size; i++) {
      this._slots.push(new Queue());
    }
  }
  get length() {
    let _length = 0;
    for (let i = 0, sl = this._slots.length; i < sl; i++) {
      _length += this._slots[i].length;
    }
    return _length;
  }
  enqueue(obj, priority) {
    priority = priority && Math.trunc(+priority) || 0;
    if (priority < 0 || priority >= this._size) {
      priority = this._size - 1;
    }
    this._slots[priority].push(obj);
  }
  dequeue() {
    for (let i = 0, sl = this._slots.length; i < sl; i++) {
      if (this._slots[i].length > 0) {
        return this._slots[i].shift();
      }
    }
    return null;
  }
  get head() {
    for (let i = 0, sl = this._slots.length; i < sl; i++) {
      if (this._slots[i].length > 0) {
        return this._slots[i].head;
      }
    }
    return null;
  }
  get tail() {
    for (let i = this._slots.length - 1; i >= 0; i--) {
      if (this._slots[i].length > 0) {
        return this._slots[i].tail;
      }
    }
    return null;
  }
}
export {
  PriorityQueue
};
