// waddler/pool-ts/PriorityQueue.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PriorityQueue_exports = {};
__export(PriorityQueue_exports, {
  PriorityQueue: () => PriorityQueue
});
module.exports = __toCommonJS(PriorityQueue_exports);
var import_Queue = require("./Queue.cjs");
class PriorityQueue {
  _size;
  _slots;
  constructor(size) {
    this._size = Math.max(Math.trunc(+size), 1);
    this._slots = [];
    for (let i = 0; i < this._size; i++) {
      this._slots.push(new import_Queue.Queue());
    }
  }
  get length() {
    let _length = 0;
    for (let i = 0, sl = this._slots.length; i < sl; i++) {
      _length += this._slots[i].length;
    }
    return _length;
  }
  enqueue(obj, priority) {
    priority = priority && Math.trunc(+priority) || 0;
    if (priority < 0 || priority >= this._size) {
      priority = this._size - 1;
    }
    this._slots[priority].push(obj);
  }
  dequeue() {
    for (let i = 0, sl = this._slots.length; i < sl; i++) {
      if (this._slots[i].length > 0) {
        return this._slots[i].shift();
      }
    }
    return null;
  }
  get head() {
    for (let i = 0, sl = this._slots.length; i < sl; i++) {
      if (this._slots[i].length > 0) {
        return this._slots[i].head;
      }
    }
    return null;
  }
  get tail() {
    for (let i = this._slots.length - 1; i >= 0; i--) {
      if (this._slots[i].length > 0) {
        return this._slots[i].tail;
      }
    }
    return null;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PriorityQueue
});
