// waddler/pool-ts/ResourceRequest.js
import { Deferred } from "./Deferred.js";
import { TimeoutError } from "./errors.js";
function fbind(fn, ctx) {
  return function bound() {
    return fn.apply(ctx);
  };
}
class ResourceRequest extends Deferred {
  _creationTimestamp;
  _timeout;
  /**
   * [constructor description]
   * @param  {number} ttl timeout in milliseconds
   * @param  {PromiseConstructor} promiseConstructor promise implementation
   */
  constructor(ttl, promiseConstructor) {
    super(promiseConstructor);
    this._creationTimestamp = Date.now();
    this._timeout = null;
    if (ttl !== void 0) {
      this.setTimeout(ttl);
    }
  }
  setTimeout(delay) {
    if (this._state !== ResourceRequest.PENDING) {
      return;
    }
    const ttl = Number.parseInt(delay.toString(), 10);
    if (Number.isNaN(ttl) || ttl <= 0) {
      throw new Error("delay must be a positive int");
    }
    const age = Date.now() - this._creationTimestamp;
    if (this._timeout) {
      this.removeTimeout();
    }
    this._timeout = setTimeout(
      fbind(this._fireTimeout, this),
      Math.max(ttl - age, 0)
    );
  }
  removeTimeout() {
    if (this._timeout) {
      clearTimeout(this._timeout);
    }
    this._timeout = null;
  }
  _fireTimeout() {
    this.reject(new TimeoutError("ResourceRequest timed out"));
  }
  reject(reason) {
    this.removeTimeout();
    super.reject(reason);
  }
  resolve(value) {
    this.removeTimeout();
    super.resolve(value);
  }
}
export {
  ResourceRequest
};
