// waddler/pool-ts/ResourceRequest.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ResourceRequest_exports = {};
__export(ResourceRequest_exports, {
  ResourceRequest: () => ResourceRequest
});
module.exports = __toCommonJS(ResourceRequest_exports);
var import_Deferred = require("./Deferred.cjs");
var import_errors = require("./errors.cjs");
function fbind(fn, ctx) {
  return function bound() {
    return fn.apply(ctx);
  };
}
class ResourceRequest extends import_Deferred.Deferred {
  _creationTimestamp;
  _timeout;
  /**
   * [constructor description]
   * @param  {number} ttl timeout in milliseconds
   * @param  {PromiseConstructor} promiseConstructor promise implementation
   */
  constructor(ttl, promiseConstructor) {
    super(promiseConstructor);
    this._creationTimestamp = Date.now();
    this._timeout = null;
    if (ttl !== void 0) {
      this.setTimeout(ttl);
    }
  }
  setTimeout(delay) {
    if (this._state !== ResourceRequest.PENDING) {
      return;
    }
    const ttl = Number.parseInt(delay.toString(), 10);
    if (Number.isNaN(ttl) || ttl <= 0) {
      throw new Error("delay must be a positive int");
    }
    const age = Date.now() - this._creationTimestamp;
    if (this._timeout) {
      this.removeTimeout();
    }
    this._timeout = setTimeout(
      fbind(this._fireTimeout, this),
      Math.max(ttl - age, 0)
    );
  }
  removeTimeout() {
    if (this._timeout) {
      clearTimeout(this._timeout);
    }
    this._timeout = null;
  }
  _fireTimeout() {
    this.reject(new import_errors.TimeoutError("ResourceRequest timed out"));
  }
  reject(reason) {
    this.removeTimeout();
    super.reject(reason);
  }
  resolve(value) {
    this.removeTimeout();
    super.resolve(value);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ResourceRequest
});
