// @netlify/zip-it-and-ship-it/dist/utils/format_result.js
import { removeUndefined } from './remove_undefined.js';
// Takes the result of zipping a function and formats it for output.
export const formatZipResult = (archive) => {
    const memory = archive.staticAnalysisResult?.config?.memory ?? archive?.config?.memory;
    const vcpu = archive.staticAnalysisResult?.config?.vcpu ?? archive?.config?.vcpu;
    const functionResult = {
        ...archive,
        staticAnalysisResult: undefined,
        eventSubscriptions: archive.staticAnalysisResult?.eventSubscriptions,
        routes: archive.staticAnalysisResult?.routes,
        excludedRoutes: archive.staticAnalysisResult?.excludedRoutes,
        runtime: archive.runtime.name,
        memory,
        region: archive.staticAnalysisResult?.config?.region ?? archive?.config?.region,
        schedule: archive.staticAnalysisResult?.config?.schedule ?? archive?.config?.schedule,
        runtimeAPIVersion: archive.staticAnalysisResult?.runtimeAPIVersion,
        vcpu,
    };
    return removeUndefined(functionResult);
};
