// waddler/sqlite/sqlite-core/dialect.js
import { Dialect, SQLDefault, SQLIdentifier, SQLRaw, SQLValues } from "../../sql-template-params.js";
class SqliteDialect extends Dialect {
  escapeParam() {
    return `?`;
  }
  escapeIdentifier(identifier) {
    return `"${identifier}"`;
  }
  checkIdentifierObject(object) {
    if (Object.values(object).includes(void 0)) {
      throw new Error(
        `you can't specify undefined parameters. maybe you want to omit it?`
      );
    }
    if (Object.keys(object).length === 0) {
      throw new Error(`you need to specify at least one parameter.`);
    }
    if (Object.keys(object).length === 1 && object.as !== void 0) {
      throw new Error(
        `you can't specify only "as" property. you have to specify "column" or "table" property along with "as".`
      );
    }
    if (!["string", "undefined"].includes(typeof object.table) || !["string", "undefined"].includes(typeof object.column) || !["string", "undefined"].includes(typeof object.as)) {
      throw new Error(
        "object properties 'table', 'column', 'as' should be of string type or omitted."
      );
    }
  }
  valueToSQL({ value, params }) {
    if (value instanceof SQLDefault) {
      return value.generateSQL().sql;
    }
    if (typeof value === "object" && value.constructor.name === "Object") {
      params.push(JSON.stringify(value));
      return this.escapeParam();
    }
    params.push(value);
    return this.escapeParam();
  }
}
class UnsafePromise {
  constructor(driver) {
    this.driver = driver;
  }
  run() {
    this.driver.run();
    return this;
  }
  all() {
    this.driver.all();
    return this;
  }
  stream() {
    return this.driver.stream();
  }
  // Allow it to be awaited (like a Promise)
  then(onfulfilled, onrejected) {
    const result = this.driver.execute();
    return Promise.resolve(result).then(onfulfilled, onrejected);
  }
}
const SQLFunctions = {
  identifier: (value) => {
    return new SQLIdentifier(value);
  },
  values: (value) => {
    return new SQLValues(value);
  },
  raw: (value) => {
    return new SQLRaw(value);
  }
};
export {
  SQLFunctions,
  SqliteDialect,
  UnsafePromise
};
