// waddler/sqlite/sqlite-core/dialect.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dialect_exports = {};
__export(dialect_exports, {
  SQLFunctions: () => SQLFunctions,
  SqliteDialect: () => SqliteDialect,
  UnsafePromise: () => UnsafePromise
});
module.exports = __toCommonJS(dialect_exports);
var import_sql_template_params = require("../../sql-template-params.cjs");
class SqliteDialect extends import_sql_template_params.Dialect {
  escapeParam() {
    return `?`;
  }
  escapeIdentifier(identifier) {
    return `"${identifier}"`;
  }
  checkIdentifierObject(object) {
    if (Object.values(object).includes(void 0)) {
      throw new Error(
        `you can't specify undefined parameters. maybe you want to omit it?`
      );
    }
    if (Object.keys(object).length === 0) {
      throw new Error(`you need to specify at least one parameter.`);
    }
    if (Object.keys(object).length === 1 && object.as !== void 0) {
      throw new Error(
        `you can't specify only "as" property. you have to specify "column" or "table" property along with "as".`
      );
    }
    if (!["string", "undefined"].includes(typeof object.table) || !["string", "undefined"].includes(typeof object.column) || !["string", "undefined"].includes(typeof object.as)) {
      throw new Error(
        "object properties 'table', 'column', 'as' should be of string type or omitted."
      );
    }
  }
  valueToSQL({ value, params }) {
    if (value instanceof import_sql_template_params.SQLDefault) {
      return value.generateSQL().sql;
    }
    if (typeof value === "object" && value.constructor.name === "Object") {
      params.push(JSON.stringify(value));
      return this.escapeParam();
    }
    params.push(value);
    return this.escapeParam();
  }
}
class UnsafePromise {
  constructor(driver) {
    this.driver = driver;
  }
  run() {
    this.driver.run();
    return this;
  }
  all() {
    this.driver.all();
    return this;
  }
  stream() {
    return this.driver.stream();
  }
  // Allow it to be awaited (like a Promise)
  then(onfulfilled, onrejected) {
    const result = this.driver.execute();
    return Promise.resolve(result).then(onfulfilled, onrejected);
  }
}
const SQLFunctions = {
  identifier: (value) => {
    return new import_sql_template_params.SQLIdentifier(value);
  },
  values: (value) => {
    return new import_sql_template_params.SQLValues(value);
  },
  raw: (value) => {
    return new import_sql_template_params.SQLRaw(value);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SQLFunctions,
  SqliteDialect,
  UnsafePromise
});
