// drizzle-orm/node-postgres/session.js
import { entityKind } from "../entity.js";
import { sql } from "../sql/sql.js";
import { NoopLogger } from "../logger.js";
import { NoopCache } from "../cache/core/index.js";
import { PgAsyncPreparedQuery, PgAsyncSession, PgAsyncTransaction } from "../pg-core/async/session.js";
import pg from "pg";
import { preparedStatementName } from "../query-name-generator.js";

//#region src/node-postgres/session.ts
const { Pool: Pool$1, types } = pg;
const noop = (val) => val;
const typeConfig = { getTypeParser: ((typeId, format) => {
	switch (typeId) {
		case types.builtins.TIMESTAMPTZ:
		case types.builtins.TIMESTAMP:
		case types.builtins.DATE:
		case types.builtins.INTERVAL:
		case 1231:
		case 1115:
		case 1185:
		case 1187:
		case 1182: return noop;
		default: return types.getTypeParser(typeId, format);
	}
}) };
var NodePgSession = class NodePgSession extends PgAsyncSession {
	static [entityKind] = "NodePgSession";
	logger;
	cache;
	constructor(client, dialect, relations, options = {}) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
		this.logger = options.logger ?? new NoopLogger();
		this.cache = options.cache ?? new NoopCache();
	}
	prepareQuery(query, mode, name, mapper, queryMetadata, cacheConfig) {
		const queryName = typeof name === "string" ? name : name === true ? preparedStatementName(query.sql, query.params) : void 0;
		const executor = async (params) => {
			return this.client.query({
				name: queryName,
				rowMode: mode === "arrays" ? "array" : void 0,
				text: query.sql,
				types: typeConfig
			}, params).then((r) => mode === "raw" ? r : r.rows);
		};
		return new PgAsyncPreparedQuery(executor, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	async transaction(transaction, config) {
		const isPool = this.client instanceof Pool$1 || Object.getPrototypeOf(this.client).constructor.name.includes("Pool");
		const session = isPool ? new NodePgSession(await this.client.connect(), this.dialect, this.relations, this.options) : this;
		const tx = new NodePgTransaction(this.dialect, session, this.relations, void 0, false);
		await tx.execute(sql`begin${config ? sql` ${tx.getTransactionConfigSQL(config)}` : void 0}`);
		try {
			const result = await transaction(tx);
			await tx.execute(sql`commit`);
			return result;
		} catch (error) {
			await tx.execute(sql`rollback`);
			throw error;
		} finally {
			if (isPool) session.client.release();
		}
	}
};
var NodePgTransaction = class NodePgTransaction extends PgAsyncTransaction {
	static [entityKind] = "NodePgTransaction";
	async transaction(transaction) {
		const savepointName = `sp${this.nestedIndex + 1}`;
		const tx = new NodePgTransaction(this.dialect, this.session, this._.relations, this.nestedIndex + 1, false);
		await tx.execute(sql.raw(`savepoint ${savepointName}`));
		try {
			const result = await transaction(tx);
			await tx.execute(sql.raw(`release savepoint ${savepointName}`));
			return result;
		} catch (err) {
			await tx.execute(sql.raw(`rollback to savepoint ${savepointName}`));
			throw err;
		}
	}
};

//#endregion
export { NodePgSession, NodePgTransaction };
