// waddler/clickhouse-core/dialect.js
import { TupleParam } from "@clickhouse/client";
import { Dialect, SQLCommonParam, SQLDefault, SQLIdentifier, SQLRaw, SQLValues } from "../sql-template-params.js";
import { getArrayDepth, makeClickHouseArray } from "./utils.js";
class ClickHouseDialect extends Dialect {
  escapeParam(lastParamIdx, typeToCast) {
    return `{param${lastParamIdx}:${typeToCast || "String"}}`;
  }
  createEmptyParams() {
    return {};
  }
  pushParams(params, param, lastParamIdx, mode = "bulk") {
    if (mode === "bulk") Object.assign(params, param);
    else params[`param${lastParamIdx}`] = param;
  }
  escapeIdentifier(identifier) {
    return `\`${identifier}\``;
  }
  checkIdentifierObject(object) {
    if (Object.values(object).includes(void 0)) {
      throw new Error(
        `you can't specify undefined parameters. maybe you want to omit it?`
      );
    }
    if (Object.keys(object).length === 0) {
      throw new Error(`you need to specify at least one parameter.`);
    }
    if (object.schema !== void 0 && object.table === void 0 && object.column !== void 0) {
      throw new Error(
        `you can't specify only "schema" and "column" properties, you need also specify "table".`
      );
    }
    if (Object.keys(object).length === 1 && object.as !== void 0) {
      throw new Error(`you can't specify only "as" property.`);
    }
    if (object.as !== void 0 && object.column === void 0 && object.table === void 0) {
      throw new Error(
        `you have to specify "column" or "table" property along with "as".`
      );
    }
    if (!["string", "undefined"].includes(typeof object.schema) || !["string", "undefined"].includes(typeof object.table) || !["string", "undefined"].includes(typeof object.column) || !["string", "undefined"].includes(typeof object.as)) {
      throw new Error(
        "object properties 'schema', 'table', 'column', 'as' should be of string type or omitted."
      );
    }
  }
  valueToSQL({ value, lastParamIdx, params, types, colIdx, paramsCount }) {
    if (value instanceof SQLDefault) {
      return value.generateSQL().sql;
    }
    if (value instanceof SQLRaw) {
      return value.generateSQL().sql;
    }
    if (typeof value === "bigint") {
      this.pushParams(params, `${value}`, lastParamIdx + paramsCount + 1, "single");
      return this.escapeParam(lastParamIdx + paramsCount + 1, types[colIdx]);
    }
    if (Array.isArray(value)) {
      const { mappedArray: mappedValue, baseTypeToCast } = makeClickHouseArray(value, types[colIdx]);
      let arrayTypeToCast;
      if (baseTypeToCast !== void 0) {
        const arrayDepth = getArrayDepth(value);
        arrayTypeToCast = baseTypeToCast;
        for (let i = 0; i < arrayDepth; i++) arrayTypeToCast = `Array(${arrayTypeToCast})`;
      }
      this.pushParams(params, mappedValue, lastParamIdx + paramsCount + 1, "single");
      return this.escapeParam(lastParamIdx + paramsCount + 1, types[colIdx] ?? arrayTypeToCast);
    }
    if (typeof value === "number" || typeof value === "boolean" || typeof value === "string" || value === null || value instanceof Date) {
      this.pushParams(params, value, lastParamIdx + paramsCount + 1, "single");
      return this.escapeParam(lastParamIdx + paramsCount + 1, types[colIdx]);
    }
    if (value instanceof Map || value instanceof TupleParam) {
      this.pushParams(params, value, lastParamIdx + paramsCount + 1, "single");
      return this.escapeParam(lastParamIdx + paramsCount + 1, types[colIdx]);
    }
    if (typeof value === "object") {
      this.pushParams(params, JSON.stringify(value), lastParamIdx + paramsCount + 1, "single");
      return this.escapeParam(lastParamIdx + paramsCount + 1, types[colIdx] ?? "JSON");
    }
    if (value === void 0) {
      throw new Error("value can't be undefined, maybe you mean sql.default?");
    }
    throw new Error(`you can't specify ${typeof value} as value.`);
  }
}
class UnsafePromise {
  constructor(driver) {
    this.driver = driver;
  }
  command() {
    this.driver.command();
    return this;
  }
  query() {
    this.driver.query();
    return this;
  }
  stream() {
    return this.driver.stream();
  }
  // Allow it to be awaited (like a Promise)
  then(onfulfilled, onrejected) {
    const result = this.driver.execute();
    return Promise.resolve(result).then(onfulfilled, onrejected);
  }
}
const SQLFunctions = {
  identifier: (value) => {
    return new SQLIdentifier(value);
  },
  values: (value, types) => {
    return new SQLValues(value, types);
  },
  param: (value, type) => {
    return new SQLCommonParam(value, type);
  },
  raw: (value) => {
    return new SQLRaw(value);
  },
  default: new SQLDefault()
};
export {
  ClickHouseDialect,
  SQLFunctions,
  UnsafePromise
};
