// waddler/clickhouse-core/dialect.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dialect_exports = {};
__export(dialect_exports, {
  ClickHouseDialect: () => ClickHouseDialect,
  SQLFunctions: () => SQLFunctions,
  UnsafePromise: () => UnsafePromise
});
module.exports = __toCommonJS(dialect_exports);
var import_client = require("@clickhouse/client");
var import_sql_template_params = require("../sql-template-params.cjs");
var import_utils = require("./utils.cjs");
class ClickHouseDialect extends import_sql_template_params.Dialect {
  escapeParam(lastParamIdx, typeToCast) {
    return `{param${lastParamIdx}:${typeToCast || "String"}}`;
  }
  createEmptyParams() {
    return {};
  }
  pushParams(params, param, lastParamIdx, mode = "bulk") {
    if (mode === "bulk") Object.assign(params, param);
    else params[`param${lastParamIdx}`] = param;
  }
  escapeIdentifier(identifier) {
    return `\`${identifier}\``;
  }
  checkIdentifierObject(object) {
    if (Object.values(object).includes(void 0)) {
      throw new Error(
        `you can't specify undefined parameters. maybe you want to omit it?`
      );
    }
    if (Object.keys(object).length === 0) {
      throw new Error(`you need to specify at least one parameter.`);
    }
    if (object.schema !== void 0 && object.table === void 0 && object.column !== void 0) {
      throw new Error(
        `you can't specify only "schema" and "column" properties, you need also specify "table".`
      );
    }
    if (Object.keys(object).length === 1 && object.as !== void 0) {
      throw new Error(`you can't specify only "as" property.`);
    }
    if (object.as !== void 0 && object.column === void 0 && object.table === void 0) {
      throw new Error(
        `you have to specify "column" or "table" property along with "as".`
      );
    }
    if (!["string", "undefined"].includes(typeof object.schema) || !["string", "undefined"].includes(typeof object.table) || !["string", "undefined"].includes(typeof object.column) || !["string", "undefined"].includes(typeof object.as)) {
      throw new Error(
        "object properties 'schema', 'table', 'column', 'as' should be of string type or omitted."
      );
    }
  }
  valueToSQL({ value, lastParamIdx, params, types, colIdx, paramsCount }) {
    if (value instanceof import_sql_template_params.SQLDefault) {
      return value.generateSQL().sql;
    }
    if (value instanceof import_sql_template_params.SQLRaw) {
      return value.generateSQL().sql;
    }
    if (typeof value === "bigint") {
      this.pushParams(params, `${value}`, lastParamIdx + paramsCount + 1, "single");
      return this.escapeParam(lastParamIdx + paramsCount + 1, types[colIdx]);
    }
    if (Array.isArray(value)) {
      const { mappedArray: mappedValue, baseTypeToCast } = (0, import_utils.makeClickHouseArray)(value, types[colIdx]);
      let arrayTypeToCast;
      if (baseTypeToCast !== void 0) {
        const arrayDepth = (0, import_utils.getArrayDepth)(value);
        arrayTypeToCast = baseTypeToCast;
        for (let i = 0; i < arrayDepth; i++) arrayTypeToCast = `Array(${arrayTypeToCast})`;
      }
      this.pushParams(params, mappedValue, lastParamIdx + paramsCount + 1, "single");
      return this.escapeParam(lastParamIdx + paramsCount + 1, types[colIdx] ?? arrayTypeToCast);
    }
    if (typeof value === "number" || typeof value === "boolean" || typeof value === "string" || value === null || value instanceof Date) {
      this.pushParams(params, value, lastParamIdx + paramsCount + 1, "single");
      return this.escapeParam(lastParamIdx + paramsCount + 1, types[colIdx]);
    }
    if (value instanceof Map || value instanceof import_client.TupleParam) {
      this.pushParams(params, value, lastParamIdx + paramsCount + 1, "single");
      return this.escapeParam(lastParamIdx + paramsCount + 1, types[colIdx]);
    }
    if (typeof value === "object") {
      this.pushParams(params, JSON.stringify(value), lastParamIdx + paramsCount + 1, "single");
      return this.escapeParam(lastParamIdx + paramsCount + 1, types[colIdx] ?? "JSON");
    }
    if (value === void 0) {
      throw new Error("value can't be undefined, maybe you mean sql.default?");
    }
    throw new Error(`you can't specify ${typeof value} as value.`);
  }
}
class UnsafePromise {
  constructor(driver) {
    this.driver = driver;
  }
  command() {
    this.driver.command();
    return this;
  }
  query() {
    this.driver.query();
    return this;
  }
  stream() {
    return this.driver.stream();
  }
  // Allow it to be awaited (like a Promise)
  then(onfulfilled, onrejected) {
    const result = this.driver.execute();
    return Promise.resolve(result).then(onfulfilled, onrejected);
  }
}
const SQLFunctions = {
  identifier: (value) => {
    return new import_sql_template_params.SQLIdentifier(value);
  },
  values: (value, types) => {
    return new import_sql_template_params.SQLValues(value, types);
  },
  param: (value, type) => {
    return new import_sql_template_params.SQLCommonParam(value, type);
  },
  raw: (value) => {
    return new import_sql_template_params.SQLRaw(value);
  },
  default: new import_sql_template_params.SQLDefault()
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ClickHouseDialect,
  SQLFunctions,
  UnsafePromise
});
