// drizzle-orm/effect-postgres/driver.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
const require_effect_postgres_codecs = require('./codecs.cjs');
const require_effect_postgres_session = require('./session.cjs');
let __entity_ts = require("../entity.cjs");
let __utils_ts = require("../utils.cjs");
let __pg_core_dialect_ts = require("../pg-core/dialect.cjs");
let effect_Effect = require("effect/Effect");
effect_Effect = require_runtime.__toESM(effect_Effect);
let __cache_core_cache_effect_ts = require("../cache/core/cache-effect.cjs");
let __effect_core_index_ts = require("../effect-core/index.cjs");
let _effect_sql_pg_PgClient = require("@effect/sql-pg/PgClient");
let effect_Layer = require("effect/Layer");
effect_Layer = require_runtime.__toESM(effect_Layer);
let __pg_core_effect_db_ts = require("../pg-core/effect/db.cjs");

//#region src/effect-postgres/driver.ts
var EffectPgDatabase = class extends __pg_core_effect_db_ts.PgEffectDatabase {
	static [__entity_ts.entityKind] = "EffectPgDatabase";
};
const DefaultServices = effect_Layer.merge(__cache_core_cache_effect_ts.EffectCache.Default, __effect_core_index_ts.EffectLogger.Default);
/**
* Creates an EffectPgDatabase instance.
*
* Requires `PgClient`, `EffectLogger`, and `EffectCache` services to be provided.
* Use `DefaultServices` to provide default (no-op) logger and cache implementations.
*
* @example
* ```ts
* // With default services (no logging, no caching)
* const db = yield* PgDrizzle.make({ relations }).pipe(
*   Effect.provide(PgDrizzle.DefaultServices),
* );
*
* // With Effect-based logging
* const db = yield* PgDrizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layer),
*   Effect.provide(PgDrizzle.DefaultServices),
* );
*
* // With custom Drizzle logger
* const db = yield* PgDrizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layerFromDrizzle(myLogger)),
*   Effect.provide(PgDrizzle.DefaultServices),
* );
* ```
*/
const make = effect_Effect.fn("PgDrizzle.make")(function* (config = {}) {
	const client = yield* _effect_sql_pg_PgClient.PgClient;
	const cache = yield* __cache_core_cache_effect_ts.EffectCache;
	const logger = yield* __effect_core_index_ts.EffectLogger;
	const dialect = new __pg_core_dialect_ts.PgDialect({
		useJitMappers: (0, __utils_ts.jitCompatCheck)(config.jit),
		codecs: config.codecs ?? require_effect_postgres_codecs.effectPgCodecs
	});
	const relations = config.relations ?? {};
	const db = new EffectPgDatabase(dialect, new require_effect_postgres_session.EffectPgSession(client, dialect, relations, {
		logger,
		cache,
		useJitMappers: (0, __utils_ts.jitCompatCheck)(config.jit)
	}), relations);
	db.$client = client;
	db.$cache = cache;
	if (db.$cache) db.$cache["invalidate"] = cache.onMutate;
	return db;
});
/**
* Convenience function that creates an EffectPgDatabase with `DefaultServices` already provided.
*/
const makeWithDefaults = (config = {}) => make(config).pipe(effect_Effect.provide(DefaultServices));

//#endregion
exports.DefaultServices = DefaultServices;
exports.EffectPgDatabase = EffectPgDatabase;
exports.make = make;
exports.makeWithDefaults = makeWithDefaults;
