// drizzle-orm/effect-postgres/session.js
import { entityKind } from "../entity.js";
import * as Effect from "effect/Effect";
import { PgEffectPreparedQuery, PgEffectSession, PgEffectTransaction } from "../pg-core/effect/session.js";

//#region src/effect-postgres/session.ts
var EffectPgSession = class extends PgEffectSession {
	static [entityKind] = "EffectPgSession";
	constructor(client, dialect, relations, options) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
	}
	prepareQuery(query, mode, _name, mapper, queryMetadata, cacheConfig) {
		const executor = (params) => {
			const q = this.client.unsafe(query.sql, params);
			if (mode === "arrays") return q.values;
			return q.withoutTransform;
		};
		return new PgEffectPreparedQuery(executor, query, mapper, mode, this.options.logger, this.options.cache, queryMetadata, cacheConfig);
	}
	transaction(transaction) {
		const { dialect, relations } = this;
		return this.client.withTransaction(Effect.gen({ self: this }, function* () {
			return yield* transaction(new EffectPgTransaction(dialect, this, relations));
		}));
	}
};
var EffectPgTransaction = class extends PgEffectTransaction {
	static [entityKind] = "EffectPgTransaction";
	transaction(transaction) {
		return this.session.transaction(transaction);
	}
};

//#endregion
export { EffectPgSession, EffectPgTransaction };
