// drizzle-orm/effect-postgres/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let effect_Effect = require("effect/Effect");
effect_Effect = require_runtime.__toESM(effect_Effect);
let __pg_core_effect_session_ts = require("../pg-core/effect/session.cjs");

//#region src/effect-postgres/session.ts
var EffectPgSession = class extends __pg_core_effect_session_ts.PgEffectSession {
	static [__entity_ts.entityKind] = "EffectPgSession";
	constructor(client, dialect, relations, options) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
	}
	prepareQuery(query, mode, _name, mapper, queryMetadata, cacheConfig) {
		const executor = (params) => {
			const q = this.client.unsafe(query.sql, params);
			if (mode === "arrays") return q.values;
			return q.withoutTransform;
		};
		return new __pg_core_effect_session_ts.PgEffectPreparedQuery(executor, query, mapper, mode, this.options.logger, this.options.cache, queryMetadata, cacheConfig);
	}
	transaction(transaction) {
		const { dialect, relations } = this;
		return this.client.withTransaction(effect_Effect.gen({ self: this }, function* () {
			return yield* transaction(new EffectPgTransaction(dialect, this, relations));
		}));
	}
};
var EffectPgTransaction = class extends __pg_core_effect_session_ts.PgEffectTransaction {
	static [__entity_ts.entityKind] = "EffectPgTransaction";
	transaction(transaction) {
		return this.session.transaction(transaction);
	}
};

//#endregion
exports.EffectPgSession = EffectPgSession;
exports.EffectPgTransaction = EffectPgTransaction;
