// @tanstack/start-plugin-core/dist/esm/import-protection-plugin/rewriteDeniedImports.js
import { getOrCreate } from "./utils.js";
import { MOCK_MODULE_ID } from "./virtualModules.js";
import { parseImportProtectionAst } from "./ast.js";
import * as t from "@babel/types";
import { generateFromAst } from "@tanstack/router-utils";
//#region src/import-protection-plugin/rewriteDeniedImports.ts
function isValidExportName(name) {
	if (name === "default" || name.length === 0) return false;
	const first = name.charCodeAt(0);
	if (!(first >= 65 && first <= 90 || first >= 97 && first <= 122 || first === 95 || first === 36)) return false;
	for (let i = 1; i < name.length; i++) {
		const ch = name.charCodeAt(i);
		if (!(ch >= 65 && ch <= 90 || ch >= 97 && ch <= 122 || ch >= 48 && ch <= 57 || ch === 95 || ch === 36)) return false;
	}
	return true;
}
function collectMockExportNamesBySource(code) {
	return collectMockExportNamesBySourceFromAst(parseImportProtectionAst(code));
}
function collectMockExportNamesBySourceFromAst(ast) {
	const namesBySource = /* @__PURE__ */ new Map();
	const memberBindingToSource = /* @__PURE__ */ new Map();
	const add = (source, name) => {
		if (name === "default" || name.length === 0) return;
		getOrCreate(namesBySource, source, () => /* @__PURE__ */ new Set()).add(name);
	};
	for (const node of ast.program.body) {
		if (t.isImportDeclaration(node)) {
			if (node.importKind === "type") continue;
			const source = node.source.value;
			for (const s of node.specifiers) {
				if (t.isImportNamespaceSpecifier(s)) {
					memberBindingToSource.set(s.local.name, source);
					continue;
				}
				if (t.isImportDefaultSpecifier(s)) {
					memberBindingToSource.set(s.local.name, source);
					continue;
				}
				if (!t.isImportSpecifier(s)) continue;
				if (s.importKind === "type") continue;
				const importedName = t.isIdentifier(s.imported) ? s.imported.name : s.imported.value;
				if (importedName === "default") continue;
				add(source, importedName);
			}
		}
		if (t.isExportNamedDeclaration(node) && node.source?.value) {
			if (node.exportKind === "type") continue;
			const source = node.source.value;
			for (const s of node.specifiers) {
				if (!t.isExportSpecifier(s)) continue;
				if (s.exportKind === "type") continue;
				add(source, s.local.name);
			}
		}
	}
	if (memberBindingToSource.size > 0) {
		const visit = (node) => {
			if (t.isMemberExpression(node)) {
				const object = node.object;
				if (t.isIdentifier(object)) {
					const source = memberBindingToSource.get(object.name);
					if (source) {
						const property = node.property;
						if (!node.computed && t.isIdentifier(property)) add(source, property.name);
						else if (node.computed && t.isStringLiteral(property)) add(source, property.value);
					}
				}
			}
			const keys = t.VISITOR_KEYS[node.type];
			if (!keys) return;
			for (const key of keys) {
				const child = node[key];
				if (Array.isArray(child)) {
					for (const item of child) if (item && typeof item === "object" && "type" in item) visit(item);
				} else if (child && typeof child === "object" && "type" in child) visit(child);
			}
		};
		visit(ast.program);
	}
	const out = /* @__PURE__ */ new Map();
	for (const [source, set] of namesBySource) out.set(source, Array.from(set).sort());
	return out;
}
/** Collect all valid named export identifiers from the given code. */
function collectNamedExports(code) {
	return collectNamedExportsFromAst(parseImportProtectionAst(code));
}
function collectIdentifiersFromPattern(pattern, add) {
	if (t.isIdentifier(pattern)) add(pattern.name);
	else if (t.isObjectPattern(pattern)) for (const prop of pattern.properties) if (t.isRestElement(prop)) collectIdentifiersFromPattern(prop.argument, add);
	else collectIdentifiersFromPattern(prop.value, add);
	else if (t.isArrayPattern(pattern)) {
		for (const elem of pattern.elements) if (elem) collectIdentifiersFromPattern(elem, add);
	} else if (t.isAssignmentPattern(pattern)) collectIdentifiersFromPattern(pattern.left, add);
	else if (t.isRestElement(pattern)) collectIdentifiersFromPattern(pattern.argument, add);
}
function collectNamedExportsFromAst(ast) {
	const names = /* @__PURE__ */ new Set();
	const add = (name) => {
		if (isValidExportName(name)) names.add(name);
	};
	for (const node of ast.program.body) if (t.isExportNamedDeclaration(node)) {
		if (node.exportKind === "type") continue;
		if (node.declaration) {
			const decl = node.declaration;
			if (t.isFunctionDeclaration(decl) || t.isClassDeclaration(decl)) {
				if (decl.id?.name) add(decl.id.name);
			} else if (t.isVariableDeclaration(decl)) for (const d of decl.declarations) collectIdentifiersFromPattern(d.id, add);
		}
		for (const s of node.specifiers) {
			if (!t.isExportSpecifier(s)) continue;
			if (s.exportKind === "type") continue;
			add(t.isIdentifier(s.exported) ? s.exported.name : s.exported.value);
		}
	}
	return Array.from(names).sort();
}
/**
* Rewrite static imports/re-exports from denied sources using Babel AST transforms.
*
* Transforms:
*   import { a as b, c } from 'denied'
* Into:
*   import __tss_deny_0 from 'tanstack-start-import-protection:mock'
*   const b = __tss_deny_0.a
*   const c = __tss_deny_0.c
*
* Also handles:
*   import def from 'denied'        -> import def from mock
*   import * as ns from 'denied'    -> import ns from mock
*   export { x } from 'denied'      -> export const x = mock.x
*   export * from 'denied'          -> removed
*   export { x as y } from 'denied' -> export const y = mock.x
*/
function rewriteDeniedImports(code, id, deniedSources, getMockModuleId = () => MOCK_MODULE_ID) {
	return rewriteDeniedImportsFromAst(parseImportProtectionAst(code), id, deniedSources, getMockModuleId);
}
function rewriteDeniedImportsFromAst(ast, id, deniedSources, getMockModuleId = () => MOCK_MODULE_ID) {
	let modified = false;
	let mockCounter = 0;
	for (let i = ast.program.body.length - 1; i >= 0; i--) {
		const node = ast.program.body[i];
		if (t.isImportDeclaration(node)) {
			if (node.importKind === "type") continue;
			if (!deniedSources.has(node.source.value)) continue;
			const mockVar = `__tss_deny_${mockCounter++}`;
			const replacements = [];
			replacements.push(t.importDeclaration([t.importDefaultSpecifier(t.identifier(mockVar))], t.stringLiteral(getMockModuleId(node.source.value))));
			for (const specifier of node.specifiers) if (t.isImportDefaultSpecifier(specifier) || t.isImportNamespaceSpecifier(specifier)) replacements.push(t.variableDeclaration("const", [t.variableDeclarator(t.identifier(specifier.local.name), t.identifier(mockVar))]));
			else if (t.isImportSpecifier(specifier)) {
				if (specifier.importKind === "type") continue;
				const importedName = t.isIdentifier(specifier.imported) ? specifier.imported.name : specifier.imported.value;
				replacements.push(t.variableDeclaration("const", [t.variableDeclarator(t.identifier(specifier.local.name), t.memberExpression(t.identifier(mockVar), t.identifier(importedName)))]));
			}
			ast.program.body.splice(i, 1, ...replacements);
			modified = true;
			continue;
		}
		if (t.isExportNamedDeclaration(node) && node.source) {
			if (node.exportKind === "type") continue;
			if (!deniedSources.has(node.source.value)) continue;
			const mockVar = `__tss_deny_${mockCounter++}`;
			const replacements = [];
			replacements.push(t.importDeclaration([t.importDefaultSpecifier(t.identifier(mockVar))], t.stringLiteral(getMockModuleId(node.source.value))));
			const exportSpecifiers = [];
			for (const specifier of node.specifiers) if (t.isExportSpecifier(specifier)) {
				if (specifier.exportKind === "type") continue;
				const localName = specifier.local.name;
				const exportedName = t.isIdentifier(specifier.exported) ? specifier.exported.name : specifier.exported.value;
				const internalVar = `__tss_reexport_${localName}`;
				replacements.push(t.variableDeclaration("const", [t.variableDeclarator(t.identifier(internalVar), t.memberExpression(t.identifier(mockVar), t.identifier(localName)))]));
				exportSpecifiers.push({
					localName: internalVar,
					exportedName
				});
			}
			if (exportSpecifiers.length > 0) replacements.push(t.exportNamedDeclaration(null, exportSpecifiers.map((s) => t.exportSpecifier(t.identifier(s.localName), t.identifier(s.exportedName)))));
			ast.program.body.splice(i, 1, ...replacements);
			modified = true;
			continue;
		}
		if (t.isExportAllDeclaration(node)) {
			if (node.exportKind === "type") continue;
			if (!deniedSources.has(node.source.value)) continue;
			ast.program.body.splice(i, 1);
			modified = true;
			continue;
		}
	}
	if (!modified) return void 0;
	const result = generateFromAst(ast, {
		sourceMaps: true,
		sourceFileName: id,
		filename: id
	});
	return {
		code: result.code,
		...result.map ? { map: result.map } : {}
	};
}
//#endregion
export { collectMockExportNamesBySource, collectNamedExports, isValidExportName, rewriteDeniedImports };
