// drizzle-orm/sqlite-core/query-builders/count.js
import { entityKind } from "../../entity.js";
import { applyMixins } from "../../utils.js";
import { SQL, sql } from "../../sql/sql.js";
import { QueryPromise } from "../../query-promise.js";

//#region src/sqlite-core/query-builders/count.ts
var SQLiteCountBuilder = class SQLiteCountBuilder extends SQL {
	static [entityKind] = "SQLiteCountBuilder";
	dialect;
	session;
	static buildCount(source, filters, parens) {
		const query = sql`select count(*) from ${source}${sql` where ${filters}`.if(filters)}`;
		return parens ? sql`(${query})` : query;
	}
	constructor(countConfig) {
		super(SQLiteCountBuilder.buildCount(countConfig.source, countConfig.filters, true).queryChunks);
		this.countConfig = countConfig;
		this.dialect = countConfig.dialect;
		this.session = countConfig.session;
		this.mapWith((e) => {
			if (typeof e === "number") return e;
			return Number(e ?? 0);
		});
	}
	executableSql;
	build() {
		if (!this.executableSql) {
			const { source, filters } = this.countConfig;
			this.executableSql = SQLiteCountBuilder.buildCount(source, filters);
		}
		return this.dialect.sqlToQuery(this.executableSql);
	}
	/** @internal */
	executeRaw(placeholderValues) {
		return this.session.prepareOneTimeQuery(this.build(), void 0, "get", (rows) => {
			const v = rows[0]?.[0];
			if (typeof v === "number") return v;
			return v ? Number(v) : 0;
		}).get(placeholderValues);
	}
	async execute(placeholderValues) {
		return await this.executeRaw(placeholderValues);
	}
};
var SQLiteSyncCountBuilder = class extends SQLiteCountBuilder {
	static [entityKind] = "SQLiteSyncCountBuilder";
	sync(placeholderValues) {
		return this.executeRaw(placeholderValues);
	}
};
applyMixins(SQLiteCountBuilder, [QueryPromise]);

//#endregion
export { SQLiteCountBuilder, SQLiteSyncCountBuilder };
