// drizzle-kit/prompts-CJ8ydMyf.js
const require_chunk = require('./chunk-D0pENJGy.js');
const require_views = require('./views-DGPo_q1l.js');

//#region src/cli/prompts.ts
var prompts_exports = /* @__PURE__ */ require_chunk.__exportAll({ resolver: () => resolver });
var import_hanji = require_views.require_hanji();
const resolver = (entity, defaultSchema = "public") => {
	return async (it) => {
		const { created, deleted } = it;
		if (created.length === 0 || deleted.length === 0) return {
			created,
			deleted,
			renamedOrMoved: []
		};
		const result = {
			created: [],
			deleted: [],
			renamedOrMoved: []
		};
		let index = 0;
		let leftMissing = [...deleted];
		do {
			const newItem = created[index];
			const { status, data } = await (0, import_hanji.render)(new require_views.ResolveSelect(newItem, [newItem, ...leftMissing.map((it) => {
				return {
					from: it,
					to: newItem
				};
			})], entity, defaultSchema));
			if (status === "aborted") {
				console.error("ERROR");
				process.exit(1);
			}
			if (require_views.isRenamePromptItem(data)) {
				const to = data.to;
				const fromEntity = `${newItem.schema ? newItem.schema !== defaultSchema ? `${newItem.schema}.` : "" : ""}${newItem.table ? `${newItem.table}.` : ""}${data.from.name}`;
				const toEntity = `${to.schema ? to.schema !== defaultSchema ? `${to.schema}.` : "" : ""}${to.table ? `${to.table}.` : ""}${to.name}`;
				console.log(`${require_views.chalk.yellow("~")} ${fromEntity} › ${toEntity} ${require_views.chalk.gray(`${entity} will be renamed/moved`)}`);
				result.renamedOrMoved.push(data);
				delete leftMissing[leftMissing.indexOf(data.from)];
				leftMissing = leftMissing.filter(Boolean);
			} else {
				console.log(`${require_views.chalk.green("+")} ${newItem.name} ${require_views.chalk.gray(`${entity} will be created`)}`);
				result.created.push(newItem);
			}
			index += 1;
		} while (index < created.length);
		console.log(require_views.chalk.gray(`--- all ${entity} conflicts resolved ---\n`));
		result.deleted.push(...leftMissing);
		return result;
	};
};

//#endregion
Object.defineProperty(exports, 'prompts_exports', {
  enumerable: true,
  get: function () {
    return prompts_exports;
  }
});
