// waddler/pg/pglite/session.js
import { SQLTemplate } from "../../sql-template.js";
import { types } from "@electric-sql/pglite";
import { WaddlerQueryError } from "../../errors/index.js";
class PGliteSQLTemplate extends SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
    this.rawQueryConfig = {
      rowMode: "object",
      parsers: {
        // [types.TIMESTAMP]: (value) => value,
        // [types.TIMESTAMPTZ]: (value) => value,
        [types.INTERVAL]: (value) => value
        // [types.DATE]: (value) => value,
      }
    };
    this.queryConfig = {
      rowMode: "array",
      parsers: {
        // [types.TIMESTAMP]: (value) => value,
        // [types.TIMESTAMPTZ]: (value) => value,
        [types.INTERVAL]: (value) => value
        // [types.DATE]: (value) => value,
      }
    };
  }
  rawQueryConfig;
  queryConfig;
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      const queryResult = await (this.options.rowMode === "array" ? this.client.query(query, params, this.queryConfig) : this.client.query(query, params, this.rawQueryConfig));
      ({ rows: finalRes, ...finalMetadata } = queryResult);
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  /**
   * For now, throws the Error.
   * Current implementation (a placeholder) will be replaced once PGlite acquires streaming support or when a suitable third-party solution is found.
   */
  // eslint-disable-next-line require-yield
  async *stream() {
    throw new Error("stream is not implemented for pglite yet.");
  }
}
export {
  PGliteSQLTemplate
};
