// waddler/pg/pglite/session.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var session_exports = {};
__export(session_exports, {
  PGliteSQLTemplate: () => PGliteSQLTemplate
});
module.exports = __toCommonJS(session_exports);
var import_sql_template = require("../../sql-template.cjs");
var import_pglite = require("@electric-sql/pglite");
var import_errors = require("../../errors/index.cjs");
class PGliteSQLTemplate extends import_sql_template.SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
    this.rawQueryConfig = {
      rowMode: "object",
      parsers: {
        // [types.TIMESTAMP]: (value) => value,
        // [types.TIMESTAMPTZ]: (value) => value,
        [import_pglite.types.INTERVAL]: (value) => value
        // [types.DATE]: (value) => value,
      }
    };
    this.queryConfig = {
      rowMode: "array",
      parsers: {
        // [types.TIMESTAMP]: (value) => value,
        // [types.TIMESTAMPTZ]: (value) => value,
        [import_pglite.types.INTERVAL]: (value) => value
        // [types.DATE]: (value) => value,
      }
    };
  }
  rawQueryConfig;
  queryConfig;
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      const queryResult = await (this.options.rowMode === "array" ? this.client.query(query, params, this.queryConfig) : this.client.query(query, params, this.rawQueryConfig));
      ({ rows: finalRes, ...finalMetadata } = queryResult);
    } catch (error) {
      throw new import_errors.WaddlerQueryError(query, params, error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  /**
   * For now, throws the Error.
   * Current implementation (a placeholder) will be replaced once PGlite acquires streaming support or when a suitable third-party solution is found.
   */
  // eslint-disable-next-line require-yield
  async *stream() {
    throw new Error("stream is not implemented for pglite yet.");
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PGliteSQLTemplate
});
