// waddler/pg/neon-http/session.js
import { types } from "@neondatabase/serverless";
import { WaddlerQueryError } from "../../errors/index.js";
import { SQLTemplate } from "../../sql-template.js";
const pgTypeConfig = {
  // @ts-expect-error
  getTypeParser: (typeId, format) => {
    if (typeId === types.builtins.INTERVAL) return (val) => val;
    if (typeId === 1187) return (val) => val;
    return types.getTypeParser(typeId, format);
  }
};
const rawQueryConfig = {
  arrayMode: false,
  fullResults: true,
  types: pgTypeConfig
};
const queryConfig = {
  arrayMode: true,
  fullResults: true,
  types: pgTypeConfig
};
class NeonHttpSQLTemplate extends SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
  }
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      const queryResult = await (this.options.rowMode === "array" ? this.client.query(query, params, queryConfig) : this.client.query(query, params, rawQueryConfig));
      ({ rows: finalRes, ...finalMetadata } = queryResult);
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  /**
   * For now, throws the Error.
   * Current implementation (a placeholder) will be replaced once @neondatabase/serverless acquires streaming support or when a suitable third-party solution is found.
   */
  // eslint-disable-next-line require-yield
  async *stream() {
    throw new Error("stream is not implemented for neon-http yet.");
  }
}
export {
  NeonHttpSQLTemplate
};
