// waddler/pg/neon-serverless/session.js
import {
  types
} from "@neondatabase/serverless";
import { WaddlerQueryError } from "../../errors/index.js";
import { SQLTemplate } from "../../sql-template.js";
const pgTypeConfig = {
  // @ts-expect-error
  getTypeParser: (typeId, format) => {
    if (typeId === types.builtins.INTERVAL) return (val) => val;
    if (typeId === 1187) return (val) => val;
    return types.getTypeParser(typeId, format);
  }
};
class NeonServerlessSQLTemplate extends SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
    const query = this.sqlWrapper.getQuery(this.dialect).sql;
    this.rawQueryConfig = {
      text: query,
      types: pgTypeConfig
    };
    this.queryConfig = {
      rowMode: "array",
      text: query,
      types: pgTypeConfig
    };
  }
  rawQueryConfig;
  queryConfig;
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      const queryResult = await (this.options.rowMode === "array" ? this.client.query(this.queryConfig, params) : await this.client.query(this.rawQueryConfig, params));
      ({ rows: finalRes, ...finalMetadata } = queryResult);
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  async *stream() {
    const queryStreamObj = this.configOptions?.extensions?.find((it) => it.name === "WaddlerPgQueryStream");
    if (!queryStreamObj) {
      throw new Error(
        'To use stream feature, you would need to provide queryStream() function to waddler extensions, example: waddler("", { extensions: [queryStream()] })'
      );
    }
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    const queryStream = new queryStreamObj.constructor(query, params, {
      types: this.queryConfig.types
      // rowMode: 'array',
    });
    this.logger.logQuery(query, params);
    try {
      const stream = this.client.query(queryStream);
      for await (const row of stream) {
        yield row;
      }
    } catch (error) {
      throw new WaddlerQueryError(query, params, error);
    }
  }
}
export {
  NeonServerlessSQLTemplate
};
