// waddler/pg/neon-serverless/session.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var session_exports = {};
__export(session_exports, {
  NeonServerlessSQLTemplate: () => NeonServerlessSQLTemplate
});
module.exports = __toCommonJS(session_exports);
var import_serverless = require("@neondatabase/serverless");
var import_errors = require("../../errors/index.cjs");
var import_sql_template = require("../../sql-template.cjs");
const pgTypeConfig = {
  // @ts-expect-error
  getTypeParser: (typeId, format) => {
    if (typeId === import_serverless.types.builtins.INTERVAL) return (val) => val;
    if (typeId === 1187) return (val) => val;
    return import_serverless.types.getTypeParser(typeId, format);
  }
};
class NeonServerlessSQLTemplate extends import_sql_template.SQLTemplate {
  constructor(sqlWrapper, client, dialect, configOptions, options = { rowMode: "object" }) {
    super(sqlWrapper, dialect, configOptions);
    this.sqlWrapper = sqlWrapper;
    this.client = client;
    this.options = options;
    const query = this.sqlWrapper.getQuery(this.dialect).sql;
    this.rawQueryConfig = {
      text: query,
      types: pgTypeConfig
    };
    this.queryConfig = {
      rowMode: "array",
      text: query,
      types: pgTypeConfig
    };
  }
  rawQueryConfig;
  queryConfig;
  async execute() {
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    let finalRes, finalMetadata;
    try {
      const queryResult = await (this.options.rowMode === "array" ? this.client.query(this.queryConfig, params) : await this.client.query(this.rawQueryConfig, params));
      ({ rows: finalRes, ...finalMetadata } = queryResult);
    } catch (error) {
      throw new import_errors.WaddlerQueryError(query, params, error);
    }
    this.logger.logQuery(query, params, finalMetadata);
    return finalRes;
  }
  async *stream() {
    const queryStreamObj = this.configOptions?.extensions?.find((it) => it.name === "WaddlerPgQueryStream");
    if (!queryStreamObj) {
      throw new Error(
        'To use stream feature, you would need to provide queryStream() function to waddler extensions, example: waddler("", { extensions: [queryStream()] })'
      );
    }
    const { sql: query, params } = this.sqlWrapper.getQuery(this.dialect);
    const queryStream = new queryStreamObj.constructor(query, params, {
      types: this.queryConfig.types
      // rowMode: 'array',
    });
    this.logger.logQuery(query, params);
    try {
      const stream = this.client.query(queryStream);
      for await (const row of stream) {
        yield row;
      }
    } catch (error) {
      throw new import_errors.WaddlerQueryError(query, params, error);
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  NeonServerlessSQLTemplate
});
