// drizzle-orm/effect-schema/schema.js
import { isWithEnum } from "../utils.js";
import { columnToSchema } from "./column.js";
import { is } from "../entity.js";
import { isTable } from "../table.js";
import { Column } from "../column.js";
import { getColumns } from "../utils.js";
import { SQL, isView } from "../sql/sql.js";
import { Schema } from "effect";

//#region src/effect-schema/schema.ts
function isOptional(schema) {
	if ((typeof schema !== "object" || schema === null) && typeof schema !== "function") return false;
	return Schema.isSchema(schema) && schema.ast?.context?.isOptional === true;
}
function isStructField(schema) {
	if (Schema.isSchema(schema)) return true;
	return isOptional(schema);
}
function handleColumns(columns, refinements, conditions) {
	const columnSchemas = {};
	for (const [key, selected] of Object.entries(columns)) {
		if (!is(selected, Column) && !is(selected, SQL) && !is(selected, SQL.Aliased) && typeof selected === "object") {
			columnSchemas[key] = handleColumns(isTable(selected) || isView(selected) ? getColumns(selected) : selected, refinements[key] ?? {}, conditions);
			continue;
		}
		const refinement = refinements[key];
		if (refinement !== void 0 && !(typeof refinement === "function" && !isStructField(refinement))) {
			columnSchemas[key] = refinement;
			continue;
		}
		const column = is(selected, Column) ? selected : void 0;
		const schema = column ? columnToSchema(column) : Schema.Any;
		const _refined = isStructField(refinement) || typeof refinement !== "function" ? schema : refinement(schema);
		const refined = isOptional(_refined) ? _refined.schema : _refined;
		if (conditions.never(column)) continue;
		else columnSchemas[key] = refined;
		if (column) {
			if (conditions.nullable(column)) columnSchemas[key] = Schema.NullOr(columnSchemas[key]);
			if (conditions.optional(column)) columnSchemas[key] = Schema.optional(Schema.UndefinedOr(columnSchemas[key]));
		}
	}
	return Schema.Struct(columnSchemas);
}
function handleEnum(enum_) {
	return Schema.Literals(enum_.enumValues);
}
const selectConditions = {
	never: () => false,
	optional: () => false,
	nullable: (column) => !column.notNull
};
const insertConditions = {
	never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
	optional: (column) => !column.notNull || column.notNull && column.hasDefault,
	nullable: (column) => !column.notNull
};
const updateConditions = {
	never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
	optional: () => true,
	nullable: (column) => !column.notNull
};
const createSelectSchema = (entity, refine) => {
	if (isWithEnum(entity)) return handleEnum(entity);
	return handleColumns(getColumns(entity), refine ?? {}, selectConditions);
};
const createInsertSchema = (entity, refine) => {
	return handleColumns(getColumns(entity), refine ?? {}, insertConditions);
};
const createUpdateSchema = (entity, refine) => {
	return handleColumns(getColumns(entity), refine ?? {}, updateConditions);
};

//#endregion
export { createInsertSchema, createSelectSchema, createUpdateSchema };
