// drizzle-orm/valibot/schema.js
import { isWithEnum } from "../utils.js";
import { columnToSchema, mapEnumValues } from "./column.js";
import { is } from "../entity.js";
import { isTable } from "../table.js";
import { Column } from "../column.js";
import { getColumns } from "../utils.js";
import { SQL, isView } from "../sql/sql.js";
import * as v from "valibot";

//#region src/valibot/schema.ts
function handleColumns(columns, refinements, conditions) {
	const columnSchemas = {};
	for (const [key, selected] of Object.entries(columns)) {
		if (!is(selected, Column) && !is(selected, SQL) && !is(selected, SQL.Aliased) && typeof selected === "object") {
			columnSchemas[key] = handleColumns(isTable(selected) || isView(selected) ? getColumns(selected) : selected, refinements[key] ?? {}, conditions);
			continue;
		}
		const refinement = refinements[key];
		if (refinement !== void 0 && typeof refinement !== "function") {
			columnSchemas[key] = refinement;
			continue;
		}
		const column = is(selected, Column) ? selected : void 0;
		const schema = column ? columnToSchema(column) : v.any();
		const refined = typeof refinement === "function" ? refinement(schema) : schema;
		if (conditions.never(column)) continue;
		else columnSchemas[key] = refined;
		if (column) {
			if (conditions.nullable(column)) columnSchemas[key] = v.nullable(columnSchemas[key]);
			if (conditions.optional(column)) columnSchemas[key] = v.optional(columnSchemas[key]);
		}
	}
	return v.object(columnSchemas);
}
const createSelectSchema = (entity, refine) => {
	if (isWithEnum(entity)) return v.enum(mapEnumValues(entity.enumValues));
	return handleColumns(getColumns(entity), refine ?? {}, {
		never: () => false,
		optional: () => false,
		nullable: (column) => !column.notNull
	});
};
const createInsertSchema = (entity, refine) => {
	return handleColumns(getColumns(entity), refine ?? {}, {
		never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
		optional: (column) => !column.notNull || column.notNull && column.hasDefault,
		nullable: (column) => !column.notNull
	});
};
const createUpdateSchema = (entity, refine) => {
	return handleColumns(getColumns(entity), refine ?? {}, {
		never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
		optional: () => true,
		nullable: (column) => !column.notNull
	});
};

//#endregion
export { createInsertSchema, createSelectSchema, createUpdateSchema };
