// drizzle-orm/valibot/schema.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
const require_utils = require('../utils.cjs');
const require_valibot_column = require('./column.cjs');
let __entity_ts = require("../entity.cjs");
let __table_ts = require("../table.cjs");
let __column_ts = require("../column.cjs");
let __utils_ts = require("../utils.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");
let valibot = require("valibot");
valibot = require_runtime.__toESM(valibot);

//#region src/valibot/schema.ts
function handleColumns(columns, refinements, conditions) {
	const columnSchemas = {};
	for (const [key, selected] of Object.entries(columns)) {
		if (!(0, __entity_ts.is)(selected, __column_ts.Column) && !(0, __entity_ts.is)(selected, __sql_sql_ts.SQL) && !(0, __entity_ts.is)(selected, __sql_sql_ts.SQL.Aliased) && typeof selected === "object") {
			columnSchemas[key] = handleColumns((0, __table_ts.isTable)(selected) || (0, __sql_sql_ts.isView)(selected) ? (0, __utils_ts.getColumns)(selected) : selected, refinements[key] ?? {}, conditions);
			continue;
		}
		const refinement = refinements[key];
		if (refinement !== void 0 && typeof refinement !== "function") {
			columnSchemas[key] = refinement;
			continue;
		}
		const column = (0, __entity_ts.is)(selected, __column_ts.Column) ? selected : void 0;
		const schema = column ? require_valibot_column.columnToSchema(column) : valibot.any();
		const refined = typeof refinement === "function" ? refinement(schema) : schema;
		if (conditions.never(column)) continue;
		else columnSchemas[key] = refined;
		if (column) {
			if (conditions.nullable(column)) columnSchemas[key] = valibot.nullable(columnSchemas[key]);
			if (conditions.optional(column)) columnSchemas[key] = valibot.optional(columnSchemas[key]);
		}
	}
	return valibot.object(columnSchemas);
}
const createSelectSchema = (entity, refine) => {
	if (require_utils.isWithEnum(entity)) return valibot.enum(require_valibot_column.mapEnumValues(entity.enumValues));
	return handleColumns((0, __utils_ts.getColumns)(entity), refine ?? {}, {
		never: () => false,
		optional: () => false,
		nullable: (column) => !column.notNull
	});
};
const createInsertSchema = (entity, refine) => {
	return handleColumns((0, __utils_ts.getColumns)(entity), refine ?? {}, {
		never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
		optional: (column) => !column.notNull || column.notNull && column.hasDefault,
		nullable: (column) => !column.notNull
	});
};
const createUpdateSchema = (entity, refine) => {
	return handleColumns((0, __utils_ts.getColumns)(entity), refine ?? {}, {
		never: (column) => column?.generated?.type === "always" || column?.generatedIdentity?.type === "always" || "identity" in (column ?? {}) && typeof column?.identity !== "undefined",
		optional: () => true,
		nullable: (column) => !column.notNull
	});
};

//#endregion
exports.createInsertSchema = createInsertSchema;
exports.createSelectSchema = createSelectSchema;
exports.createUpdateSchema = createUpdateSchema;
